---
title:  "INTEGRATED TROPHOBLAST DATA PREPARATION"
author: "Matthew J Shannon"
date:   "17/10/2022"
output: html_document
---

# Load in all dependencies

```{r, include = FALSE}
#devtools::install_github  ('chris-mcginnis-ucsf/DoubletFinder')
#devtools::install_github  ('satijalab/seurat-wrappers')
#remotes::install_github   ("mojaveazure/seurat-disk")
#devtools::install_github  ("velocyto-team/velocyto.R")
#BiocManager::install      ("pcaMethods")
#BiocManager::install      ('MAST')
#BiocManager::install      ('org.Hs.eg.db')
#BiocManager::install      (version = "3.12")ikAOQ
#BiocManager::install      ('Seurat')
#BiocManager::install      ('readxl')
#BiocManager::install      ('modes')
#BiocManager::install      ('pheatmap')
#BiocManager::install      ('limma')
#BiocManager::install      ('clustree')
#BiocManager::install      ('clusterProfiler')
#BiocManager::install      ('EnhancedVolcano')
#install.packages          ('fields')
#install.packages          ('plotly')
#install.packages          ("VennDiagram")
 library                   (Seurat)
 library                   (readr)
 library                   (tidyverse)
 library                   (Matrix)
 library                   (dplyr)
 library                   (ggplot2)
 library                   (cowplot)
 library                   (tibble)
 library                   (readxl)
 library                   (sctransform)
 library                   (fields)
 library                   (KernSmooth)
 library                   (ROCR)
 library                   (parallel)
 library                   (reshape2)
 library                   (pheatmap)
 library                   (DoubletFinder)
 library                   (limma)
 library                   (SeuratWrappers)
 library                   (SeuratDisk)
 library                   (plotly)
 library                   (clustree)
 library                   (velocyto.R)
 library                   (MAST)
 library                   (EnhancedVolcano)
 library                   (clusterProfiler)
 library                   (AnnotationDbi)
 library                   (org.Hs.eg.db)
 library                   (VennDiagram)
```

## Read in cell cycle marker genes
 
```{r}
s.genes   <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
```


# Placental Villi Dataset Prepation, Quality Control, and Sample Integration

## Read in the Beristain single-cell Placenta data

### Patient 277

```{r}
# Read in the data for placental villi sample one
sample1 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Beristain/P1")

# Read in the respective velocity data
          ldat1            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/1_12weeks.loom")
colnames (ldat1$spliced)   <- paste        (substring (colnames (ldat1$spliced),   15, 30), sep = "")
colnames (ldat1$spliced)   <- paste                   (colnames (ldat1$spliced),   "-1",    sep = "")
colnames (ldat1$unspliced) <- paste        (substring (colnames (ldat1$unspliced), 15, 30), sep = "")
colnames (ldat1$unspliced) <- paste                   (colnames (ldat1$unspliced), "-1",    sep = "")
colnames (ldat1$ambiguous) <- paste        (substring (colnames (ldat1$ambiguous), 15, 30), sep = "")
colnames (ldat1$ambiguous) <- paste                   (colnames (ldat1$ambiguous), "-1",    sep = "")

# Convert to a seurat object
sample1 <- CreateSeuratObject (counts = sample1, min.cells = 1, min.features = 1, project = "B_P277")

for (i in names (x = ldat1)) { sample1[[i]] <- CreateAssayObject(counts = ldat1[[i]]) }

# Add the sample metadata
sample1 <- AddMetaData (sample1, metadata = "MPI",      col.name = "Model"      ) # Trophoblast model
sample1 <- AddMetaData (sample1, metadata = "B_P277",   col.name = "ID"         ) # Sample identifier
sample1 <- AddMetaData (sample1, metadata = "B_P277",   col.name = "rawID"      ) # Raw sample identifier (for samples with multiple files)
sample1 <- AddMetaData (sample1, metadata = "1",        col.name = "nSex"       ) # Numeric biological sex of sample
sample1 <- AddMetaData (sample1, metadata = "Male",     col.name = "Sex"        ) # Biological sex of sample
sample1 <- AddMetaData (sample1, metadata = "6.4",      col.name = "rawGA"      ) # Raw gestational age of sample
sample1 <- AddMetaData (sample1, metadata = "Early",    col.name = "binGA"      ) # Bin gestational age of sample
sample1 <- AddMetaData (sample1, metadata = "1",        col.name = "nGA"        ) # Numeric gestational age of sample
sample1 <- AddMetaData (sample1, metadata = "Early",    col.name = "GA"         ) # Gestational age of sample
sample1 <- AddMetaData (sample1, metadata = "Placenta", col.name = "Tissue"     ) # Sample tissue type (either Placenta, Decidua, or PB)
sample1 <- AddMetaData (sample1, metadata = "1",        col.name = "Batch"      ) # Sequencing batch
sample1 <- AddMetaData (sample1, metadata = "1",        col.name = "Source"     ) # Data source (Beristain or Vento-Tormo)
sample1 <- AddMetaData (sample1, metadata = "in vivo",  col.name = "Type"       ) # Data type (in vivo or in vitro)
sample1 <- AddMetaData (sample1, metadata = "24",       col.name = "MA"         ) # Maternal Age (years)
sample1 <- AddMetaData (sample1, metadata = "20.90",    col.name = "BMI"        ) # Patient Body Mass Index
sample1 <- AddMetaData (sample1, metadata = "No",       col.name = "Smoke"      ) # Did the patient smoke during pregnancy?
sample1 <- AddMetaData (sample1, metadata = "No",       col.name = "Diabetic"   ) # Is the patient living with diabetes (T1D or T2D)?
sample1 <- AddMetaData (sample1, metadata = "None",     col.name = "Medications") # Did the patient consume anti-inflammatory and/or anti-hypertensive medications during pregnancy?
sample1 <- AddMetaData (sample1, metadata = "All",      col.name = "Gate"       ) # Were cells enriched prior to sequencing?
sample1 <- AddMetaData (sample1, metadata = "First",    col.name = "Trimester"  ) # Sample trimester

# Normalize the sample data and add cell cycle scores
sample1 <- NormalizeData    (sample1)
sample1 <- CellCycleScoring (sample1, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample1)
dim   (sample1)
sample1@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample1 [["percent.mt"]]   <- PercentageFeatureSet (sample1, pattern = "^MT-")
ribo.genes                 <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample1), value = TRUE);
percent.ribo               <- Matrix::colSums (x = GetAssayData (sample1, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample1, slot   = 'counts')              );
sample1 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample1, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample1, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample1, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample1@meta.data
```

```{r}
# Filter sample one
fltsample1 <- subset (sample1, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample1, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample1, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample1, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample1
fltsample1
```

```{r}
normsample1 <- NormalizeData        ( fltsample1)
normsample1 <- ScaleData            (normsample1)
normsample1 <- FindVariableFeatures (normsample1, selection.method = "vst", nfeatures = 2000)
normsample1 <- RunPCA               (normsample1)
normsample1 <- RunUMAP              (normsample1, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample1 <- paramSweep_v3    (normsample1, PCs = 1:50, sct = FALSE)
sweep.stats_sample1    <- summarizeSweep   (sweep.res.list_sample1,  GT  = FALSE)
bcmvn_sample1          <- find.pK          (sweep.stats_sample1)
annotations1           <- normsample1@meta.data$ClusteringResults
homotypic.prop1        <- modelHomotypic   (annotations1)
nExp_poi1              <- round            (0.025 * length (normsample1@meta.data$ID))  ## Assuming 2.5% doublet formation rate
nExp_poi.adj1          <- round            (nExp_poi1 *    (1 - homotypic.prop1)     )
normsample1            <- doubletFinder_v3 (normsample1,
                                            PCs        = 1:50,
                                            pN         = 0.25,
                                            pK         = 0.09,
                                            nExp       = nExp_poi1,
                                            reuse.pANN = FALSE,
                                            sct        = FALSE)
normsample1            <- doubletFinder_v3 (normsample1,
                                            PCs        = 1:50,
                                            pN         = 0.25,
                                            pK         = 0.09,
                                            nExp       = nExp_poi.adj1,
                                            reuse.pANN = "pANN_0.25_0.09_77",
                                            sct        = FALSE)
```

```{r}
normsample1@meta.data
DimPlot (normsample1, group.by = "DF.classifications_0.25_0.09_77")
table   (normsample1$DF.classifications_0.25_0.09_77)
```

```{r}
# Subset to remove identified cell doublets
fltsample1 <- subset (normsample1, subset = DF.classifications_0.25_0.09_77 == 'Singlet')
fltsample1
table (fltsample1$DF.classifications_0.25_0.09_77)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample1)
rm (sweep.res.list_sample1)
rm (sweep.stats_sample1)
rm (bcmvn_sample1)
rm (annotations1)
rm (homotypic.prop1)
rm (nExp_poi1)
rm (nExp_poi.adj1)
rm (normsample1)
rm (ldat1)
```

### Patient 89

```{r}
# Read in the data for placental villi sample two
sample2 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Beristain/P2")

# Read in the respective velocity data
          ldat2            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/2_12weeks.loom")
colnames (ldat2$spliced)   <- paste        (substring (colnames (ldat2$spliced),   15, 30), sep = "")
colnames (ldat2$spliced)   <- paste                   (colnames (ldat2$spliced),   "-1",    sep = "")
colnames (ldat2$unspliced) <- paste        (substring (colnames (ldat2$unspliced), 15, 30), sep = "")
colnames (ldat2$unspliced) <- paste                   (colnames (ldat2$unspliced), "-1",    sep = "")
colnames (ldat2$ambiguous) <- paste        (substring (colnames (ldat2$ambiguous), 15, 30), sep = "")
colnames (ldat2$ambiguous) <- paste                   (colnames (ldat2$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample2 <- CreateSeuratObject (counts = sample2, min.cells = 1, min.features = 1, project = "B_P89")

for (i in names (x = ldat2)) { sample2[[i]] <- CreateAssayObject(counts = ldat2[[i]]) }

# Add the sample metadata
sample2 <- AddMetaData (sample2, metadata = "MPI",      col.name = "Model"      )
sample2 <- AddMetaData (sample2, metadata = "B_P89",    col.name = "ID"         )
sample2 <- AddMetaData (sample2, metadata = "B_P89",    col.name = "rawID"      )
sample2 <- AddMetaData (sample2, metadata = "0",        col.name = "nSex"       )
sample2 <- AddMetaData (sample2, metadata = "Female",   col.name = "Sex"        )
sample2 <- AddMetaData (sample2, metadata = "6.4",      col.name = "rawGA"      )
sample2 <- AddMetaData (sample2, metadata = "Early",    col.name = "binGA"      )
sample2 <- AddMetaData (sample2, metadata = "1",        col.name = "nGA"        )
sample2 <- AddMetaData (sample2, metadata = "Early",    col.name = "GA"         )
sample2 <- AddMetaData (sample2, metadata = "Placenta", col.name = "Tissue"     )
sample2 <- AddMetaData (sample2, metadata = "1",        col.name = "Batch"      )
sample2 <- AddMetaData (sample2, metadata = "1",        col.name = "Source"     )
sample2 <- AddMetaData (sample2, metadata = "in vivo",  col.name = "Type"       )
sample2 <- AddMetaData (sample2, metadata = "22",       col.name = "MA"         )
sample2 <- AddMetaData (sample2, metadata = "24.21",    col.name = "BMI"        )
sample2 <- AddMetaData (sample2, metadata = "No",       col.name = "Smoke"      )
sample2 <- AddMetaData (sample2, metadata = "No",       col.name = "Diabetic"   )
sample2 <- AddMetaData (sample2, metadata = "None",     col.name = "Medications")
sample2 <- AddMetaData (sample2, metadata = "All",      col.name = "Gate"       )
sample2 <- AddMetaData (sample2, metadata = "First",    col.name = "Trimester"  )

# Normalize the sample data and add cell cycle scores
sample2 <- NormalizeData    (sample2)
sample2 <- CellCycleScoring (sample2, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample2)
dim   (sample2)
sample2@meta.data
```

```{r} 
# Add sample quality control metrics (% MT and % RP data)
sample2 [["percent.mt"]]   <- PercentageFeatureSet (sample2, pattern = "^MT-")
ribo.genes                 <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample2), value = TRUE);
percent.ribo               <- Matrix::colSums (x = GetAssayData (sample2, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample2, slot   = 'counts')              );
sample2 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample2, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample2, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample2, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample2@meta.data
```

```{r}
# Filter sample two
fltsample2 <- subset (sample2, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample2, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample2, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample2, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample two before and after filtering to observe the effect of filtering on this sample
   sample2
fltsample2
```

```{r}
normsample2 <- NormalizeData        ( fltsample2)
normsample2 <- ScaleData            (normsample2)
normsample2 <- FindVariableFeatures (normsample2, selection.method = "vst", nfeatures = 2000)
normsample2 <- RunPCA               (normsample2)
normsample2 <- RunUMAP              (normsample2, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample2 <- paramSweep_v3    (normsample2, PCs = 1:50, sct = FALSE)
sweep.stats_sample2    <- summarizeSweep   (sweep.res.list_sample2,  GT  = FALSE)
bcmvn_sample2          <- find.pK          (sweep.stats_sample2)
annotations2           <- normsample2@meta.data$ClusteringResults
homotypic.prop2        <- modelHomotypic   (annotations2)
nExp_poi2              <- round            (0.031 * length (normsample2@meta.data$ID))  ## Assuming 3.1% doublet formation rate
nExp_poi.adj2          <- round            (nExp_poi2 *    (1 - homotypic.prop2)     )
normsample2            <- doubletFinder_v3 (normsample2,
                                            PCs        = 1:50,
                                            pN         = 0.25,
                                            pK         = 0.09,
                                            nExp       = nExp_poi2,
                                            reuse.pANN = FALSE,
                                            sct        = FALSE)
normsample2            <- doubletFinder_v3 (normsample2,
                                            PCs        = 1:50,
                                            pN         = 0.25,
                                            pK         = 0.09,
                                            nExp       = nExp_poi.adj2,
                                            reuse.pANN = "pANN_0.25_0.09_97",
                                            sct        = FALSE)
```

```{r}
normsample2@meta.data
DimPlot (normsample2, group.by = "DF.classifications_0.25_0.09_97")
table   (normsample2$DF.classifications_0.25_0.09_97)
```

```{r}
# Subset to remove identified cell doublets
fltsample2 <- subset (normsample2, subset = DF.classifications_0.25_0.09_97 == 'Singlet')
fltsample2
table (fltsample2$DF.classifications_0.25_0.09_97)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample2)
rm (sweep.res.list_sample2)
rm (sweep.stats_sample2)
rm (bcmvn_sample2)
rm (annotations2)
rm (homotypic.prop2)
rm (nExp_poi2)
rm (nExp_poi.adj2)
rm (normsample2)
rm (ldat2)
```

### Patient 21

```{r}
# Read in the data for placental villi sample three
sample3 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Beristain/P3")

# Read in the respective velocity data
          ldat3            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/3_6weeks.loom")
colnames (ldat3$spliced)   <- paste        (substring (colnames (ldat3$spliced),   15, 30), sep = "")
colnames (ldat3$spliced)   <- paste                   (colnames (ldat3$spliced),   "-1",    sep = "")
colnames (ldat3$unspliced) <- paste        (substring (colnames (ldat3$unspliced), 15, 30), sep = "")
colnames (ldat3$unspliced) <- paste                   (colnames (ldat3$unspliced), "-1",    sep = "")
colnames (ldat3$ambiguous) <- paste        (substring (colnames (ldat3$ambiguous), 15, 30), sep = "")
colnames (ldat3$ambiguous) <- paste                   (colnames (ldat3$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample3 <- CreateSeuratObject (counts = sample3, min.cells = 1, min.features = 1, project = "B_P21")

for (i in names (x = ldat3)) { sample3[[i]] <- CreateAssayObject(counts = ldat3[[i]]) }

# Add the sample metadata
sample3 <- AddMetaData (sample3, metadata = "MPI",      col.name = "Model"      )
sample3 <- AddMetaData (sample3, metadata = "B_P21",    col.name = "ID"         )
sample3 <- AddMetaData (sample3, metadata = "B_P21",    col.name = "rawID"      )
sample3 <- AddMetaData (sample3, metadata = "0",        col.name = "nSex"       )
sample3 <- AddMetaData (sample3, metadata = "Female",   col.name = "Sex"        )
sample3 <- AddMetaData (sample3, metadata = "11.2",     col.name = "rawGA"      )
sample3 <- AddMetaData (sample3, metadata = "Late",     col.name = "binGA"      )
sample3 <- AddMetaData (sample3, metadata = "2",        col.name = "nGA"        )
sample3 <- AddMetaData (sample3, metadata = "Late",     col.name = "GA"         )
sample3 <- AddMetaData (sample3, metadata = "Placenta", col.name = "Tissue"     )
sample3 <- AddMetaData (sample3, metadata = "1",        col.name = "Batch"      )
sample3 <- AddMetaData (sample3, metadata = "1",        col.name = "Source"     )
sample3 <- AddMetaData (sample3, metadata = "in vivo",  col.name = "Type"       )
sample3 <- AddMetaData (sample3, metadata = "30",       col.name = "MA"         )
sample3 <- AddMetaData (sample3, metadata = "21.16",    col.name = "BMI"        )
sample3 <- AddMetaData (sample3, metadata = "No",       col.name = "Smoke"      )
sample3 <- AddMetaData (sample3, metadata = "No",       col.name = "Diabetic"   )
sample3 <- AddMetaData (sample3, metadata = "None",     col.name = "Medications")
sample3 <- AddMetaData (sample3, metadata = "All",      col.name = "Gate"       )
sample3 <- AddMetaData (sample3, metadata = "First",    col.name = "Trimester"  )

# Normalize the sample data and add cell cycle scores
sample3 <- NormalizeData    (sample3)
sample3 <- CellCycleScoring (sample3,s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample3)
dim   (sample3)
sample3@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample3 [["percent.mt"]]   <- PercentageFeatureSet (sample3, pattern = "^MT-")
ribo.genes                 <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample3), value = TRUE);
percent.ribo               <- Matrix::colSums (x = GetAssayData (sample3, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample3, slot   = 'counts')              );
sample3 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample3, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample3, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample3, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample3@meta.data
```

```{r}
# Filter sample three
fltsample3 <- subset (sample3, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample3, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample3, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample3, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample three before and after filtering to observe the effect of filtering on this sample
   sample3
fltsample3
```

```{r}
normsample3 <- NormalizeData        ( fltsample3)
normsample3 <- ScaleData            (normsample3)
normsample3 <- FindVariableFeatures (normsample3, selection.method = "vst", nfeatures = 2000)
normsample3 <- RunPCA               (normsample3)
normsample3 <- RunUMAP              (normsample3, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample3 <- paramSweep_v3    (normsample3, PCs = 1:50, sct = FALSE)
sweep.stats_sample3    <- summarizeSweep   (sweep.res.list_sample3,  GT  = FALSE)
bcmvn_sample3          <- find.pK          (sweep.stats_sample3)
annotations3           <- normsample3@meta.data$ClusteringResults
homotypic.prop3        <- modelHomotypic   (annotations3)
nExp_poi3              <- round            (0.029 * length (normsample3@meta.data$ID) )  ## Assuming 2.9% doublet formation rate
nExp_poi.adj3          <- round            (nExp_poi3 * (1 - homotypic.prop3)        )
normsample3            <- doubletFinder_v3 (normsample3,
                                            PCs        = 1:50,
                                            pN         = 0.25,
                                            pK         = 0.09,
                                            nExp       = nExp_poi3,
                                            reuse.pANN = FALSE,
                                            sct        = FALSE)
normsample3            <- doubletFinder_v3 (normsample3,
                                            PCs        = 1:50, 
                                            pN         = 0.25,
                                            pK         = 0.09,
                                            nExp       = nExp_poi.adj3,
                                            reuse.pANN = "pANN_0.25_0.09_52",
                                            sct        = FALSE)
```

```{r}
normsample3@meta.data
DimPlot (normsample3, group.by = "DF.classifications_0.25_0.09_52")
table   (normsample3$DF.classifications_0.25_0.09_52)
```

```{r}
# Subset to remove identified cell doublets
fltsample3 <- subset (normsample3, subset = DF.classifications_0.25_0.09_52 == 'Singlet')
fltsample3
table (fltsample3$DF.classifications_0.25_0.09_52)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample3)
rm (sweep.res.list_sample3)
rm (sweep.stats_sample3)
rm (bcmvn_sample3)
rm (annotations3)
rm (homotypic.prop3)
rm (nExp_poi3)
rm (nExp_poi.adj3)
rm (normsample3)
rm (ldat3)
```

### Patient 130

```{r}
# Read in the data for placental villi sample four
sample4 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Beristain/P4")

# Read in the respective velocity data
          ldat4            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/4_6weeks.loom")
colnames (ldat4$spliced)   <- paste        (substring (colnames (ldat4$spliced),   15, 30), sep = "")
colnames (ldat4$spliced)   <- paste                   (colnames (ldat4$spliced),   "-1",    sep = "")
colnames (ldat4$unspliced) <- paste        (substring (colnames (ldat4$unspliced), 15, 30), sep = "")
colnames (ldat4$unspliced) <- paste                   (colnames (ldat4$unspliced), "-1",    sep = "")
colnames (ldat4$ambiguous) <- paste        (substring (colnames (ldat4$ambiguous), 15, 30), sep = "")
colnames (ldat4$ambiguous) <- paste                   (colnames (ldat4$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample4 <- CreateSeuratObject (counts = sample4, min.cells = 1, min.features = 1, project = "B_P130")

for (i in names (x = ldat4)) { sample4[[i]] <- CreateAssayObject(counts = ldat4[[i]]) }

# Add the sample metadata
sample4 <- AddMetaData (sample4, metadata = "MPI",      col.name = "Model"      )
sample4 <- AddMetaData (sample4, metadata = "B_P130",   col.name = "ID"         )
sample4 <- AddMetaData (sample4, metadata = "B_P130",   col.name = "rawID"      )
sample4 <- AddMetaData (sample4, metadata = "1",        col.name = "nSex"       )
sample4 <- AddMetaData (sample4, metadata = "Male",     col.name = "Sex"        )
sample4 <- AddMetaData (sample4, metadata = "11.4",     col.name = "rawGA"      )
sample4 <- AddMetaData (sample4, metadata = "Late",     col.name = "binGA"      )
sample4 <- AddMetaData (sample4, metadata = "2",        col.name = "nGA"        )
sample4 <- AddMetaData (sample4, metadata = "Late",     col.name = "GA"         )
sample4 <- AddMetaData (sample4, metadata = "Placenta", col.name = "Tissue"     )
sample4 <- AddMetaData (sample4, metadata = "1",        col.name = "Batch"      )
sample4 <- AddMetaData (sample4, metadata = "1",        col.name = "Source"     )
sample4 <- AddMetaData (sample4, metadata = "in vivo",  col.name = "Type"       )
sample4 <- AddMetaData (sample4, metadata = "32",       col.name = "MA"         )
sample4 <- AddMetaData (sample4, metadata = "22.86",    col.name = "BMI"        )
sample4 <- AddMetaData (sample4, metadata = "No",       col.name = "Smoke"      )
sample4 <- AddMetaData (sample4, metadata = "No",       col.name = "Diabetic"   )
sample4 <- AddMetaData (sample4, metadata = "None",     col.name = "Medications")
sample4 <- AddMetaData (sample4, metadata = "All",      col.name = "Gate"       )
sample4 <- AddMetaData (sample4, metadata = "First",    col.name = "Trimester"  )

# Normalize the sample data and add cell cycle scores
sample4 <- NormalizeData    (sample4)
sample4 <- CellCycleScoring (sample4, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample4)
dim   (sample4)
sample4@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample4 [["percent.mt"]]   <- PercentageFeatureSet (sample4, pattern = "^MT-")
ribo.genes                 <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample4), value = TRUE);
percent.ribo               <- Matrix::colSums (x = GetAssayData (sample4, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample4, slot   = 'counts')              );
sample4 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample4, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample4, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample4, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample4@meta.data
```

```{r}
# Filter sample four
fltsample4 <- subset (sample4, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot and decide threshold for % mtRNA
VlnPlot (fltsample4, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample4, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample4, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample four before and after filtering to observe the effect of filtering on this sample
   sample4
fltsample4
```

```{r}
normsample4 <- NormalizeData        ( fltsample4)
normsample4 <- ScaleData            (normsample4)
normsample4 <- FindVariableFeatures (normsample4, selection.method = "vst", nfeatures = 2000)
normsample4 <- RunPCA               (normsample4)
normsample4 <- RunUMAP              (normsample4, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample4 <- paramSweep_v3    (normsample4, PCs = 1:50, sct = FALSE)
sweep.stats_sample4    <- summarizeSweep   (sweep.res.list_sample4,  GT  = FALSE)
bcmvn_sample4          <- find.pK          (sweep.stats_sample4)
annotations4           <- normsample4@meta.data$ClusteringResults
homotypic.prop4        <- modelHomotypic   (annotations4)
nExp_poi4              <- round            (0.031 * length (normsample4@meta.data$ID))  ## Assuming 3.1% doublet formation rate
nExp_poi.adj4          <- round            (nExp_poi4 *    (1 - homotypic.prop4)     )
normsample4            <- doubletFinder_v3 (normsample4,
                                            PCs        = 1:50,
                                            pN         = 0.25,
                                            pK         = 0.09,
                                            nExp       = nExp_poi4,
                                            reuse.pANN = FALSE,
                                            sct        = FALSE)
normsample4            <- doubletFinder_v3 (normsample4,
                                            PCs        = 1:50,
                                            pN         = 0.25,
                                            pK         = 0.09,
                                            nExp       = nExp_poi.adj4,
                                            reuse.pANN = "pANN_0.25_0.09_101",
                                            sct        = FALSE)
```

```{r}
normsample4@meta.data
DimPlot (normsample4, group.by = "DF.classifications_0.25_0.09_101")
table   (normsample4$DF.classifications_0.25_0.09_101)
```

```{r}
# Subset to remove identified cell doublets
fltsample4 <- subset (normsample4, subset = DF.classifications_0.25_0.09_101 == 'Singlet')
fltsample4
table (fltsample4$DF.classifications_0.25_0.09_101)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample4)
rm (sweep.res.list_sample4)
rm (sweep.stats_sample4)
rm (bcmvn_sample4)
rm (annotations4)
rm (homotypic.prop4)
rm (nExp_poi4)
rm (nExp_poi.adj4)
rm (normsample4)
rm (ldat4)
```

### Patient 244

```{r}
# Read in the data for placental villi sample five
sample5 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Beristain/P5")

# Read in the respective velocity data
          ldat5            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/sc244_6w.loom")
colnames (ldat5$spliced)   <- paste        (substring (colnames (ldat5$spliced),   18, 33), sep = "")
colnames (ldat5$spliced)   <- paste                   (colnames (ldat5$spliced),   "-1",    sep = "")
colnames (ldat5$unspliced) <- paste        (substring (colnames (ldat5$unspliced), 18, 33), sep = "")
colnames (ldat5$unspliced) <- paste                   (colnames (ldat5$unspliced), "-1",    sep = "")
colnames (ldat5$ambiguous) <- paste        (substring (colnames (ldat5$ambiguous), 18, 33), sep = "")
colnames (ldat5$ambiguous) <- paste                   (colnames (ldat5$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample5 <- CreateSeuratObject (counts = sample5, min.cells = 1, min.features = 1, project = "B_P244")

for (i in names (x = ldat5)) { sample5[[i]] <- CreateAssayObject(counts = ldat5[[i]]) }

# Add the sample metadata
sample5 <- AddMetaData (sample5, metadata = "MPI",      col.name = "Model"      )
sample5 <- AddMetaData (sample5, metadata = "B_P244",   col.name = "ID"         )
sample5 <- AddMetaData (sample5, metadata = "B_P244",   col.name = "rawID"      )
sample5 <- AddMetaData (sample5, metadata = "0",        col.name = "nSex"       )
sample5 <- AddMetaData (sample5, metadata = "Female",   col.name = "Sex"        )
sample5 <- AddMetaData (sample5, metadata = "6.2",      col.name = "rawGA"      )
sample5 <- AddMetaData (sample5, metadata = "Early",    col.name = "binGA"      )
sample5 <- AddMetaData (sample5, metadata = "1",        col.name = "nGA"        )
sample5 <- AddMetaData (sample5, metadata = "Early",    col.name = "GA"         )
sample5 <- AddMetaData (sample5, metadata = "Placenta", col.name = "Tissue"     )
sample5 <- AddMetaData (sample5, metadata = "2",        col.name = "Batch"      )
sample5 <- AddMetaData (sample5, metadata = "1",        col.name = "Source"     )
sample5 <- AddMetaData (sample5, metadata = "in vivo",  col.name = "Type"       )
sample5 <- AddMetaData (sample5, metadata = "23",       col.name = "MA"         )
sample5 <- AddMetaData (sample5, metadata = "23.53",    col.name = "BMI"        )
sample5 <- AddMetaData (sample5, metadata = "No",       col.name = "Smoke"      )
sample5 <- AddMetaData (sample5, metadata = "No",       col.name = "Diabetic"   )
sample5 <- AddMetaData (sample5, metadata = "None",     col.name = "Medications")
sample5 <- AddMetaData (sample5, metadata = "All",      col.name = "Gate"       )
sample5 <- AddMetaData (sample5, metadata = "First",    col.name = "Trimester"  )

# Normalize the sample data and add cell cycle scores
sample5 <- NormalizeData    (sample5)
sample5 <- CellCycleScoring (sample5, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample5)
dim   (sample5)
sample5@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample5 [["percent.mt"]]   <- PercentageFeatureSet (sample5, pattern = "^MT-")
ribo.genes                 <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample5), value = TRUE);
percent.ribo               <- Matrix::colSums (x = GetAssayData (sample5, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample5, slot   = 'counts')              );
sample5 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample5, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample5, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample5, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample5@meta.data
```

```{r}
# Filter sample five
fltsample5 <- subset (sample5, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample5, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample5, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample5, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample five before and after filtering to observe the effect of filtering on this sample
   sample5
fltsample5
```

```{r}
normsample5 <- NormalizeData        ( fltsample5)
normsample5 <- ScaleData            (normsample5)
normsample5 <- FindVariableFeatures (normsample5, selection.method = "vst", nfeatures = 2000)
normsample5 <- RunPCA               (normsample5)
normsample5 <- RunUMAP              (normsample5, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample5 <- paramSweep_v3    (normsample5, PCs = 1:50, sct = FALSE)
sweep.stats_sample5    <- summarizeSweep   (sweep.res.list_sample5,  GT  = FALSE)
bcmvn_sample5          <- find.pK          (sweep.stats_sample5)
annotations5           <- normsample5@meta.data$ClusteringResults
homotypic.prop5        <- modelHomotypic   (annotations5)
nExp_poi5              <- round            (0.007 * length (normsample5@meta.data$ID))  ## Assuming 0.7% doublet formation rate
nExp_poi.adj5          <- round            (nExp_poi5 *    (1 - homotypic.prop5)     )
normsample5            <- doubletFinder_v3 (normsample5,
                                            PCs        = 1:50,
                                            pN         = 0.25,
                                            pK         = 0.09,
                                            nExp       = nExp_poi5,
                                            reuse.pANN = FALSE,
                                            sct        = FALSE)
normsample5            <- doubletFinder_v3 (normsample5,
                                            PCs        = 1:50,
                                            pN         = 0.25,
                                            pK         = 0.09,
                                            nExp       = nExp_poi.adj5,
                                            reuse.pANN = "pANN_0.25_0.09_4",
                                            sct        = FALSE)
```

```{r}
normsample5@meta.data
DimPlot (normsample5, group.by = "DF.classifications_0.25_0.09_4")
table   (normsample5$DF.classifications_0.25_0.09_4)
```

```{r}
# Subset to remove identified cell doublets
fltsample5 <- subset (normsample5, subset = DF.classifications_0.25_0.09_4 == 'Singlet')
fltsample5
table (fltsample5$DF.classifications_0.25_0.09_4)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample5)
rm (sweep.res.list_sample5)
rm (sweep.stats_sample5)
rm (bcmvn_sample5)
rm (annotations5)
rm (homotypic.prop5)
rm (nExp_poi5)
rm (nExp_poi.adj5)
rm (normsample5)
rm (ldat5)
```

### Patient 295

```{r}
# Read in the data for placental villi sample six
sample6 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Beristain/P6")

# Read in the respective velocity data
          ldat6            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/sc295_6w.loom")
colnames (ldat6$spliced)   <- paste        (substring (colnames (ldat6$spliced),   18, 33), sep = "")
colnames (ldat6$spliced)   <- paste                   (colnames (ldat6$spliced),   "-1",    sep = "")
colnames (ldat6$unspliced) <- paste        (substring (colnames (ldat6$unspliced), 18, 33), sep = "")
colnames (ldat6$unspliced) <- paste                   (colnames (ldat6$unspliced), "-1",    sep = "")
colnames (ldat6$ambiguous) <- paste        (substring (colnames (ldat6$ambiguous), 18, 33), sep = "")
colnames (ldat6$ambiguous) <- paste                   (colnames (ldat6$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample6 <- CreateSeuratObject (counts = sample6, min.cells = 1, min.features = 1, project = "B_P295")

for (i in names (x = ldat6)) { sample6[[i]] <- CreateAssayObject(counts = ldat6[[i]]) }

# Add the sample metadata
sample6 <- AddMetaData (sample6, metadata = "MPI",      col.name = "Model"      )
sample6 <- AddMetaData (sample6, metadata = "B_P295",   col.name = "ID"         )
sample6 <- AddMetaData (sample6, metadata = "B_P295",   col.name = "rawID"      )
sample6 <- AddMetaData (sample6, metadata = "1",        col.name = "nSex"       )
sample6 <- AddMetaData (sample6, metadata = "Male",     col.name = "Sex"        )
sample6 <- AddMetaData (sample6, metadata = "6.5",      col.name = "rawGA"      )
sample6 <- AddMetaData (sample6, metadata = "Early",    col.name = "binGA"      )
sample6 <- AddMetaData (sample6, metadata = "1",        col.name = "nGA"        )
sample6 <- AddMetaData (sample6, metadata = "Early",    col.name = "GA"         )
sample6 <- AddMetaData (sample6, metadata = "Placenta", col.name = "Tissue"     )
sample6 <- AddMetaData (sample6, metadata = "2",        col.name = "Batch"      )
sample6 <- AddMetaData (sample6, metadata = "1",        col.name = "Source"     )
sample6 <- AddMetaData (sample6, metadata = "in vivo",  col.name = "Type"       )
sample6 <- AddMetaData (sample6, metadata = "30",       col.name = "MA"         )
sample6 <- AddMetaData (sample6, metadata = "23.51",    col.name = "BMI"        )
sample6 <- AddMetaData (sample6, metadata = "No",       col.name = "Smoke"      )
sample6 <- AddMetaData (sample6, metadata = "No",       col.name = "Diabetic"   )
sample6 <- AddMetaData (sample6, metadata = "None",     col.name = "Medications")
sample6 <- AddMetaData (sample6, metadata = "All",      col.name = "Gate"       )
sample6 <- AddMetaData (sample6, metadata = "First",    col.name = "Trimester"  )

# Normalize the sample data and add cell cycle scores
sample6 <- NormalizeData    (sample6)
sample6 <- CellCycleScoring (sample6, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample6)
dim   (sample6)
sample6@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample6 [["percent.mt"]]   <- PercentageFeatureSet (sample6, pattern = "^MT-")
ribo.genes                 <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample6), value = TRUE);
percent.ribo               <- Matrix::colSums (x = GetAssayData (sample6, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample6, slot   = 'counts')              );
sample6 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample6, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample6, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample6, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample6@meta.data
```

```{r}
# Filter sample six
fltsample6 <- subset (sample6, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample6, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample6, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample6, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample six before and after filtering to observe the effect of filtering on this sample
   sample6
fltsample6
```

```{r}
normsample6 <- NormalizeData        ( fltsample6)
normsample6 <- ScaleData            (normsample6)
normsample6 <- FindVariableFeatures (normsample6, selection.method = "vst", nfeatures = 2000)
normsample6 <- RunPCA               (normsample6)
normsample6 <- RunUMAP              (normsample6, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample6 <- paramSweep_v3    (normsample6, PCs = 1:50, sct = FALSE)
sweep.stats_sample6    <- summarizeSweep   (sweep.res.list_sample6,  GT  = FALSE)
bcmvn_sample6          <- find.pK          (sweep.stats_sample6)
annotations6           <- normsample6@meta.data$ClusteringResults
homotypic.prop6        <- modelHomotypic   (annotations6)
nExp_poi6              <- round            (0.012 * length (normsample6@meta.data$ID))  ## Assuming 1.2% doublet formation rate
nExp_poi.adj6          <- round            (nExp_poi6 *    (1 - homotypic.prop6)     )
normsample6            <- doubletFinder_v3 (normsample6,
                                            PCs        = 1:50,
                                            pN         = 0.25,
                                            pK         = 0.09,
                                            nExp       = nExp_poi6,
                                            reuse.pANN = FALSE,
                                            sct        = FALSE)
normsample6            <- doubletFinder_v3 (normsample6,
                                            PCs        = 1:50,
                                            pN         = 0.25,
                                            pK         = 0.09,
                                            nExp       = nExp_poi.adj6,
                                            reuse.pANN = "pANN_0.25_0.09_10",
                                            sct        = FALSE)
```

```{r}
normsample6@meta.data
DimPlot (normsample6, group.by = "DF.classifications_0.25_0.09_10")
table   (normsample6$DF.classifications_0.25_0.09_10)
```

```{r}
# Subset to remove identified cell doublets
fltsample6 <- subset (normsample6, subset = DF.classifications_0.25_0.09_10 == 'Singlet')
fltsample6
table (fltsample6$DF.classifications_0.25_0.09_10)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample6)
rm (sweep.res.list_sample6)
rm (sweep.stats_sample6)
rm (bcmvn_sample6)
rm (annotations6)
rm (homotypic.prop6)
rm (nExp_poi6)
rm (nExp_poi.adj6)
rm (normsample6)
rm (ldat6)
```

### Patient 899

```{r}
# Read in the data for placental villi sample seven
sample7 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Beristain/P7")

# Read in the respective velocity data
          ldat7            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/sc899_12w.loom")
colnames (ldat7$spliced)   <- paste        (substring (colnames (ldat7$spliced),   18, 33), sep = "")
colnames (ldat7$spliced)   <- paste                   (colnames (ldat7$spliced),   "-1",    sep = "")
colnames (ldat7$unspliced) <- paste        (substring (colnames (ldat7$unspliced), 18, 33), sep = "")
colnames (ldat7$unspliced) <- paste                   (colnames (ldat7$unspliced), "-1",    sep = "")
colnames (ldat7$ambiguous) <- paste        (substring (colnames (ldat7$ambiguous), 18, 33), sep = "")
colnames (ldat7$ambiguous) <- paste                   (colnames (ldat7$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample7 <- CreateSeuratObject (counts = sample7, min.cells = 1, min.features = 1, project = "B_P899")

for (i in names (x = ldat7)) { sample7[[i]] <- CreateAssayObject(counts = ldat7[[i]]) }

# Add the sample metadata
sample7 <- AddMetaData (sample7, metadata = "MPI",      col.name = "Model"      )
sample7 <- AddMetaData (sample7, metadata = "B_P899",   col.name = "ID"         )
sample7 <- AddMetaData (sample7, metadata = "B_P899",   col.name = "rawID"      )
sample7 <- AddMetaData (sample7, metadata = "0",        col.name = "nSex"       )
sample7 <- AddMetaData (sample7, metadata = "Female",   col.name = "Sex"        )
sample7 <- AddMetaData (sample7, metadata = "11.4",     col.name = "rawGA"      )
sample7 <- AddMetaData (sample7, metadata = "Late",     col.name = "binGA"      )
sample7 <- AddMetaData (sample7, metadata = "2",        col.name = "nGA"        )
sample7 <- AddMetaData (sample7, metadata = "Late",     col.name = "GA"         )
sample7 <- AddMetaData (sample7, metadata = "Placenta", col.name = "Tissue"     )
sample7 <- AddMetaData (sample7, metadata = "2",        col.name = "Batch"      )
sample7 <- AddMetaData (sample7, metadata = "1",        col.name = "Source"     )
sample7 <- AddMetaData (sample7, metadata = "in vivo",  col.name = "Type"       )
sample7 <- AddMetaData (sample7, metadata = "20",       col.name = "MA"         )
sample7 <- AddMetaData (sample7, metadata = "22.08",    col.name = "BMI"        )
sample7 <- AddMetaData (sample7, metadata = "No",       col.name = "Smoke"      )
sample7 <- AddMetaData (sample7, metadata = "No",       col.name = "Diabetic"   )
sample7 <- AddMetaData (sample7, metadata = "None",     col.name = "Medications")
sample7 <- AddMetaData (sample7, metadata = "All",      col.name = "Gate"       )
sample7 <- AddMetaData (sample7, metadata = "First",    col.name = "Trimester"  )

# Normalize the sample data and add cell cycle scores
sample7 <- NormalizeData    (sample7)
sample7 <- CellCycleScoring (sample7, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample7)
dim   (sample7)
sample7@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample7 [["percent.mt"]]   <- PercentageFeatureSet (sample7, pattern = "^MT-")
ribo.genes                 <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample7), value = TRUE);
percent.ribo               <- Matrix::colSums (x = GetAssayData (sample7, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample7, slot   = 'counts')              );
sample7 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample7, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample7, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample7, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample7@meta.data
```

```{r}
# Filter sample seven
fltsample7 <- subset (sample7, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample7, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample7, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample7, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample seven before and after filtering to observe the effect of filtering on this sample
   sample7
fltsample7
```

```{r}
normsample7 <- NormalizeData        ( fltsample7)
normsample7 <- ScaleData            (normsample7)
normsample7 <- FindVariableFeatures (normsample7, selection.method = "vst", nfeatures = 2000)
normsample7 <- RunPCA               (normsample7)
normsample7 <- RunUMAP              (normsample7, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample7 <- paramSweep_v3    (normsample7, PCs = 1:50, sct = FALSE)
sweep.stats_sample7    <- summarizeSweep   (sweep.res.list_sample7,  GT  = FALSE)
bcmvn_sample7          <- find.pK          (sweep.stats_sample7)
annotations7           <- normsample7@meta.data$ClusteringResults
homotypic.prop7        <- modelHomotypic   (annotations7)
nExp_poi7              <- round            (0.023 * length (normsample7@meta.data$ID))  ## Assuming 2.3% doublet formation rate
nExp_poi.adj7          <- round            (nExp_poi7 *    (1 - homotypic.prop7)     )
normsample7            <- doubletFinder_v3 (normsample7,
                                            PCs        = 1:50,
                                            pN         = 0.25,
                                            pK         = 0.09,
                                            nExp       = nExp_poi7,
                                            reuse.pANN = FALSE,
                                            sct        = FALSE)
normsample7            <- doubletFinder_v3 (normsample7,
                                            PCs        = 1:50,
                                            pN         = 0.25,
                                            pK         = 0.09,
                                            nExp       = nExp_poi.adj7,
                                            reuse.pANN = "pANN_0.25_0.09_38",
                                            sct        = FALSE)
```

```{r}
normsample7@meta.data
DimPlot (normsample7, group.by = "DF.classifications_0.25_0.09_38")
table   (normsample7$DF.classifications_0.25_0.09_38)
```

```{r}
# Subset to remove identified cell doublets
fltsample7 <- subset (normsample7, subset = DF.classifications_0.25_0.09_38 == 'Singlet')
fltsample7
table (fltsample7$DF.classifications_0.25_0.09_38)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample7)
rm (sweep.res.list_sample7)
rm (sweep.stats_sample7)
rm (bcmvn_sample7)
rm (annotations7)
rm (homotypic.prop7)
rm (nExp_poi7)
rm (nExp_poi.adj7)
rm (normsample7)
rm (ldat7)
```

## Read in the Vento-Tormo single-cell Placenta data

### Patient D9

```{r}
# Patient D9

# Read in the data for placental villi sample eight
sample8 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Vento-Tormo/Vento-Tormo_P9/All")

# Read in the respective velocity data
          ldat8            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/FCA7196226.loom")
colnames (ldat8$spliced)   <- paste        (substring (colnames (ldat8$spliced),   12, 27), sep = "")
colnames (ldat8$spliced)   <- paste                   (colnames (ldat8$spliced),   "-1",    sep = "")
colnames (ldat8$unspliced) <- paste        (substring (colnames (ldat8$unspliced), 12, 27), sep = "")
colnames (ldat8$unspliced) <- paste                   (colnames (ldat8$unspliced), "-1",    sep = "")
colnames (ldat8$ambiguous) <- paste        (substring (colnames (ldat8$ambiguous), 12, 27), sep = "")
colnames (ldat8$ambiguous) <- paste                   (colnames (ldat8$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample8 <- CreateSeuratObject (counts = sample8, min.cells = 1, min.features = 1, project = "VT_P9")

for (i in names (x = ldat8)) { sample8[[i]] <- CreateAssayObject(counts = ldat8[[i]]) }

# Add the sample metadata
sample8 <- AddMetaData (sample8, metadata = "MPI",      col.name = "Model"    )
sample8 <- AddMetaData (sample8, metadata = "VT_P9",    col.name = "ID"       )
sample8 <- AddMetaData (sample8, metadata = "VT_P9",    col.name = "rawID"    )
sample8 <- AddMetaData (sample8, metadata = "1",        col.name = "nSex"     )
sample8 <- AddMetaData (sample8, metadata = "Male",     col.name = "Sex"      )
sample8 <- AddMetaData (sample8, metadata = "12.0",     col.name = "rawGA"    )
sample8 <- AddMetaData (sample8, metadata = "Late",     col.name = "binGA"    )
sample8 <- AddMetaData (sample8, metadata = "2",        col.name = "nGA"      )
sample8 <- AddMetaData (sample8, metadata = "Late",     col.name = "GA"       )
sample8 <- AddMetaData (sample8, metadata = "Placenta", col.name = "Tissue"   )
sample8 <- AddMetaData (sample8, metadata = "3",        col.name = "Batch"    )
sample8 <- AddMetaData (sample8, metadata = "2",        col.name = "Source"   )
sample8 <- AddMetaData (sample8, metadata = "in vivo",  col.name = "Type"     )
sample8 <- AddMetaData (sample8, metadata = "Third",    col.name = "Pregnancy")
sample8 <- AddMetaData (sample8, metadata = "All",      col.name = "Gate"     )
sample8 <- AddMetaData (sample8, metadata = "First",    col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample8 <- NormalizeData    (sample8)
sample8 <- CellCycleScoring (sample8, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample8)
dim   (sample8)
sample8@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample8 [["percent.mt"]]   <- PercentageFeatureSet (sample8, pattern = "^MT-")
ribo.genes                 <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample8), value = TRUE);
percent.ribo               <- Matrix::colSums (x = GetAssayData (sample8, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample8, slot   = 'counts')              );
sample8 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample8, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample8, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample8, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample8@meta.data
```

```{r}
# Filter sample eight
fltsample8 <- subset (sample8, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample8, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample8, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample8, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample8
fltsample8
```

```{r}
normsample8 <- NormalizeData        ( fltsample8)
normsample8 <- ScaleData            (normsample8)
normsample8 <- FindVariableFeatures (normsample8, selection.method = "vst", nfeatures = 2000)
normsample8 <- RunPCA               (normsample8)
normsample8 <- RunUMAP              (normsample8, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample8 <- paramSweep_v3    (normsample8, PCs = 1:50, sct = FALSE)
sweep.stats_sample8    <- summarizeSweep   (sweep.res.list_sample8,  GT  = FALSE)
bcmvn_sample8          <- find.pK          (sweep.stats_sample8)
annotations8           <- normsample8@meta.data$ClusteringResults
homotypic.prop8        <- modelHomotypic   (annotations8)
nExp_poi8              <- round            (0.039 * length (normsample8@meta.data$ID))  ## Assuming 3.9% doublet formation rate
nExp_poi.adj8          <- round            (nExp_poi8 *    (1 - homotypic.prop8)     )
normsample8            <- doubletFinder_v3 (normsample8,
                                            PCs        = 1:50,
                                            pN         = 0.25,
                                            pK         = 0.09,
                                            nExp       = nExp_poi8,
                                            reuse.pANN = FALSE,
                                            sct        = FALSE)
normsample8            <- doubletFinder_v3 (normsample8,
                                            PCs        = 1:50,
                                            pN         = 0.25,
                                            pK         = 0.09,
                                            nExp       = nExp_poi.adj8,
                                            reuse.pANN = "pANN_0.25_0.09_116",
                                            sct        = FALSE)
```

```{r}
normsample8@meta.data
DimPlot (normsample8, group.by = "DF.classifications_0.25_0.09_116")
table   (normsample8$DF.classifications_0.25_0.09_116)
```

```{r}
# Subset to remove identified cell doublets
fltsample8 <- subset (normsample8, subset = DF.classifications_0.25_0.09_116 == 'Singlet')
fltsample8
table (fltsample8$DF.classifications_0.25_0.09_116)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample8)
rm (sweep.res.list_sample8)
rm (sweep.stats_sample8)
rm (bcmvn_sample8)
rm (annotations8)
rm (homotypic.prop8)
rm (nExp_poi8)
rm (nExp_poi.adj8)
rm (normsample8)
rm (ldat8)
```

### Patient D10

```{r}
# Patient D10

# Read in the data for placental villi sample nine (file 1)
sample9 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Vento-Tormo/Vento-Tormo_P10/All.1")

# Read in the respective velocity data
          ldat9            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/FCA7474064.loom")
colnames (ldat9$spliced)   <- paste        (substring (colnames (ldat9$spliced),   12, 27), sep = "")
colnames (ldat9$spliced)   <- paste                   (colnames (ldat9$spliced),   "-1",    sep = "")
colnames (ldat9$unspliced) <- paste        (substring (colnames (ldat9$unspliced), 12, 27), sep = "")
colnames (ldat9$unspliced) <- paste                   (colnames (ldat9$unspliced), "-1",    sep = "")
colnames (ldat9$ambiguous) <- paste        (substring (colnames (ldat9$ambiguous), 12, 27), sep = "")
colnames (ldat9$ambiguous) <- paste                   (colnames (ldat9$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample9 <- CreateSeuratObject (counts = sample9, min.cells = 1, min.features = 1, project = "VT_P10.1")

for (i in names (x = ldat9)) { sample9[[i]] <- CreateAssayObject(counts = ldat9[[i]]) }

# Add the sample metadata
sample9 <- AddMetaData (sample9, metadata = "MPI",      col.name = "Model"    )
sample9 <- AddMetaData (sample9, metadata = "VT_P10",   col.name = "ID"       )
sample9 <- AddMetaData (sample9, metadata = "VT_P10.1", col.name = "rawID"    )
sample9 <- AddMetaData (sample9, metadata = "1",        col.name = "nSex"     )
sample9 <- AddMetaData (sample9, metadata = "Male",     col.name = "Sex"      )
sample9 <- AddMetaData (sample9, metadata = "9.6",      col.name = "rawGA"    )
sample9 <- AddMetaData (sample9, metadata = "Mid",      col.name = "binGA"    )
sample9 <- AddMetaData (sample9, metadata = "2",        col.name = "nGA"      )
sample9 <- AddMetaData (sample9, metadata = "Late",     col.name = "GA"       )
sample9 <- AddMetaData (sample9, metadata = "Placenta", col.name = "Tissue"   )
sample9 <- AddMetaData (sample9, metadata = "3",        col.name = "Batch"    )
sample9 <- AddMetaData (sample9, metadata = "2",        col.name = "Source"   )
sample9 <- AddMetaData (sample9, metadata = "in vivo",  col.name = "Type"     )
sample9 <- AddMetaData (sample9, metadata = "Second",   col.name = "Pregnancy")
sample9 <- AddMetaData (sample9, metadata = "All",      col.name = "Gate"     )
sample9 <- AddMetaData (sample9, metadata = "First",    col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample9 <- NormalizeData    (sample9)
sample9 <- CellCycleScoring (sample9, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample9)
dim   (sample9)
sample9@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample9 [["percent.mt"]]   <- PercentageFeatureSet (sample9, pattern = "^MT-")
ribo.genes                 <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample9), value = TRUE);
percent.ribo               <- Matrix::colSums (x = GetAssayData (sample9, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample9, slot   = 'counts')              );
sample9 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample9, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample9, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample9, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample9@meta.data
```

```{r}
# Filter sample nine (file 1)
fltsample9 <- subset (sample9, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample9, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample9, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample9, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample9
fltsample9
```

```{r}
normsample9 <- NormalizeData        ( fltsample9)
normsample9 <- ScaleData            (normsample9)
normsample9 <- FindVariableFeatures (normsample9, selection.method = "vst", nfeatures = 2000)
normsample9 <- RunPCA               (normsample9)
normsample9 <- RunUMAP              (normsample9, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample9 <- paramSweep_v3    (normsample9, PCs = 1:50, sct = FALSE)
sweep.stats_sample9    <- summarizeSweep   (sweep.res.list_sample9,  GT  = FALSE)
bcmvn_sample9          <- find.pK          (sweep.stats_sample9)
annotations9           <- normsample9@meta.data$ClusteringResults
homotypic.prop9        <- modelHomotypic   (annotations9)
nExp_poi9              <- round            (0.037 * length (normsample9@meta.data$ID) )  ## Assuming 7.5% doublet formation rate
nExp_poi.adj9          <- round            (nExp_poi9 *    (1 - homotypic.prop9)     )
normsample9            <- doubletFinder_v3 (normsample9,
                                            PCs        = 1:50,
                                            pN         = 0.25,
                                            pK         = 0.09,
                                            nExp       = nExp_poi9,
                                            reuse.pANN = FALSE,
                                            sct        = FALSE)
normsample9            <- doubletFinder_v3 (normsample9,
                                            PCs        = 1:50,
                                            pN         = 0.25,
                                            pK         = 0.09,
                                            nExp       = nExp_poi.adj9,
                                            reuse.pANN = "pANN_0.25_0.09_163",
                                            sct        = FALSE)
```

```{r}
normsample9@meta.data
DimPlot (normsample9, group.by = "DF.classifications_0.25_0.09_163")
table   (normsample9$DF.classifications_0.25_0.09_163)
```

```{r}
# Subset to remove identified cell doublets
fltsample9 <- subset (normsample9, subset = DF.classifications_0.25_0.09_163 == 'Singlet')
fltsample9
table (fltsample9$DF.classifications_0.25_0.09_163)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample9)
rm (sweep.res.list_sample9)
rm (sweep.stats_sample9)
rm (bcmvn_sample9)
rm (annotations9)
rm (homotypic.prop9)
rm (nExp_poi9)
rm (nExp_poi.adj9)
rm (normsample9)
rm (ldat9)
```

```{r}
# Patient D10

# Read in the data for placental villi sample nine (file 2)
sample10 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Vento-Tormo/Vento-Tormo_P10/All.2")

# Read in the respective velocity data
          ldat10            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/FCA7474065.loom")
colnames (ldat10$spliced)   <- paste        (substring (colnames (ldat10$spliced),   12, 27), sep = "")
colnames (ldat10$spliced)   <- paste                   (colnames (ldat10$spliced),   "-1",    sep = "")
colnames (ldat10$unspliced) <- paste        (substring (colnames (ldat10$unspliced), 12, 27), sep = "")
colnames (ldat10$unspliced) <- paste                   (colnames (ldat10$unspliced), "-1",    sep = "")
colnames (ldat10$ambiguous) <- paste        (substring (colnames (ldat10$ambiguous), 12, 27), sep = "")
colnames (ldat10$ambiguous) <- paste                   (colnames (ldat10$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample10 <- CreateSeuratObject (counts = sample10, min.cells = 1, min.features = 1, project = "VT_P10.2")

for (i in names (x = ldat10)) { sample10[[i]] <- CreateAssayObject(counts = ldat10[[i]]) }

# Add the sample metadata
sample10 <- AddMetaData (sample10, metadata = "MPI",      col.name = "Model"    )
sample10 <- AddMetaData (sample10, metadata = "VT_P10",   col.name = "ID"       )
sample10 <- AddMetaData (sample10, metadata = "VT_P10.2", col.name = "rawID"    )
sample10 <- AddMetaData (sample10, metadata = "1",        col.name = "nSex"     )
sample10 <- AddMetaData (sample10, metadata = "Male",     col.name = "Sex"      )
sample10 <- AddMetaData (sample10, metadata = "9.6",      col.name = "rawGA"    )
sample10 <- AddMetaData (sample10, metadata = "Mid",      col.name = "binGA"    )
sample10 <- AddMetaData (sample10, metadata = "2",        col.name = "nGA"      )
sample10 <- AddMetaData (sample10, metadata = "Late",     col.name = "GA"       )
sample10 <- AddMetaData (sample10, metadata = "Placenta", col.name = "Tissue"   )
sample10 <- AddMetaData (sample10, metadata = "3",        col.name = "Batch"    )
sample10 <- AddMetaData (sample10, metadata = "2",        col.name = "Source"   )
sample10 <- AddMetaData (sample10, metadata = "in vivo",  col.name = "Type"     )
sample10 <- AddMetaData (sample10, metadata = "Second",   col.name = "Pregnancy")
sample10 <- AddMetaData (sample10, metadata = "All",      col.name = "Gate"     )
sample10 <- AddMetaData (sample10, metadata = "First",    col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample10 <- NormalizeData    (sample10)
sample10 <- CellCycleScoring (sample10, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample10)
dim   (sample10)
sample10@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample10 [["percent.mt"]]   <- PercentageFeatureSet (sample10, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample10), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample10, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample10, slot   = 'counts')              );
sample10 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample10, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample10, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample10, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample10@meta.data
```

```{r}
# Filter sample nine (file 2)
fltsample10 <- subset (sample10, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample10, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample10, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample10, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample10
fltsample10
```

```{r}
normsample10 <- NormalizeData        ( fltsample10)
normsample10 <- ScaleData            (normsample10)
normsample10 <- FindVariableFeatures (normsample10, selection.method = "vst", nfeatures = 2000)
normsample10 <- RunPCA               (normsample10)
normsample10 <- RunUMAP              (normsample10, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample10 <- paramSweep_v3    (normsample10, PCs = 1:50, sct = FALSE)
sweep.stats_sample10    <- summarizeSweep   (sweep.res.list_sample10,  GT  = FALSE)
bcmvn_sample10          <- find.pK          (sweep.stats_sample10)
annotations10           <- normsample10@meta.data$ClusteringResults
homotypic.prop10        <- modelHomotypic   (annotations10)
nExp_poi10              <- round            (0.036 * length (normsample10@meta.data$ID))  ## Assuming 3.6% doublet formation rate
nExp_poi.adj10          <- round            (nExp_poi10 *   (1 - homotypic.prop10)     )
normsample10            <- doubletFinder_v3 (normsample10,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi10,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample10            <- doubletFinder_v3 (normsample10,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj10,
                                             reuse.pANN = "pANN_0.25_0.09_156",
                                             sct        = FALSE)
```

```{r}
normsample10@meta.data
DimPlot (normsample10, group.by = "DF.classifications_0.25_0.09_156")
table   (normsample10$DF.classifications_0.25_0.09_156)
```

```{r}
# Subset to remove identified cell doublets
fltsample10 <- subset (normsample10, subset = DF.classifications_0.25_0.09_156 == 'Singlet')
fltsample10
table (fltsample10$DF.classifications_0.25_0.09_156)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample10)
rm (sweep.res.list_sample10)
rm (sweep.stats_sample10)
rm (bcmvn_sample10)
rm (annotations10)
rm (homotypic.prop10)
rm (nExp_poi10)
rm (nExp_poi.adj10)
rm (normsample10)
rm (ldat10)
```

```{r}
# Patient D10

# Read in the data for placental villi sample ten (file 3)
sample11 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Vento-Tormo/Vento-Tormo_P10/EPCAM+")

# Read in the respective velocity data
          ldat11            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/FCA7474066.loom")
colnames (ldat11$spliced)   <- paste        (substring (colnames (ldat11$spliced),   12, 27), sep = "")
colnames (ldat11$spliced)   <- paste                   (colnames (ldat11$spliced),   "-1",    sep = "")
colnames (ldat11$unspliced) <- paste        (substring (colnames (ldat11$unspliced), 12, 27), sep = "")
colnames (ldat11$unspliced) <- paste                   (colnames (ldat11$unspliced), "-1",    sep = "")
colnames (ldat11$ambiguous) <- paste        (substring (colnames (ldat11$ambiguous), 12, 27), sep = "")
colnames (ldat11$ambiguous) <- paste                   (colnames (ldat11$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample11 <- CreateSeuratObject (counts = sample11, min.cells = 1, min.features = 1, project = "VT_P10.3")

for (i in names (x = ldat11)) { sample11[[i]] <- CreateAssayObject(counts = ldat11[[i]]) }

# Add the sample metadata
sample11 <- AddMetaData (sample11, metadata = "MPI",      col.name = "Model"    )
sample11 <- AddMetaData (sample11, metadata = "VT_P10",   col.name = "ID"       )
sample11 <- AddMetaData (sample11, metadata = "VT_P10.3", col.name = "rawID"    )
sample11 <- AddMetaData (sample11, metadata = "1",        col.name = "nSex"     )
sample11 <- AddMetaData (sample11, metadata = "Male",     col.name = "Sex"      )
sample11 <- AddMetaData (sample11, metadata = "9.6",      col.name = "rawGA"    )
sample11 <- AddMetaData (sample11, metadata = "Mid",      col.name = "binGA"    )
sample11 <- AddMetaData (sample11, metadata = "2",        col.name = "nGA"      )
sample11 <- AddMetaData (sample11, metadata = "Late",     col.name = "GA"       )
sample11 <- AddMetaData (sample11, metadata = "Placenta", col.name = "Tissue"   )
sample11 <- AddMetaData (sample11, metadata = "3",        col.name = "Batch"    )
sample11 <- AddMetaData (sample11, metadata = "2",        col.name = "Source"   )
sample11 <- AddMetaData (sample11, metadata = "in vivo",  col.name = "Type"     )
sample11 <- AddMetaData (sample11, metadata = "Second",   col.name = "Pregnancy")
sample11 <- AddMetaData (sample11, metadata = "EPCAM+",   col.name = "Gate"     )
sample11 <- AddMetaData (sample11, metadata = "First",    col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample11 <- NormalizeData    (sample11)
sample11 <- CellCycleScoring (sample11, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample11)
dim   (sample11)
sample11@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample11 [["percent.mt"]]   <- PercentageFeatureSet (sample11, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample11), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample11, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample11, slot   = 'counts')              );
sample11 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample11, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample11, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample11, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample11@meta.data
```

```{r}
# Filter sample nine (file 2)
fltsample11 <- subset (sample11, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample11, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample11, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample11, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample11
fltsample11
```

```{r}
normsample11 <- NormalizeData        ( fltsample11)
normsample11 <- ScaleData            (normsample11)
normsample11 <- FindVariableFeatures (normsample11, selection.method = "vst", nfeatures = 2000)
normsample11 <- RunPCA               (normsample11)
normsample11 <- RunUMAP              (normsample11, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample11 <- paramSweep_v3    (normsample11, PCs = 1:50, sct = FALSE)
sweep.stats_sample11    <- summarizeSweep   (sweep.res.list_sample11,  GT  = FALSE)
bcmvn_sample11          <- find.pK          (sweep.stats_sample11)
annotations11           <- normsample11@meta.data$ClusteringResults
homotypic.prop11        <- modelHomotypic   (annotations11)
nExp_poi11              <- round            (0.039 * length (normsample11@meta.data$ID))  ## Assuming 3.9% doublet formation rate
nExp_poi.adj11          <- round            (nExp_poi11 *   (1 - homotypic.prop11)     )
normsample11            <- doubletFinder_v3 (normsample11,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi11,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample11            <- doubletFinder_v3 (normsample11,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj11,
                                             reuse.pANN = "pANN_0.25_0.09_179",
                                             sct        = FALSE)
```

```{r}
normsample11@meta.data
DimPlot (normsample11, group.by = "DF.classifications_0.25_0.09_179")
table   (normsample11$DF.classifications_0.25_0.09_179)
```

```{r}
# Subset to remove identified cell doublets
fltsample11 <- subset (normsample11, subset = DF.classifications_0.25_0.09_179 == 'Singlet')
fltsample11
table (fltsample11$DF.classifications_0.25_0.09_179)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample11)
rm (sweep.res.list_sample11)
rm (sweep.stats_sample11)
rm (bcmvn_sample11)
rm (annotations11)
rm (homotypic.prop11)
rm (nExp_poi11)
rm (nExp_poi.adj11)
rm (normsample11)
rm (ldat11)
```

### Patient D11

```{r}
# Patient D11

# Read in the data for placental villi sample ten
sample12 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Vento-Tormo/Vento-Tormo_P11/All")

# Read in the respective velocity data
          ldat12            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/FCA7474068.loom")
colnames (ldat12$spliced)   <- paste        (substring (colnames (ldat12$spliced),   12, 27), sep = "")
colnames (ldat12$spliced)   <- paste                   (colnames (ldat12$spliced),   "-1",    sep = "")
colnames (ldat12$unspliced) <- paste        (substring (colnames (ldat12$unspliced), 12, 27), sep = "")
colnames (ldat12$unspliced) <- paste                   (colnames (ldat12$unspliced), "-1",    sep = "")
colnames (ldat12$ambiguous) <- paste        (substring (colnames (ldat12$ambiguous), 12, 27), sep = "")
colnames (ldat12$ambiguous) <- paste                   (colnames (ldat12$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample12 <- CreateSeuratObject (counts = sample12, min.cells = 1, min.features = 1, project = "VT_P11.1")

for (i in names (x = ldat12)) { sample12[[i]] <- CreateAssayObject (counts = ldat12[[i]]) }

# Add the sample metadata
sample12 <- AddMetaData (sample12, metadata = "MPI",          col.name = "Model"    )
sample12 <- AddMetaData (sample12, metadata = "VT_P11",       col.name = "ID"       )
sample12 <- AddMetaData (sample12, metadata = "VT_P11.1",     col.name = "rawID"    )
sample12 <- AddMetaData (sample12, metadata = "0",            col.name = "nSex"     )
sample12 <- AddMetaData (sample12, metadata = "Female",       col.name = "Sex"      )
sample12 <- AddMetaData (sample12, metadata = "6.7",          col.name = "rawGA"    )
sample12 <- AddMetaData (sample12, metadata = "Early",        col.name = "binGA"    )
sample12 <- AddMetaData (sample12, metadata = "1",            col.name = "nGA"      )
sample12 <- AddMetaData (sample12, metadata = "Early",        col.name = "GA"       )
sample12 <- AddMetaData (sample12, metadata = "Placenta",     col.name = "Tissue"   )
sample12 <- AddMetaData (sample12, metadata = "3",            col.name = "Batch"    )
sample12 <- AddMetaData (sample12, metadata = "2",            col.name = "Source"   )
sample12 <- AddMetaData (sample12, metadata = "in vivo",      col.name = "Type"     )
sample12 <- AddMetaData (sample12, metadata = "Undetermined", col.name = "Pregnancy")
sample12 <- AddMetaData (sample12, metadata = "All",          col.name = "Gate"     )
sample12 <- AddMetaData (sample12, metadata = "First",        col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample12 <- NormalizeData    (sample12)
sample12 <- CellCycleScoring (sample12, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample12)
dim   (sample12)
sample12@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample12 [["percent.mt"]]   <- PercentageFeatureSet (sample12, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample12), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample12, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample12, slot   = 'counts')              );
sample12 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample12, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample12, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample12, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample12@meta.data
```

```{r}
# Filter sample ten
fltsample12 <- subset (sample12, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample12, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample12, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample12, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample12
fltsample12
```

```{r}
normsample12 <- NormalizeData        ( fltsample12)
normsample12 <- ScaleData            (normsample12)
normsample12 <- FindVariableFeatures (normsample12, selection.method = "vst", nfeatures = 2000)
normsample12 <- RunPCA               (normsample12)
normsample12 <- RunUMAP              (normsample12, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample12 <- paramSweep_v3    (normsample12, PCs = 1:50, sct = FALSE)
sweep.stats_sample12    <- summarizeSweep   (sweep.res.list_sample12,  GT  = FALSE)
bcmvn_sample12          <- find.pK          (sweep.stats_sample12)
annotations12           <- normsample12@meta.data$ClusteringResults
homotypic.prop12        <- modelHomotypic   (annotations12)
nExp_poi12              <- round            (0.039 * length (normsample12@meta.data$ID))  ## Assuming 3.9% doublet formation rate
nExp_poi.adj12          <- round            (nExp_poi12 *   (1 - homotypic.prop12)     )
normsample12            <- doubletFinder_v3 (normsample12,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi12,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample12            <- doubletFinder_v3 (normsample12,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj12,
                                             reuse.pANN = "pANN_0.25_0.09_130",
                                             sct        = FALSE)
```

```{r}
normsample12@meta.data
DimPlot (normsample12, group.by = "DF.classifications_0.25_0.09_130")
table   (normsample12$DF.classifications_0.25_0.09_130)
```

```{r}
# Subset to remove identified cell doublets
fltsample12 <- subset (normsample12, subset = DF.classifications_0.25_0.09_130 == 'Singlet')
fltsample12
table (fltsample12$DF.classifications_0.25_0.09_130)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample12)
rm (sweep.res.list_sample12)
rm (sweep.stats_sample12)
rm (bcmvn_sample12)
rm (annotations12)
rm (homotypic.prop12)
rm (nExp_poi12)
rm (nExp_poi.adj12)
rm (normsample12)
rm (ldat12)
```

```{r}
# Patient D11

# Read in the data for placental villi sample ten (file 3)
sample13 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Vento-Tormo/Vento-Tormo_P11/EPCAM+")

# Read in the respective velocity data
          ldat13            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/FCA7474069.loom")
colnames (ldat13$spliced)   <- paste        (substring (colnames (ldat13$spliced),   12, 27), sep = "")
colnames (ldat13$spliced)   <- paste                   (colnames (ldat13$spliced),   "-1",    sep = "")
colnames (ldat13$unspliced) <- paste        (substring (colnames (ldat13$unspliced), 12, 27), sep = "")
colnames (ldat13$unspliced) <- paste                   (colnames (ldat13$unspliced), "-1",    sep = "")
colnames (ldat13$ambiguous) <- paste        (substring (colnames (ldat13$ambiguous), 12, 27), sep = "")
colnames (ldat13$ambiguous) <- paste                   (colnames (ldat13$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample13 <- CreateSeuratObject (counts = sample13, min.cells = 1, min.features = 1, project = "VT_P11.2")

for (i in names (x = ldat13)) { sample13[[i]] <- CreateAssayObject (counts = ldat13[[i]]) }

# Add the sample metadata
sample13 <- AddMetaData (sample13, metadata = "MPI",          col.name = "Model"    )
sample13 <- AddMetaData (sample13, metadata = "VT_P11",       col.name = "ID"       )
sample13 <- AddMetaData (sample13, metadata = "VT_P11.2",     col.name = "rawID"    )
sample13 <- AddMetaData (sample13, metadata = "0",            col.name = "nSex"     )
sample13 <- AddMetaData (sample13, metadata = "Female",       col.name = "Sex"      )
sample13 <- AddMetaData (sample13, metadata = "6.7",          col.name = "rawGA"    )
sample13 <- AddMetaData (sample13, metadata = "Early",        col.name = "binGA"    )
sample13 <- AddMetaData (sample13, metadata = "1",            col.name = "nGA"      )
sample13 <- AddMetaData (sample13, metadata = "Early",        col.name = "GA"       )
sample13 <- AddMetaData (sample13, metadata = "Placenta",     col.name = "Tissue"   )
sample13 <- AddMetaData (sample13, metadata = "3",            col.name = "Batch"    )
sample13 <- AddMetaData (sample13, metadata = "2",            col.name = "Source"   )
sample13 <- AddMetaData (sample13, metadata = "in vivo",      col.name = "Type"     )
sample13 <- AddMetaData (sample13, metadata = "Undetermined", col.name = "Pregnancy")
sample13 <- AddMetaData (sample13, metadata = "EPCAM+",       col.name = "Gate"     )
sample13 <- AddMetaData (sample13, metadata = "First",        col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample13 <- NormalizeData    (sample13)
sample13 <- CellCycleScoring (sample13, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample13)
dim   (sample13)
sample13@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample13 [["percent.mt"]]   <- PercentageFeatureSet (sample13, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample13), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample13, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample13, slot   = 'counts')              );
sample13 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample13, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample13, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample13, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample13@meta.data
```

```{r}
# Filter sample nine (file 2)
fltsample13 <- subset (sample13, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample13, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample13, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample13, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample13
fltsample13
```

```{r}
normsample13 <- NormalizeData        ( fltsample13)
normsample13 <- ScaleData            (normsample13)
normsample13 <- FindVariableFeatures (normsample13, selection.method = "vst", nfeatures = 2000)
normsample13 <- RunPCA               (normsample13)
normsample13 <- RunUMAP              (normsample13, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample13 <- paramSweep_v3    (normsample13, PCs = 1:50, sct = FALSE)
sweep.stats_sample13    <- summarizeSweep   (sweep.res.list_sample13,  GT  = FALSE)
bcmvn_sample13          <- find.pK          (sweep.stats_sample13)
annotations13           <- normsample13@meta.data$ClusteringResults
homotypic.prop13        <- modelHomotypic   (annotations13)
nExp_poi13              <- round            (0.006 * length (normsample13@meta.data$ID))  ## Assuming 0.6% doublet formation rate
nExp_poi.adj13          <- round            (nExp_poi13 *   (1 - homotypic.prop13)     )
normsample13            <- doubletFinder_v3 (normsample13,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi13,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample13            <- doubletFinder_v3 (normsample13,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj13,
                                             reuse.pANN = "pANN_0.25_0.09_1",
                                             sct        = FALSE)
```

```{r}
normsample13@meta.data
DimPlot (normsample13, group.by = "DF.classifications_0.25_0.09_1")
table   (normsample13$DF.classifications_0.25_0.09_1)
```

```{r}
# Subset to remove identified cell doublets
fltsample13 <- subset (normsample13, subset = DF.classifications_0.25_0.09_1 == 'Singlet')
fltsample13
table (fltsample13$DF.classifications_0.25_0.09_1)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample13)
rm (sweep.res.list_sample13)
rm (sweep.stats_sample13)
rm (bcmvn_sample13)
rm (annotations13)
rm (homotypic.prop13)
rm (nExp_poi13)
rm (nExp_poi.adj13)
rm (normsample13)
rm (ldat13)
```

### Patient D12

```{r}
# Patient D12

# Read in the data for placental villi sample eleven
sample14 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Vento-Tormo/Vento-Tormo_P12/All")

# Read in the respective velocity data
          ldat14            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/FCA7511884.loom")
colnames (ldat14$spliced)   <- paste        (substring (colnames (ldat14$spliced),   12, 27), sep = "")
colnames (ldat14$spliced)   <- paste                   (colnames (ldat14$spliced),   "-1",    sep = "")
colnames (ldat14$unspliced) <- paste        (substring (colnames (ldat14$unspliced), 12, 27), sep = "")
colnames (ldat14$unspliced) <- paste                   (colnames (ldat14$unspliced), "-1",    sep = "")
colnames (ldat14$ambiguous) <- paste        (substring (colnames (ldat14$ambiguous), 12, 27), sep = "")
colnames (ldat14$ambiguous) <- paste                   (colnames (ldat14$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample14 <- CreateSeuratObject (counts = sample14, min.cells = 1, min.features = 1, project = "VT_P12.1")

for (i in names (x = ldat14)) { sample14[[i]] <- CreateAssayObject(counts = ldat14[[i]]) }

# Add the sample metadata
sample14 <- AddMetaData (sample14, metadata = "MPI",      col.name = "Model"    )
sample14 <- AddMetaData (sample14, metadata = "VT_P12",   col.name = "ID"       )
sample14 <- AddMetaData (sample14, metadata = "VT_P12.1", col.name = "rawID"    )
sample14 <- AddMetaData (sample14, metadata = "0",        col.name = "nSex"     )
sample14 <- AddMetaData (sample14, metadata = "Female",   col.name = "Sex"      )
sample14 <- AddMetaData (sample14, metadata = "10.0",     col.name = "rawGA"    )
sample14 <- AddMetaData (sample14, metadata = "Mid",      col.name = "binGA"    )
sample14 <- AddMetaData (sample14, metadata = "2",        col.name = "nGA"      )
sample14 <- AddMetaData (sample14, metadata = "Late",     col.name = "GA"       )
sample14 <- AddMetaData (sample14, metadata = "Placenta", col.name = "Tissue"   )
sample14 <- AddMetaData (sample14, metadata = "3",        col.name = "Batch"    )
sample14 <- AddMetaData (sample14, metadata = "2",        col.name = "Source"   )
sample14 <- AddMetaData (sample14, metadata = "in vivo",  col.name = "Type"     )
sample14 <- AddMetaData (sample14, metadata = "Fourth",   col.name = "Pregnancy")
sample14 <- AddMetaData (sample14, metadata = "All",      col.name = "Gate"     )
sample14 <- AddMetaData (sample14, metadata = "First",    col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample14 <- NormalizeData    (sample14)
sample14 <- CellCycleScoring (sample14, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample14)
dim   (sample14)
sample14@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample14 [["percent.mt"]]   <- PercentageFeatureSet (sample14, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample14), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample14, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample14, slot   = 'counts')              );
sample14 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample14, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample14, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample14, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample14@meta.data
```

```{r}
# Filter sample eleven
fltsample14 <- subset (sample14, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample14, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample14, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample14, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample14
fltsample14
```

```{r}
normsample14 <- NormalizeData        ( fltsample14)
normsample14 <- ScaleData            (normsample14)
normsample14 <- FindVariableFeatures (normsample14, selection.method = "vst", nfeatures = 2000)
normsample14 <- RunPCA               (normsample14)
normsample14 <- RunUMAP              (normsample14, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample14 <- paramSweep_v3    (normsample14, PCs = 1:50, sct = FALSE)
sweep.stats_sample14    <- summarizeSweep   (sweep.res.list_sample14,  GT  = FALSE)
bcmvn_sample14          <- find.pK          (sweep.stats_sample14)
annotations14           <- normsample14@meta.data$ClusteringResults
homotypic.prop14        <- modelHomotypic   (annotations14)
nExp_poi14              <- round            (0.035 * length (normsample14@meta.data$ID))  ## Assuming 3.5% doublet formation rate
nExp_poi.adj14          <- round            (nExp_poi14 *   (1 - homotypic.prop14)     )
normsample14            <- doubletFinder_v3 (normsample14,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi14,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample14            <- doubletFinder_v3 (normsample14,
                                             PCs        = 1:50,
                                             pN         = 0.25, 
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj14,
                                             reuse.pANN = "pANN_0.25_0.09_140",
                                             sct        = FALSE)
```

```{r}
normsample14@meta.data
DimPlot (normsample14, group.by = "DF.classifications_0.25_0.09_140")
table   (normsample14$DF.classifications_0.25_0.09_140)
```

```{r}
# Subset to remove identified cell doublets
fltsample14 <- subset (normsample14, subset = DF.classifications_0.25_0.09_140 == 'Singlet')
fltsample14
table (fltsample14$DF.classifications_0.25_0.09_140)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample14)
rm (sweep.res.list_sample14)
rm (sweep.stats_sample14)
rm (bcmvn_sample14)
rm (annotations14)
rm (homotypic.prop14)
rm (nExp_poi14)
rm (nExp_poi.adj14)
rm (normsample14)
rm (ldat14)
```

```{r}
# Patient D12

# Read in the data for placental villi sample ten (file 3)
sample15 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Vento-Tormo/Vento-Tormo_P12/EPCAM+")

# Read in the respective velocity data
          ldat15            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/FCA7511886.loom")
colnames (ldat15$spliced)   <- paste        (substring (colnames (ldat15$spliced),   12, 27), sep = "")
colnames (ldat15$spliced)   <- paste                   (colnames (ldat15$spliced),   "-1",    sep = "")
colnames (ldat15$unspliced) <- paste        (substring (colnames (ldat15$unspliced), 12, 27), sep = "")
colnames (ldat15$unspliced) <- paste                   (colnames (ldat15$unspliced), "-1",    sep = "")
colnames (ldat15$ambiguous) <- paste        (substring (colnames (ldat15$ambiguous), 12, 27), sep = "")
colnames (ldat15$ambiguous) <- paste                   (colnames (ldat15$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample15 <- CreateSeuratObject (counts = sample15, min.cells = 1, min.features = 1, project = "VT_P12.2")

for (i in names (x = ldat15)) { sample15[[i]] <- CreateAssayObject (counts = ldat15[[i]]) }

# Add the sample metadata
sample15 <- AddMetaData (sample15, metadata = "MPI",      col.name = "Model"    )
sample15 <- AddMetaData (sample15, metadata = "VT_P12",   col.name = "ID"       )
sample15 <- AddMetaData (sample15, metadata = "VT_P12.2", col.name = "rawID"    )
sample15 <- AddMetaData (sample15, metadata = "0",        col.name = "nSex"     )
sample15 <- AddMetaData (sample15, metadata = "Female",   col.name = "Sex"      )
sample15 <- AddMetaData (sample15, metadata = "10.0",     col.name = "rawGA"    )
sample15 <- AddMetaData (sample15, metadata = "Mid",      col.name = "binGA"    )
sample15 <- AddMetaData (sample15, metadata = "2",        col.name = "nGA"      )
sample15 <- AddMetaData (sample15, metadata = "Late",     col.name = "GA"       )
sample15 <- AddMetaData (sample15, metadata = "Placenta", col.name = "Tissue"   )
sample15 <- AddMetaData (sample15, metadata = "3",        col.name = "Batch"    )
sample15 <- AddMetaData (sample15, metadata = "2",        col.name = "Source"   )
sample15 <- AddMetaData (sample15, metadata = "in vivo",  col.name = "Type"     )
sample15 <- AddMetaData (sample15, metadata = "Fourth",   col.name = "Pregnancy")
sample15 <- AddMetaData (sample15, metadata = "EPCAM+",   col.name = "Gate"     )
sample15 <- AddMetaData (sample15, metadata = "First",    col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample15 <- NormalizeData    (sample15)
sample15 <- CellCycleScoring (sample15, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample15)
dim   (sample15)
sample15@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample15 [["percent.mt"]]   <- PercentageFeatureSet (sample15, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample15), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample15, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample15, slot   = 'counts')              );
sample15 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample15, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample15, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample15, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample15@meta.data
```

```{r}
# Filter sample nine (file 2)
fltsample15 <- subset (sample15, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample15, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample15, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample15, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample15
fltsample15
```

```{r}
normsample15 <- NormalizeData        ( fltsample15)
normsample15 <- ScaleData            (normsample15)
normsample15 <- FindVariableFeatures (normsample15, selection.method = "vst", nfeatures = 2000)
normsample15 <- RunPCA               (normsample15)
normsample15 <- RunUMAP              (normsample15, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample15 <- paramSweep_v3    (normsample15, PCs = 1:50, sct = FALSE)
sweep.stats_sample15    <- summarizeSweep   (sweep.res.list_sample15,  GT  = FALSE)
bcmvn_sample15          <- find.pK          (sweep.stats_sample15)
annotations15           <- normsample15@meta.data$ClusteringResults
homotypic.prop15        <- modelHomotypic   (annotations15)
nExp_poi15              <- round            (0.025 * length (normsample15@meta.data$ID))  ## Assuming 2.5% doublet formation rate
nExp_poi.adj15          <- round            (nExp_poi15 *   (1 - homotypic.prop15)     )
normsample15            <- doubletFinder_v3 (normsample15,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi15,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample15            <- doubletFinder_v3 (normsample15,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj15,
                                             reuse.pANN = "pANN_0.25_0.09_67",
                                             sct        = FALSE)
```

```{r}
normsample15@meta.data
DimPlot (normsample15, group.by = "DF.classifications_0.25_0.09_67")
table   (normsample15$DF.classifications_0.25_0.09_67)
```

```{r}
# Subset to remove identified cell doublets
fltsample15 <- subset (normsample15, subset = DF.classifications_0.25_0.09_67 == 'Singlet')
fltsample15
table (fltsample15$DF.classifications_0.25_0.09_67)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample15)
rm (sweep.res.list_sample15)
rm (sweep.stats_sample15)
rm (bcmvn_sample15)
rm (annotations15)
rm (homotypic.prop15)
rm (nExp_poi15)
rm (nExp_poi.adj15)
rm (normsample15)
rm (ldat15)
```

```{r}
# Patient D12

# Read in the data for placental villi sample ten (file 3)
sample16 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Vento-Tormo/Vento-Tormo_P12/HLA-G+")

# Read in the respective velocity data
          ldat16            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/FCA7511885.loom")
colnames (ldat16$spliced)   <- paste        (substring (colnames (ldat16$spliced),   12, 27), sep = "")
colnames (ldat16$spliced)   <- paste                   (colnames (ldat16$spliced),   "-1",    sep = "")
colnames (ldat16$unspliced) <- paste        (substring (colnames (ldat16$unspliced), 12, 27), sep = "")
colnames (ldat16$unspliced) <- paste                   (colnames (ldat16$unspliced), "-1",    sep = "")
colnames (ldat16$ambiguous) <- paste        (substring (colnames (ldat16$ambiguous), 12, 27), sep = "")
colnames (ldat16$ambiguous) <- paste                   (colnames (ldat16$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample16 <- CreateSeuratObject (counts = sample16, min.cells = 1, min.features = 1, project = "VT_P12.3")

for (i in names (x = ldat16)) { sample16[[i]] <- CreateAssayObject (counts = ldat16[[i]]) }

# Add the sample metadata
sample16 <- AddMetaData (sample16, metadata = "MPI",      col.name = "Model"    )
sample16 <- AddMetaData (sample16, metadata = "VT_P12",   col.name = "ID"       )
sample16 <- AddMetaData (sample16, metadata = "VT_P12.3", col.name = "rawID"    )
sample16 <- AddMetaData (sample16, metadata = "0",        col.name = "nSex"     )
sample16 <- AddMetaData (sample16, metadata = "Female",   col.name = "Sex"      )
sample16 <- AddMetaData (sample16, metadata = "10.0",     col.name = "rawGA"    )
sample16 <- AddMetaData (sample16, metadata = "Mid",      col.name = "binGA"    )
sample16 <- AddMetaData (sample16, metadata = "2",        col.name = "nGA"      )
sample16 <- AddMetaData (sample16, metadata = "Late",     col.name = "GA"       )
sample16 <- AddMetaData (sample16, metadata = "Placenta", col.name = "Tissue"   )
sample16 <- AddMetaData (sample16, metadata = "3",        col.name = "Batch"    )
sample16 <- AddMetaData (sample16, metadata = "2",        col.name = "Source"   )
sample16 <- AddMetaData (sample16, metadata = "in vivo",  col.name = "Type"     )
sample16 <- AddMetaData (sample16, metadata = "Fourth",   col.name = "Pregnancy")
sample16 <- AddMetaData (sample16, metadata = "HLA-G+",   col.name = "Gate"     )
sample16 <- AddMetaData (sample16, metadata = "First",    col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample16 <- NormalizeData    (sample16)
sample16 <- CellCycleScoring (sample16, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample16)
dim   (sample16)
sample16@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample16 [["percent.mt"]]   <- PercentageFeatureSet (sample16, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample16), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample16, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample16, slot   = 'counts')              );
sample16 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample16, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample16, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample16, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample16@meta.data
```

```{r}
# Filter sample nine (file 2)
fltsample16 <- subset (sample16, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample16, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample16, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample16, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample16
fltsample16
```

```{r}
normsample16 <- NormalizeData        ( fltsample16)
normsample16 <- ScaleData            (normsample16)
normsample16 <- FindVariableFeatures (normsample16, selection.method = "vst", nfeatures = 2000)
normsample16 <- RunPCA               (normsample16)
normsample16 <- RunUMAP              (normsample16, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample16 <- paramSweep_v3    (normsample16, PCs = 1:50, sct = FALSE)
sweep.stats_sample16    <- summarizeSweep   (sweep.res.list_sample16,  GT  = FALSE)
bcmvn_sample16          <- find.pK          (sweep.stats_sample16)
annotations16           <- normsample16@meta.data$ClusteringResults
homotypic.prop16        <- modelHomotypic   (annotations16)
nExp_poi16              <- round            (0.030 * length (normsample16@meta.data$ID))  ## Assuming 3.0% doublet formation rate
nExp_poi.adj16          <- round            (nExp_poi16 *   (1 - homotypic.prop16)     )
normsample16            <- doubletFinder_v3 (normsample16,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi16,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample16            <- doubletFinder_v3 (normsample16,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj16,
                                             reuse.pANN = "pANN_0.25_0.09_105",
                                             sct        = FALSE)
```

```{r}
normsample16@meta.data
DimPlot (normsample16, group.by = "DF.classifications_0.25_0.09_105")
table   (normsample16$DF.classifications_0.25_0.09_105)
```

```{r}
# Subset to remove identified cell doublets
fltsample16 <- subset (normsample16, subset = DF.classifications_0.25_0.09_105 == 'Singlet')
fltsample16
table (fltsample16$DF.classifications_0.25_0.09_105)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample16)
rm (sweep.res.list_sample16)
rm (sweep.stats_sample16)
rm (bcmvn_sample16)
rm (annotations16)
rm (homotypic.prop16)
rm (nExp_poi16)
rm (nExp_poi.adj16)
rm (normsample16)
rm (ldat16)
```

## Read in the Vento-Tormo single-cell Maternal data

### Patient D6 Decidua

```{r}
# Patient D6

# Read in the data for decidual sample file one
sample17 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Vento-Tormo/Vento-Tormo_D6/CD45-")

# Read in the respective velocity data
          ldat17            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/FCA7167223.loom")
colnames (ldat17$spliced)   <- paste        (substring (colnames (ldat17$spliced),   12, 27), sep = "")
colnames (ldat17$spliced)   <- paste                   (colnames (ldat17$spliced),   "-1",    sep = "")
colnames (ldat17$unspliced) <- paste        (substring (colnames (ldat17$unspliced), 12, 27), sep = "")
colnames (ldat17$unspliced) <- paste                   (colnames (ldat17$unspliced), "-1",    sep = "")
colnames (ldat17$ambiguous) <- paste        (substring (colnames (ldat17$ambiguous), 12, 27), sep = "")
colnames (ldat17$ambiguous) <- paste                   (colnames (ldat17$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample17 <- CreateSeuratObject (counts = sample17, min.cells = 1, min.features = 1, project = "VT_D6.1")

for (i in names (x = ldat17)) { sample17[[i]] <- CreateAssayObject(counts = ldat17[[i]]) }

# Add the sample metadata
sample17 <- AddMetaData (sample17, metadata = "MPI",     col.name = "Model"    )
sample17 <- AddMetaData (sample17, metadata = "VT_D6",   col.name = "ID"       )
sample17 <- AddMetaData (sample17, metadata = "VT_D6.1", col.name = "rawID"    )
sample17 <- AddMetaData (sample17, metadata = "0",       col.name = "nSex"     )
sample17 <- AddMetaData (sample17, metadata = "Female",  col.name = "Sex"      )
sample17 <- AddMetaData (sample17, metadata = "12.0",    col.name = "rawGA"    )
sample17 <- AddMetaData (sample17, metadata = "Late",    col.name = "binGA"    )
sample17 <- AddMetaData (sample17, metadata = "2",       col.name = "nGA"      )
sample17 <- AddMetaData (sample17, metadata = "Late",    col.name = "GA"       )
sample17 <- AddMetaData (sample17, metadata = "Decidua", col.name = "Tissue"   )
sample17 <- AddMetaData (sample17, metadata = "3",       col.name = "Batch"    )
sample17 <- AddMetaData (sample17, metadata = "2",       col.name = "Source"   )
sample17 <- AddMetaData (sample17, metadata = "in vivo", col.name = "Type"     )
sample17 <- AddMetaData (sample17, metadata = "Second",  col.name = "Pregnancy")
sample17 <- AddMetaData (sample17, metadata = "CD45-",   col.name = "Gate"     )
sample17 <- AddMetaData (sample17, metadata = "First",   col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample17 <- NormalizeData    (sample17)
sample17 <- CellCycleScoring (sample17, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample17)
dim   (sample17)
sample17@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample17 [["percent.mt"]]   <- PercentageFeatureSet (sample17, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample17), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample17, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample17, slot   = 'counts')              );
sample17 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample17, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample17, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample17, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample17@meta.data
```

```{r}
# Filter decidual file one
fltsample17 <- subset (sample17, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample17, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample17, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample17, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample17
fltsample17
```

```{r}
normsample17 <- NormalizeData        ( fltsample17)
normsample17 <- ScaleData            (normsample17)
normsample17 <- FindVariableFeatures (normsample17, selection.method = "vst", nfeatures = 2000)
normsample17 <- RunPCA               (normsample17)
normsample17 <- RunUMAP              (normsample17, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample17 <- paramSweep_v3    (normsample17, PCs = 1:50, sct = FALSE)
sweep.stats_sample17    <- summarizeSweep   (sweep.res.list_sample17,  GT  = FALSE)
bcmvn_sample17          <- find.pK          (sweep.stats_sample17)
annotations17           <- normsample17@meta.data$ClusteringResults
homotypic.prop17        <- modelHomotypic   (annotations17)
nExp_poi17              <- round            (0.055 * length (normsample17@meta.data$ID))  ## Assuming 5.5% doublet formation rate
nExp_poi.adj17          <- round            (nExp_poi17 *   (1 - homotypic.prop17)     )
normsample17            <- doubletFinder_v3 (normsample17,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi17,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample17            <- doubletFinder_v3 (normsample17,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj17,
                                             reuse.pANN = "pANN_0.25_0.09_302",
                                             sct        = FALSE)
```

```{r}
normsample17@meta.data
DimPlot (normsample17, group.by = "DF.classifications_0.25_0.09_302")
table   (normsample17$DF.classifications_0.25_0.09_302)
```

```{r}
# Subset to remove identified cell doublets
fltsample17 <- subset (normsample17, subset = DF.classifications_0.25_0.09_302 == 'Singlet')
fltsample17
table (fltsample17$DF.classifications_0.25_0.09_302)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample17)
rm (sweep.res.list_sample17)
rm (sweep.stats_sample17)
rm (bcmvn_sample17)
rm (annotations17)
rm (homotypic.prop17)
rm (nExp_poi17)
rm (nExp_poi.adj17)
rm (normsample17)
rm (ldat17)
```

```{r}
# Patient D6

# Read in the data for decidual sample file two
sample18 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Vento-Tormo/Vento-Tormo_D6/CD45+.1")

# Read in the respective velocity data
          ldat18            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/FCA7167219.loom")
colnames (ldat18$spliced)   <- paste        (substring (colnames (ldat18$spliced),   12, 27), sep = "")
colnames (ldat18$spliced)   <- paste                   (colnames (ldat18$spliced),   "-1",    sep = "")
colnames (ldat18$unspliced) <- paste        (substring (colnames (ldat18$unspliced), 12, 27), sep = "")
colnames (ldat18$unspliced) <- paste                   (colnames (ldat18$unspliced), "-1",    sep = "")
colnames (ldat18$ambiguous) <- paste        (substring (colnames (ldat18$ambiguous), 12, 27), sep = "")
colnames (ldat18$ambiguous) <- paste                   (colnames (ldat18$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample18 <- CreateSeuratObject (counts = sample18, min.cells = 1, min.features = 1, project = "VT_D6.2")

for (i in names (x = ldat18)) { sample18[[i]] <- CreateAssayObject(counts = ldat18[[i]]) }

# Add the sample metadata
sample18 <- AddMetaData (sample18, metadata = "MPI",     col.name = "Model"    )
sample18 <- AddMetaData (sample18, metadata = "VT_D6",   col.name = "ID"       )
sample18 <- AddMetaData (sample18, metadata = "VT_D6.2", col.name = "rawID"    )
sample18 <- AddMetaData (sample18, metadata = "0",       col.name = "nSex"     )
sample18 <- AddMetaData (sample18, metadata = "Female",  col.name = "Sex"      )
sample18 <- AddMetaData (sample18, metadata = "12.0",    col.name = "rawGA"    )
sample18 <- AddMetaData (sample18, metadata = "Late",    col.name = "binGA"    )
sample18 <- AddMetaData (sample18, metadata = "2",       col.name = "nGA"      )
sample18 <- AddMetaData (sample18, metadata = "Late",    col.name = "GA"       )
sample18 <- AddMetaData (sample18, metadata = "Decidua", col.name = "Tissue"   )
sample18 <- AddMetaData (sample18, metadata = "3",       col.name = "Batch"    )
sample18 <- AddMetaData (sample18, metadata = "2",       col.name = "Source"   )
sample18 <- AddMetaData (sample18, metadata = "in vivo", col.name = "Type"     )
sample18 <- AddMetaData (sample18, metadata = "Second",  col.name = "Pregnancy")
sample18 <- AddMetaData (sample18, metadata = "CD45+",   col.name = "Gate"     )
sample18 <- AddMetaData (sample18, metadata = "First",   col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample18 <- NormalizeData    (sample18)
sample18 <- CellCycleScoring (sample18, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample18)
dim   (sample18)
sample18@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample18 [["percent.mt"]]   <- PercentageFeatureSet (sample18, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample18), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample18, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample18, slot   = 'counts')              );
sample18 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample18, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample18, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample18, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample18@meta.data
```

```{r}
# Filter decidual file two
fltsample18 <- subset (sample18, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample18, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample18, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample18, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample18
fltsample18
```

```{r}
normsample18 <- NormalizeData        ( fltsample18)
normsample18 <- ScaleData            (normsample18)
normsample18 <- FindVariableFeatures (normsample18, selection.method = "vst", nfeatures = 2000)
normsample18 <- RunPCA               (normsample18)
normsample18 <- RunUMAP              (normsample18, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample18 <- paramSweep_v3    (normsample18, PCs = 1:50, sct = FALSE)
sweep.stats_sample18    <- summarizeSweep   (sweep.res.list_sample18,  GT  = FALSE)
bcmvn_sample18          <- find.pK          (sweep.stats_sample18)
annotations18           <- normsample18@meta.data$ClusteringResults
homotypic.prop18        <- modelHomotypic   (annotations18)
nExp_poi18              <- round            (0.004 * length (normsample18@meta.data$ID))  ## Assuming 0.4% doublet formation rate
nExp_poi.adj18          <- round            (nExp_poi18 *   (1 - homotypic.prop18)     )
normsample18            <- doubletFinder_v3 (normsample18,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi18,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample18            <- doubletFinder_v3 (normsample18,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj18,
                                             reuse.pANN = "pANN_0.25_0.09_2",
                                             sct        = FALSE)
```

```{r}
normsample18@meta.data
DimPlot (normsample18, group.by = "DF.classifications_0.25_0.09_2")
table   (normsample18$DF.classifications_0.25_0.09_2)
```

```{r}
# Subset to remove identified cell doublets
fltsample18 <- subset (normsample18, subset = DF.classifications_0.25_0.09_2 == 'Singlet')
fltsample18
table (fltsample18$DF.classifications_0.25_0.09_2)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample18)
rm (sweep.res.list_sample18)
rm (sweep.stats_sample18)
rm (bcmvn_sample18)
rm (annotations18)
rm (homotypic.prop18)
rm (nExp_poi18)
rm (nExp_poi.adj18)
rm (normsample18)
rm (ldat18)
```

```{r}
# Patient D6

# Read in the data for decidual sample file three
sample19 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Vento-Tormo/Vento-Tormo_D6/CD45+.2")

# Read in the respective velocity data
          ldat19            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/FCA7167224.loom")
colnames (ldat19$spliced)   <- paste        (substring (colnames (ldat19$spliced),   12, 27), sep = "")
colnames (ldat19$spliced)   <- paste                   (colnames (ldat19$spliced),   "-1",    sep = "")
colnames (ldat19$unspliced) <- paste        (substring (colnames (ldat19$unspliced), 12, 27), sep = "")
colnames (ldat19$unspliced) <- paste                   (colnames (ldat19$unspliced), "-1",    sep = "")
colnames (ldat19$ambiguous) <- paste        (substring (colnames (ldat19$ambiguous), 12, 27), sep = "")
colnames (ldat19$ambiguous) <- paste                   (colnames (ldat19$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample19 <- CreateSeuratObject (counts = sample19, min.cells = 1, min.features = 1, project = "VT_D6.3")

for (i in names (x = ldat19)) { sample19[[i]] <- CreateAssayObject(counts = ldat19[[i]]) }

# Add the sample metadata
sample19 <- AddMetaData (sample19, metadata = "MPI",     col.name = "Model"    )
sample19 <- AddMetaData (sample19, metadata = "VT_D6",   col.name = "ID"       )
sample19 <- AddMetaData (sample19, metadata = "VT_D6.3", col.name = "rawID"    )
sample19 <- AddMetaData (sample19, metadata = "0",       col.name = "nSex"     )
sample19 <- AddMetaData (sample19, metadata = "Female",  col.name = "Sex"      )
sample19 <- AddMetaData (sample19, metadata = "12.0",    col.name = "rawGA"    )
sample19 <- AddMetaData (sample19, metadata = "Late",    col.name = "binGA"    )
sample19 <- AddMetaData (sample19, metadata = "2",       col.name = "nGA"      )
sample19 <- AddMetaData (sample19, metadata = "Late",    col.name = "GA"       )
sample19 <- AddMetaData (sample19, metadata = "Decidua", col.name = "Tissue"   )
sample19 <- AddMetaData (sample19, metadata = "3",       col.name = "Batch"    )
sample19 <- AddMetaData (sample19, metadata = "2",       col.name = "Source"   )
sample19 <- AddMetaData (sample19, metadata = "in vivo", col.name = "Type"     )
sample19 <- AddMetaData (sample19, metadata = "Second",  col.name = "Pregnancy")
sample19 <- AddMetaData (sample19, metadata = "CD45+",   col.name = "Gate"     )
sample19 <- AddMetaData (sample19, metadata = "First",   col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample19 <- NormalizeData    (sample19)
sample19 <- CellCycleScoring (sample19, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample19)
dim   (sample19)
sample19@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample19 [["percent.mt"]]   <- PercentageFeatureSet (sample19, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample19), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample19, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample19, slot   = 'counts')              );
sample19 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample19, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample19, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample19, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample19@meta.data
```

```{r}
# Filter decidual file three
fltsample19 <- subset (sample19, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample19, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample19, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample19, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample19
fltsample19
```

```{r}
normsample19 <- NormalizeData        ( fltsample19)
normsample19 <- ScaleData            (normsample19)
normsample19 <- FindVariableFeatures (normsample19, selection.method = "vst", nfeatures = 2000)
normsample19 <- RunPCA               (normsample19)
normsample19 <- RunUMAP              (normsample19, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample19 <- paramSweep_v3    (normsample19, PCs = 1:50, sct = FALSE)
sweep.stats_sample19    <- summarizeSweep   (sweep.res.list_sample19,  GT  = FALSE)
bcmvn_sample19          <- find.pK          (sweep.stats_sample19)
annotations19           <- normsample19@meta.data$ClusteringResults
homotypic.prop19        <- modelHomotypic   (annotations19)
nExp_poi19              <- round            (0.019 * length (normsample19@meta.data$ID))  ## Assuming 1.5% doublet formation rate
nExp_poi.adj19          <- round            (nExp_poi19 *   (1-homotypic.prop19)       )
normsample19            <- doubletFinder_v3 (normsample19,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi19,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample19            <- doubletFinder_v3 (normsample19,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj19,
                                             reuse.pANN = "pANN_0.25_0.09_41",
                                             sct        = FALSE)
```

```{r}
normsample19@meta.data
DimPlot (normsample19, group.by = "DF.classifications_0.25_0.09_41")
table   (normsample19$DF.classifications_0.25_0.09_41)
```

```{r}
# Subset to remove identified cell doublets
fltsample19 <- subset (normsample19, subset = DF.classifications_0.25_0.09_41 == 'Singlet')
fltsample19
table (fltsample19$DF.classifications_0.25_0.09_41)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample19)
rm (sweep.res.list_sample19)
rm (sweep.stats_sample19)
rm (bcmvn_sample19)
rm (annotations19)
rm (homotypic.prop19)
rm (nExp_poi19)
rm (nExp_poi.adj19)
rm (normsample19)
rm (ldat19)
```

### Patient D6 Maternal peripheral blood

```{r}
# Patient D6 - Blood

# Read in the data for placental villi sample ten (file 3)
sample20 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Vento-Tormo/Vento-Tormo_B6/PB.1")

# Read in the respective velocity data
          ldat20            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/FCA7167231.loom")
colnames (ldat20$spliced)   <- paste        (substring (colnames (ldat20$spliced),   12, 27), sep = "")
colnames (ldat20$spliced)   <- paste                   (colnames (ldat20$spliced),   "-1",    sep = "")
colnames (ldat20$unspliced) <- paste        (substring (colnames (ldat20$unspliced), 12, 27), sep = "")
colnames (ldat20$unspliced) <- paste                   (colnames (ldat20$unspliced), "-1",    sep = "")
colnames (ldat20$ambiguous) <- paste        (substring (colnames (ldat20$ambiguous), 12, 27), sep = "")
colnames (ldat20$ambiguous) <- paste                   (colnames (ldat20$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample20 <- CreateSeuratObject (counts = sample20, min.cells = 1, min.features = 1, project = "VT_B6.1")

# Add the sample metadata
sample20 <- AddMetaData (sample20, metadata = "MPI",     col.name = "Model"    )
sample20 <- AddMetaData (sample20, metadata = "VT_D6",   col.name = "ID"       )
sample20 <- AddMetaData (sample20, metadata = "VT_B6.1", col.name = "rawID"    )
sample20 <- AddMetaData (sample20, metadata = "0",       col.name = "nSex"     )
sample20 <- AddMetaData (sample20, metadata = "Female",  col.name = "Sex"      )
sample20 <- AddMetaData (sample20, metadata = "12.0",    col.name = "rawGA"    )
sample20 <- AddMetaData (sample20, metadata = "Late",    col.name = "binGA"    )
sample20 <- AddMetaData (sample20, metadata = "2",       col.name = "nGA"      )
sample20 <- AddMetaData (sample20, metadata = "Late",    col.name = "GA"       )
sample20 <- AddMetaData (sample20, metadata = "PB",      col.name = "Tissue"   )
sample20 <- AddMetaData (sample20, metadata = "3",       col.name = "Batch"    )
sample20 <- AddMetaData (sample20, metadata = "2",       col.name = "Source"   )
sample20 <- AddMetaData (sample20, metadata = "in vivo", col.name = "Type"     )
sample20 <- AddMetaData (sample20, metadata = "Second",  col.name = "Pregnancy")
sample20 <- AddMetaData (sample20, metadata = "PB",      col.name = "Gate"     )
sample20 <- AddMetaData (sample20, metadata = "First",   col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample20 <- NormalizeData    (sample20)
sample20 <- CellCycleScoring (sample20, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample20)
dim   (sample20)
sample20@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample20 [["percent.mt"]]   <- PercentageFeatureSet (sample20, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample20), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample20, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample20, slot   = 'counts')              );
sample20 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample20, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample20, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample20, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample20@meta.data
```

```{r}
# Filter sample nine (file 2)
fltsample20 <- subset (sample20, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample20, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample20, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample20, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample20
fltsample20
```

```{r}
normsample20 <- NormalizeData        ( fltsample20)
normsample20 <- ScaleData            (normsample20)
normsample20 <- FindVariableFeatures (normsample20, selection.method = "vst", nfeatures = 2000)
normsample20 <- RunPCA               (normsample20)
normsample20 <- RunUMAP              (normsample20, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample20 <- paramSweep_v3    (normsample20, PCs = 1:50, sct = FALSE)
sweep.stats_sample20    <- summarizeSweep   (sweep.res.list_sample20,  GT  = FALSE)
bcmvn_sample20          <- find.pK          (sweep.stats_sample20)
annotations20           <- normsample20@meta.data$ClusteringResults
homotypic.prop20        <- modelHomotypic   (annotations20)
nExp_poi20              <- round            (0.012 * length (normsample20@meta.data$ID))  ## Assuming 1.2% doublet formation rate
nExp_poi.adj20          <- round            (nExp_poi20 *   (1 - homotypic.prop20)     )
normsample20            <- doubletFinder_v3 (normsample20,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi20,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample20            <- doubletFinder_v3 (normsample20,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj20,
                                             reuse.pANN = "pANN_0.25_0.09_13",
                                             sct        = FALSE)
```

```{r}
normsample20@meta.data
DimPlot (normsample20, group.by = "DF.classifications_0.25_0.09_13")
table   (normsample20$DF.classifications_0.25_0.09_13)
```

```{r}
# Subset to remove identified cell doublets
fltsample20 <- subset (normsample20, subset = DF.classifications_0.25_0.09_13 == 'Singlet')
fltsample20
table (fltsample20$DF.classifications_0.25_0.09_13)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample20)
rm (sweep.res.list_sample20)
rm (sweep.stats_sample20)
rm (bcmvn_sample20)
rm (annotations20)
rm (homotypic.prop20)
rm (nExp_poi20)
rm (nExp_poi.adj20)
rm (normsample20)
rm (ldat20)
```

```{r}
# Patient D6

# Read in the data for placental villi sample ten (file 3)
sample21 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Vento-Tormo/Vento-Tormo_B6/PB.2")

# Read in the respective velocity data
          ldat21            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/FCA7167232.loom")
colnames (ldat21$spliced)   <- paste        (substring (colnames (ldat21$spliced),   12, 27), sep = "")
colnames (ldat21$spliced)   <- paste                   (colnames (ldat21$spliced),   "-1",    sep = "")
colnames (ldat21$unspliced) <- paste        (substring (colnames (ldat21$unspliced), 12, 27), sep = "")
colnames (ldat21$unspliced) <- paste                   (colnames (ldat21$unspliced), "-1",    sep = "")
colnames (ldat21$ambiguous) <- paste        (substring (colnames (ldat21$ambiguous), 12, 27), sep = "")
colnames (ldat21$ambiguous) <- paste                   (colnames (ldat21$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample21 <- CreateSeuratObject (counts = sample21, min.cells = 1, min.features = 1, project = "VT_B6.2")

for (i in names (x = ldat21)) { sample21[[i]] <- CreateAssayObject (counts = ldat21[[i]]) }

# Add the sample metadata
sample21 <- AddMetaData (sample21, metadata = "MPI",     col.name = "Model"    )
sample21 <- AddMetaData (sample21, metadata = "VT_D6",   col.name = "ID"       )
sample21 <- AddMetaData (sample21, metadata = "VT_B6.2", col.name = "rawID"    )
sample21 <- AddMetaData (sample21, metadata = "0",       col.name = "nSex"     )
sample21 <- AddMetaData (sample21, metadata = "Female",  col.name = "Sex"      )
sample21 <- AddMetaData (sample21, metadata = "12.0",    col.name = "rawGA"    )
sample21 <- AddMetaData (sample21, metadata = "Late",    col.name = "binGA"    )
sample21 <- AddMetaData (sample21, metadata = "2",       col.name = "nGA"      )
sample21 <- AddMetaData (sample21, metadata = "Late",    col.name = "GA"       )
sample21 <- AddMetaData (sample21, metadata = "PB",      col.name = "Tissue"   )
sample21 <- AddMetaData (sample21, metadata = "3",       col.name = "Batch"    )
sample21 <- AddMetaData (sample21, metadata = "2",       col.name = "Source"   )
sample21 <- AddMetaData (sample21, metadata = "in vivo", col.name = "Type"     )
sample21 <- AddMetaData (sample21, metadata = "Second",  col.name = "Pregnancy")
sample21 <- AddMetaData (sample21, metadata = "PB",      col.name = "Gate"     )
sample21 <- AddMetaData (sample21, metadata = "First",   col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample21 <- NormalizeData    (sample21)
sample21 <- CellCycleScoring (sample21, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample21)
dim   (sample21)
sample21@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample21 [["percent.mt"]]   <- PercentageFeatureSet (sample21, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample21), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample21, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample21, slot   = 'counts')              );
sample21 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample21, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample21, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample21, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample21@meta.data
```

```{r}
# Filter sample nine (file 2)
fltsample21 <- subset (sample21, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample21, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample21, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample21, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample21
fltsample21
```

```{r}
normsample21 <- NormalizeData        ( fltsample21)
normsample21 <- ScaleData            (normsample21)
normsample21 <- FindVariableFeatures (normsample21, selection.method = "vst", nfeatures = 2000)
normsample21 <- RunPCA               (normsample21)
normsample21 <- RunUMAP              (normsample21, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample21 <- paramSweep_v3    (normsample21, PCs = 1:50, sct = FALSE)
sweep.stats_sample21    <- summarizeSweep   (sweep.res.list_sample21,  GT  = FALSE)
bcmvn_sample21          <- find.pK          (sweep.stats_sample21)
annotations21           <- normsample21@meta.data$ClusteringResults
homotypic.prop21        <- modelHomotypic   (annotations21)
nExp_poi21              <- round            (0.016 * length (normsample21@meta.data$ID))  ## Assuming 1.6% doublet formation rate
nExp_poi.adj21          <- round            (nExp_poi21 *   (1 - homotypic.prop21)     )
normsample21            <- doubletFinder_v3 (normsample21,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi21,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample21            <- doubletFinder_v3 (normsample21,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj21,
                                             reuse.pANN = "pANN_0.25_0.09_18",
                                             sct        = FALSE)
```

```{r}
normsample21@meta.data
DimPlot (normsample21, group.by = "DF.classifications_0.25_0.09_18")
table   (normsample21$DF.classifications_0.25_0.09_18)
```

```{r}
# Subset to remove identified cell doublets
fltsample21 <- subset (normsample21, subset = DF.classifications_0.25_0.09_18 == 'Singlet')
fltsample21
table (fltsample21$DF.classifications_0.25_0.09_18)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample21)
rm (sweep.res.list_sample21)
rm (sweep.stats_sample21)
rm (bcmvn_sample21)
rm (annotations21)
rm (homotypic.prop21)
rm (nExp_poi21)
rm (nExp_poi.adj21)
rm (normsample21)
rm (ldat21)
```

### Patient D9 Decidua

```{r}
# Patient D9

# Read in the data for decidual sample file four
sample22 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Vento-Tormo/Vento-Tormo_D9/CD45-")

# Read in the respective velocity data
          ldat22            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/FCA7196225.loom")
colnames (ldat22$spliced)   <- paste        (substring (colnames (ldat22$spliced),   12, 27), sep = "")
colnames (ldat22$spliced)   <- paste                   (colnames (ldat22$spliced),   "-1",    sep = "")
colnames (ldat22$unspliced) <- paste        (substring (colnames (ldat22$unspliced), 12, 27), sep = "")
colnames (ldat22$unspliced) <- paste                   (colnames (ldat22$unspliced), "-1",    sep = "")
colnames (ldat22$ambiguous) <- paste        (substring (colnames (ldat22$ambiguous), 12, 27), sep = "")
colnames (ldat22$ambiguous) <- paste                   (colnames (ldat22$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample22 <- CreateSeuratObject (counts = sample22, min.cells = 1, min.features = 1, project = "VT_D9.1")

for (i in names (x = ldat22)) { sample22[[i]] <- CreateAssayObject(counts = ldat22[[i]]) }

# Add the sample metadata
sample22 <- AddMetaData (sample22, metadata = "MPI",     col.name = "Model"    )
sample22 <- AddMetaData (sample22, metadata = "VT_D9",   col.name = "ID"       )
sample22 <- AddMetaData (sample22, metadata = "VT_D9.1", col.name = "rawID"    )
sample22 <- AddMetaData (sample22, metadata = "1",       col.name = "nSex"     )
sample22 <- AddMetaData (sample22, metadata = "Male",    col.name = "Sex"      )
sample22 <- AddMetaData (sample22, metadata = "12.0",    col.name = "rawGA"    )
sample22 <- AddMetaData (sample22, metadata = "Late",    col.name = "binGA"    )
sample22 <- AddMetaData (sample22, metadata = "2",       col.name = "nGA"      )
sample22 <- AddMetaData (sample22, metadata = "Late",    col.name = "GA"       )
sample22 <- AddMetaData (sample22, metadata = "Decidua", col.name = "Tissue"   )
sample22 <- AddMetaData (sample22, metadata = "3",       col.name = "Batch"    )
sample22 <- AddMetaData (sample22, metadata = "2",       col.name = "Source"   )
sample22 <- AddMetaData (sample22, metadata = "in vivo", col.name = "Type"     )
sample22 <- AddMetaData (sample22, metadata = "Third",   col.name = "Pregnancy")
sample22 <- AddMetaData (sample22, metadata = "CD45-",   col.name = "Gate"     )
sample22 <- AddMetaData (sample22, metadata = "First",   col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample22 <- NormalizeData      (sample22)

sample22 <- CellCycleScoring   (sample22, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample22)
dim   (sample22)
sample22@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample22 [["percent.mt"]]   <- PercentageFeatureSet (sample22, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample22), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample22, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample22, slot   = 'counts')              );
sample22 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample22, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample22, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample22, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample22@meta.data
```

```{r}
# Filter decidual file four
fltsample22 <- subset (sample22, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample22, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample22, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample22, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample22
fltsample22
```

```{r}
normsample22 <- NormalizeData        ( fltsample22)
normsample22 <- ScaleData            (normsample22)
normsample22 <- FindVariableFeatures (normsample22, selection.method = "vst", nfeatures = 2000)
normsample22 <- RunPCA               (normsample22)
normsample22 <- RunUMAP              (normsample22, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample22 <- paramSweep_v3    (normsample22, PCs = 1:50, sct = FALSE)
sweep.stats_sample22    <- summarizeSweep   (sweep.res.list_sample22,  GT  = FALSE)
bcmvn_sample22          <- find.pK          (sweep.stats_sample22)
annotations22           <- normsample22@meta.data$ClusteringResults
homotypic.prop22        <- modelHomotypic   (annotations22)
nExp_poi22              <- round            (0.034 * length (normsample22@meta.data$ID))  ## Assuming 2.5% doublet formation rate
nExp_poi.adj22          <- round            (nExp_poi22 *   (1-homotypic.prop22)       )
normsample22            <- doubletFinder_v3 (normsample22,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi22,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample22            <- doubletFinder_v3 (normsample22,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj22,
                                             reuse.pANN = "pANN_0.25_0.09_136",
                                             sct        = FALSE)
```

```{r}
normsample22@meta.data
DimPlot (normsample22, group.by = "DF.classifications_0.25_0.09_136")
table   (normsample22$DF.classifications_0.25_0.09_136)
```

```{r}
# Subset to remove identified cell doublets
fltsample22 <- subset (normsample22, subset = DF.classifications_0.25_0.09_136 == 'Singlet')
fltsample22
table (fltsample22$DF.classifications_0.25_0.09_136)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample22)
rm (sweep.res.list_sample22)
rm (sweep.stats_sample22)
rm (bcmvn_sample22)
rm (annotations22)
rm (homotypic.prop22)
rm (nExp_poi22)
rm (nExp_poi.adj22)
rm (normsample22)
rm (ldat22)
```

```{r}
# Patient D9

# Read in the data for decidual sample file five
sample23 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Vento-Tormo/Vento-Tormo_D9/CD45+")

# Read in the respective velocity data
          ldat23            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/FCA7196224.loom")
colnames (ldat23$spliced)   <- paste        (substring (colnames (ldat23$spliced),   12, 27), sep = "")
colnames (ldat23$spliced)   <- paste                   (colnames (ldat23$spliced),   "-1",    sep = "")
colnames (ldat23$unspliced) <- paste        (substring (colnames (ldat23$unspliced), 12, 27), sep = "")
colnames (ldat23$unspliced) <- paste                   (colnames (ldat23$unspliced), "-1",    sep = "")
colnames (ldat23$ambiguous) <- paste        (substring (colnames (ldat23$ambiguous), 12, 27), sep = "")
colnames (ldat23$ambiguous) <- paste                   (colnames (ldat23$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample23 <- CreateSeuratObject (counts = sample23, min.cells = 1, min.features = 1, project = "VT_D9.2")

for (i in names (x = ldat23)) { sample23[[i]] <- CreateAssayObject(counts = ldat23[[i]]) }

# Add the sample metadata
sample23 <- AddMetaData (sample23, metadata = "MPI",     col.name = "Model"    )
sample23 <- AddMetaData (sample23, metadata = "VT_D9",   col.name = "ID"       )
sample23 <- AddMetaData (sample23, metadata = "VT_D9.2", col.name = "rawID"    )
sample23 <- AddMetaData (sample23, metadata = "1",       col.name = "nSex"     )
sample23 <- AddMetaData (sample23, metadata = "Male",    col.name = "Sex"      )
sample23 <- AddMetaData (sample23, metadata = "12.0",    col.name = "rawGA"    )
sample23 <- AddMetaData (sample23, metadata = "Late",    col.name = "binGA"    )
sample23 <- AddMetaData (sample23, metadata = "2",       col.name = "nGA"      )
sample23 <- AddMetaData (sample23, metadata = "Late",    col.name = "GA"       )
sample23 <- AddMetaData (sample23, metadata = "Decidua", col.name = "Tissue"   )
sample23 <- AddMetaData (sample23, metadata = "3",       col.name = "Batch"    )
sample23 <- AddMetaData (sample23, metadata = "2",       col.name = "Source"   )
sample23 <- AddMetaData (sample23, metadata = "in vivo", col.name = "Type"     )
sample23 <- AddMetaData (sample23, metadata = "Third",   col.name = "Pregnancy")
sample23 <- AddMetaData (sample23, metadata = "CD45+",   col.name = "Gate"     )
sample23 <- AddMetaData (sample23, metadata = "First",   col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample23 <- NormalizeData    (sample23)
sample23 <- CellCycleScoring (sample23, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample23)
dim   (sample23)
sample23@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample23 [["percent.mt"]]   <- PercentageFeatureSet (sample23, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample23), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample23, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample23, slot   = 'counts')              );
sample23 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample23, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample23, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample23, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample23@meta.data
```

```{r}
# Filter decidual file five
fltsample23 <- subset (sample23, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample23, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample23, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample23, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample23
fltsample23
```

```{r}
normsample23 <- NormalizeData        ( fltsample23)
normsample23 <- ScaleData            (normsample23)
normsample23 <- FindVariableFeatures (normsample23, selection.method = "vst", nfeatures = 2000)
normsample23 <- RunPCA               (normsample23)
normsample23 <- RunUMAP              (normsample23, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample23 <- paramSweep_v3    (normsample23, PCs = 1:50, sct = FALSE)
sweep.stats_sample23    <- summarizeSweep   (sweep.res.list_sample23,  GT  = FALSE)
bcmvn_sample23          <- find.pK          (sweep.stats_sample23)
annotations23           <- normsample23@meta.data$ClusteringResults
homotypic.prop23        <- modelHomotypic   (annotations23)
nExp_poi23              <- round            (0.061 * length (normsample23@meta.data$ID))  ## Assuming 6.1% doublet formation rate
nExp_poi.adj23          <- round            (nExp_poi23 *   (1-homotypic.prop23)       )
normsample23            <- doubletFinder_v3 (normsample23,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi23,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample23            <- doubletFinder_v3 (normsample23,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj23,
                                             reuse.pANN = "pANN_0.25_0.09_436",
                                             sct        = FALSE)
```

```{r}
normsample23@meta.data
DimPlot (normsample23, group.by = "DF.classifications_0.25_0.09_436")
table   (normsample23$DF.classifications_0.25_0.09_436)
```

```{r}
# Subset to remove identified cell doublets
fltsample23 <- subset (normsample23, subset = DF.classifications_0.25_0.09_436 == 'Singlet')
fltsample23
table (fltsample23$DF.classifications_0.25_0.09_436)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample23)
rm (sweep.res.list_sampl23)
rm (sweep.stats_sample23)
rm (bcmvn_sample23)
rm (annotations23)
rm (homotypic.prop23)
rm (nExp_poi23)
rm (nExp_poi.adj23)
rm (normsample23)
rm (ldat23)
```

### Patient D9 Maternal peripheral blood

```{r}
# Patient D9

# Read in the data for placental villi sample ten (file 3)
sample24 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Vento-Tormo/Vento-Tormo_B9/PB")

# Read in the respective velocity data
          ldat24            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/FCA7196229.loom")
colnames (ldat24$spliced)   <- paste        (substring (colnames (ldat24$spliced),   12, 27), sep = "")
colnames (ldat24$spliced)   <- paste                   (colnames (ldat24$spliced),   "-1",    sep = "")
colnames (ldat24$unspliced) <- paste        (substring (colnames (ldat24$unspliced), 12, 27), sep = "")
colnames (ldat24$unspliced) <- paste                   (colnames (ldat24$unspliced), "-1",    sep = "")
colnames (ldat24$ambiguous) <- paste        (substring (colnames (ldat24$ambiguous), 12, 27), sep = "")
colnames (ldat24$ambiguous) <- paste                   (colnames (ldat24$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample24 <- CreateSeuratObject (counts = sample24, min.cells = 1, min.features = 1, project = "VT_B9.1")

for (i in names (x = ldat24)) { sample24[[i]] <- CreateAssayObject (counts = ldat24[[i]]) }

# Add the sample metadata
sample24 <- AddMetaData (sample24, metadata = "MPI",     col.name = "Model"    )
sample24 <- AddMetaData (sample24, metadata = "VT_D9",   col.name = "ID"       )
sample24 <- AddMetaData (sample24, metadata = "VT_B9.1", col.name = "rawID"    )
sample24 <- AddMetaData (sample24, metadata = "1",       col.name = "nSex"     )
sample24 <- AddMetaData (sample24, metadata = "Male",    col.name = "Sex"      )
sample24 <- AddMetaData (sample24, metadata = "12.0",    col.name = "rawGA"    )
sample24 <- AddMetaData (sample24, metadata = "Late",    col.name = "binGA"    )
sample24 <- AddMetaData (sample24, metadata = "2",       col.name = "nGA"      )
sample24 <- AddMetaData (sample24, metadata = "Late",    col.name = "GA"       )
sample24 <- AddMetaData (sample24, metadata = "PB",      col.name = "Tissue"   )
sample24 <- AddMetaData (sample24, metadata = "3",       col.name = "Batch"    )
sample24 <- AddMetaData (sample24, metadata = "2",       col.name = "Source"   )
sample24 <- AddMetaData (sample24, metadata = "in vivo", col.name = "Type"     )
sample24 <- AddMetaData (sample24, metadata = "Third",   col.name = "Pregnancy")
sample24 <- AddMetaData (sample24, metadata = "PB",      col.name = "Gate"     )
sample24 <- AddMetaData (sample24, metadata = "First",   col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample24 <- NormalizeData    (sample24)
sample24 <- CellCycleScoring (sample24, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample24)
dim   (sample24)
sample24@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample24 [["percent.mt"]]   <- PercentageFeatureSet (sample24, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample24), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample24, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample24, slot   = 'counts')              );
sample24 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample24, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample24, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample24, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample24@meta.data
```

```{r}
# Filter sample nine (file 2)
fltsample24 <- subset (sample24, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample24, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample24, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample24, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample24
fltsample24
```

```{r}
normsample24 <- NormalizeData        ( fltsample24)
normsample24 <- ScaleData            (normsample24)
normsample24 <- FindVariableFeatures (normsample24, selection.method = "vst", nfeatures = 2000)
normsample24 <- RunPCA               (normsample24)
normsample24 <- RunUMAP              (normsample24, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample24 <- paramSweep_v3    (normsample24, PCs = 1:50, sct = FALSE)
sweep.stats_sample24    <- summarizeSweep   (sweep.res.list_sample24,  GT  = FALSE)
bcmvn_sample24          <- find.pK          (sweep.stats_sample24)
annotations24           <- normsample24@meta.data$ClusteringResults
homotypic.prop24        <- modelHomotypic   (annotations24)
nExp_poi24              <- round            (0.023 * length (normsample24@meta.data$ID))  ## Assuming 2.3% doublet formation rate
nExp_poi.adj24          <- round            (nExp_poi24 *   (1 - homotypic.prop24)     )
normsample24            <- doubletFinder_v3 (normsample24,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi24,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample24            <- doubletFinder_v3 (normsample24,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj24,
                                             reuse.pANN = "pANN_0.25_0.09_70",
                                             sct        = FALSE)
```

```{r}
normsample24@meta.data
DimPlot (normsample24, group.by = "DF.classifications_0.25_0.09_70")
table   (normsample24$DF.classifications_0.25_0.09_70)
```

```{r}
# Subset to remove identified cell doublets
fltsample24 <- subset (normsample24, subset = DF.classifications_0.25_0.09_70 == 'Singlet')
fltsample24
table (fltsample24$DF.classifications_0.25_0.09_70)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample24)
rm (sweep.res.list_sample24)
rm (sweep.stats_sample24)
rm (bcmvn_sample24)
rm (annotations24)
rm (homotypic.prop24)
rm (nExp_poi24)
rm (nExp_poi.adj24)
rm (normsample24)
rm (ldat24)
```

### Patient D10 Decidua

```{r}
# Patient D10

# Read in the data for decidual sample file six
sample25 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Vento-Tormo/Vento-Tormo_D10/CD45-")

# Read in the respective velocity data
          ldat25            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/FCA7474063.loom")
colnames (ldat25$spliced)   <- paste        (substring (colnames (ldat25$spliced),   12, 27), sep = "")
colnames (ldat25$spliced)   <- paste                   (colnames (ldat25$spliced),   "-1",    sep = "")
colnames (ldat25$unspliced) <- paste        (substring (colnames (ldat25$unspliced), 12, 27), sep = "")
colnames (ldat25$unspliced) <- paste                   (colnames (ldat25$unspliced), "-1",    sep = "")
colnames (ldat25$ambiguous) <- paste        (substring (colnames (ldat25$ambiguous), 12, 27), sep = "")
colnames (ldat25$ambiguous) <- paste                   (colnames (ldat25$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample25 <- CreateSeuratObject (counts = sample25, min.cells = 1, min.features = 1, project = "VT_D10.1")

for (i in names (x = ldat25)) { sample25[[i]] <- CreateAssayObject(counts = ldat25[[i]]) }

# Add the sample metadata
sample25 <- AddMetaData (sample25, metadata = "MPI",      col.name = "Model"    )
sample25 <- AddMetaData (sample25, metadata = "VT_D10",   col.name = "ID"       )
sample25 <- AddMetaData (sample25, metadata = "VT_D10.1", col.name = "rawID"    )
sample25 <- AddMetaData (sample25, metadata = "1",        col.name = "nSex"     )
sample25 <- AddMetaData (sample25, metadata = "Male",     col.name = "Sex"      )
sample25 <- AddMetaData (sample25, metadata = "9.6",      col.name = "rawGA"    )
sample25 <- AddMetaData (sample25, metadata = "Mid",      col.name = "binGA"    )
sample25 <- AddMetaData (sample25, metadata = "2",        col.name = "nGA"      )
sample25 <- AddMetaData (sample25, metadata = "Late",     col.name = "GA"       )
sample25 <- AddMetaData (sample25, metadata = "Decidua",  col.name = "Tissue"   )
sample25 <- AddMetaData (sample25, metadata = "3",        col.name = "Batch"    )
sample25 <- AddMetaData (sample25, metadata = "2",        col.name = "Source"   )
sample25 <- AddMetaData (sample25, metadata = "in vivo",  col.name = "Type"     )
sample25 <- AddMetaData (sample25, metadata = "Second",   col.name = "Pregnancy")
sample25 <- AddMetaData (sample25, metadata = "CD45-",    col.name = "Gate"     )
sample25 <- AddMetaData (sample25, metadata = "First",    col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample25 <- NormalizeData    (sample25)
sample25 <- CellCycleScoring (sample25, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample25)
dim   (sample25)
sample25@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample25 [["percent.mt"]]   <- PercentageFeatureSet (sample25, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample25), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample25, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample25, slot   = 'counts')              );
sample25 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample25, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample25, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample25, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample25@meta.data
```

```{r}
# Filter decidual file six
fltsample25 <- subset (sample25, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample25, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample25, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample25, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample25
fltsample25
```

```{r}
normsample25 <- NormalizeData        ( fltsample25)
normsample25 <- ScaleData            (normsample25)
normsample25 <- FindVariableFeatures (normsample25, selection.method = "vst", nfeatures = 2000)
normsample25 <- RunPCA               (normsample25)
normsample25 <- RunUMAP              (normsample25, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample25 <- paramSweep_v3    (normsample25, PCs = 1:50, sct = FALSE)
sweep.stats_sample25    <- summarizeSweep   (sweep.res.list_sample25,  GT  = FALSE)
bcmvn_sample25          <- find.pK          (sweep.stats_sample25)
annotations25           <- normsample25@meta.data$ClusteringResults
homotypic.prop25        <- modelHomotypic   (annotations25)
nExp_poi25              <- round            (0.018 * length (normsample25@meta.data$ID))  ## Assuming 1.8% doublet formation rate
nExp_poi.adj25          <- round            (nExp_poi25 *   (1-homotypic.prop25)       )
normsample25            <- doubletFinder_v3 (normsample25,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi25,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample25            <- doubletFinder_v3 (normsample25,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj25,
                                             reuse.pANN = "pANN_0.25_0.09_41",
                                             sct        = FALSE)
```

```{r}
normsample25@meta.data
DimPlot (normsample25, group.by = "DF.classifications_0.25_0.09_41")
table   (normsample25$DF.classifications_0.25_0.09_41)
```

```{r}
# Subset to remove identified cell doublets
fltsample25 <- subset (normsample25, subset = DF.classifications_0.25_0.09_41 == 'Singlet')
fltsample25
table (fltsample25$DF.classifications_0.25_0.09_41)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample25)
rm (sweep.res.list_sample25)
rm (sweep.stats_sample25)
rm (bcmvn_sample25)
rm (annotations25)
rm (homotypic.prop25)
rm (nExp_poi25)
rm (nExp_poi.adj25)
rm (normsample25)
rm (ldat25)
```

```{r}
# Patient D10

# Read in the data for decidual sample file seven
sample26 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Vento-Tormo/Vento-Tormo_D10/CD45+")

# Read in the respective velocity data
          ldat26            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/FCA7474062.loom")
colnames (ldat26$spliced)   <- paste        (substring (colnames (ldat26$spliced),   12, 27), sep = "")
colnames (ldat26$spliced)   <- paste                   (colnames (ldat26$spliced),   "-1",    sep = "")
colnames (ldat26$unspliced) <- paste        (substring (colnames (ldat26$unspliced), 12, 27), sep = "")
colnames (ldat26$unspliced) <- paste                   (colnames (ldat26$unspliced), "-1",    sep = "")
colnames (ldat26$ambiguous) <- paste        (substring (colnames (ldat26$ambiguous), 12, 27), sep = "")
colnames (ldat26$ambiguous) <- paste                   (colnames (ldat26$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample26 <- CreateSeuratObject (counts = sample26, min.cells = 1, min.features = 1, project = "VT_D10.2")

for (i in names (x = ldat26)) { sample26[[i]] <- CreateAssayObject(counts = ldat26[[i]]) }

# Add the sample metadata
sample26 <- AddMetaData (sample26, metadata = "MPI",      col.name = "Model"    )
sample26 <- AddMetaData (sample26, metadata = "VT_D10",   col.name = "ID"       )
sample26 <- AddMetaData (sample26, metadata = "VT_D10.2", col.name = "rawID"    )
sample26 <- AddMetaData (sample26, metadata = "1",        col.name = "nSex"     )
sample26 <- AddMetaData (sample26, metadata = "Male",     col.name = "Sex"      )
sample26 <- AddMetaData (sample26, metadata = "9.6",      col.name = "rawGA"    )
sample26 <- AddMetaData (sample26, metadata = "Mid",      col.name = "binGA"    )
sample26 <- AddMetaData (sample26, metadata = "2",        col.name = "nGA"      )
sample26 <- AddMetaData (sample26, metadata = "Late",     col.name = "GA"       )
sample26 <- AddMetaData (sample26, metadata = "Decidua",  col.name = "Tissue"   )
sample26 <- AddMetaData (sample26, metadata = "3",        col.name = "Batch"    )
sample26 <- AddMetaData (sample26, metadata = "2",        col.name = "Source"   )
sample26 <- AddMetaData (sample26, metadata = "in vivo",  col.name = "Type"     )
sample26 <- AddMetaData (sample26, metadata = "Second",   col.name = "Pregnancy")
sample26 <- AddMetaData (sample26, metadata = "CD45+",    col.name = "Gate"     )
sample26 <- AddMetaData (sample26, metadata = "First",    col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample26 <- NormalizeData    (sample26)
sample26 <- CellCycleScoring (sample26, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample26)
dim   (sample26)
sample26@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample26 [["percent.mt"]]   <- PercentageFeatureSet (sample26, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample26), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample26, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample26, slot   = 'counts')              );
sample26 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample26, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample26, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample26, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample26@meta.data
```

```{r}
# Filter decidual file seven
fltsample26 <- subset (sample26, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample26, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample26, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample26, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample26
fltsample26
```

```{r}
normsample26 <- NormalizeData        ( fltsample26)
normsample26 <- ScaleData            (normsample26)
normsample26 <- FindVariableFeatures (normsample26, selection.method = "vst", nfeatures = 2000)
normsample26 <- RunPCA               (normsample26)
normsample26 <- RunUMAP              (normsample26, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample26 <- paramSweep_v3    (normsample26, PCs = 1:50, sct = FALSE)
sweep.stats_sample26    <- summarizeSweep   (sweep.res.list_sample26,  GT  = FALSE)
bcmvn_sample26          <- find.pK          (sweep.stats_sample26)
annotations26           <- normsample26@meta.data$ClusteringResults
homotypic.prop26        <- modelHomotypic   (annotations26)
nExp_poi26              <- round            (0.024 * length (normsample26@meta.data$ID))  ## Assuming 2.4% doublet formation rate
nExp_poi.adj26          <- round            (nExp_poi26 *   (1-homotypic.prop26)       )
normsample26            <- doubletFinder_v3 (normsample26,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi26,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample26normsample26            <- doubletFinder_v3 (normsample26,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj26,
                                             reuse.pANN = "pANN_0.25_0.09_50",
                                             sct        = FALSE)
```

```{r}
normsample26@meta.data
DimPlot (normsample26, group.by = "DF.classifications_0.25_0.09_50")
table   (normsample26$DF.classifications_0.25_0.09_50)
```

```{r}
# Subset to remove identified cell doublets
fltsample26 <- subset (normsample26, subset = DF.classifications_0.25_0.09_50 == 'Singlet')
fltsample26
table (fltsample26$DF.classifications_0.25_0.09_50)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample26)
rm (sweep.res.list_sample26)
rm (sweep.stats_sample26)
rm (bcmvn_sample26)
rm (annotations26)
rm (homotypic.prop26)
rm (nExp_poi26)
rm (nExp_poi.adj26)
rm (normsample26)
rm (ldat26)
```

### Patient D12 Decidua

```{r}
# Patient D12

# Read in the data for decidual sample file eight
sample27 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Vento-Tormo/Vento-Tormo_D12/CD45-")

# Read in the respective velocity data
          ldat27            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/FCA7511882.loom")
colnames (ldat27$spliced)   <- paste        (substring (colnames (ldat27$spliced),   12, 27), sep = "")
colnames (ldat27$spliced)   <- paste                   (colnames (ldat27$spliced),   "-1",    sep = "")
colnames (ldat27$unspliced) <- paste        (substring (colnames (ldat27$unspliced), 12, 27), sep = "")
colnames (ldat27$unspliced) <- paste                   (colnames (ldat27$unspliced), "-1",    sep = "")
colnames (ldat27$ambiguous) <- paste        (substring (colnames (ldat27$ambiguous), 12, 27), sep = "")
colnames (ldat27$ambiguous) <- paste                   (colnames (ldat27$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample27 <- CreateSeuratObject (counts = sample27, min.cells = 1, min.features = 1, project = "VT_D12.1")

for (i in names (x = ldat27)) { sample27[[i]] <- CreateAssayObject(counts = ldat27[[i]]) }

# Add the sample metadata
sample27 <- AddMetaData (sample27, metadata = "MPI",      col.name = "Model"    )
sample27 <- AddMetaData (sample27, metadata = "VT_D12",   col.name = "ID"       )
sample27 <- AddMetaData (sample27, metadata = "VT_D12.1", col.name = "rawID"    )
sample27 <- AddMetaData (sample27, metadata = "0",        col.name = "nSex"     )
sample27 <- AddMetaData (sample27, metadata = "Female",   col.name = "Sex"      )
sample27 <- AddMetaData (sample27, metadata = "10.0",     col.name = "rawGA"    )
sample27 <- AddMetaData (sample27, metadata = "Mid",      col.name = "binGA"    )
sample27 <- AddMetaData (sample27, metadata = "2",        col.name = "nGA"      )
sample27 <- AddMetaData (sample27, metadata = "Late",     col.name = "GA"       )
sample27 <- AddMetaData (sample27, metadata = "Decidua",  col.name = "Tissue"   )
sample27 <- AddMetaData (sample27, metadata = "3",        col.name = "Batch"    )
sample27 <- AddMetaData (sample27, metadata = "2",        col.name = "Source"   )
sample27 <- AddMetaData (sample27, metadata = "in vivo",  col.name = "Type"     )
sample27 <- AddMetaData (sample27, metadata = "Fourth",   col.name = "Pregnancy")
sample27 <- AddMetaData (sample27, metadata = "CD45-",    col.name = "Gate"     )
sample27 <- AddMetaData (sample27, metadata = "First",    col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample27 <- NormalizeData    (sample27)
sample27 <- CellCycleScoring (sample27, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample27)
dim   (sample27)
sample27@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample27 [["percent.mt"]]   <- PercentageFeatureSet (sample27, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample27), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample27, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample27, slot   = 'counts')              );
sample27 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample27, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample27, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample27, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample27@meta.data
```

```{r}
# Filter decidual file eight
fltsample27 <- subset (sample27, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample27, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample27, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample27, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample27
fltsample27
```

```{r}
normsample27 <- NormalizeData        ( fltsample27)
normsample27 <- ScaleData            (normsample27)
normsample27 <- FindVariableFeatures (normsample27, selection.method = "vst", nfeatures = 2000)
normsample27 <- RunPCA               (normsample27)
normsample27 <- RunUMAP              (normsample27, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample27 <- paramSweep_v3    (normsample27, PCs = 1:50, sct = FALSE)
sweep.stats_sample27    <- summarizeSweep   (sweep.res.list_sample27,  GT  = FALSE)
bcmvn_sample27          <- find.pK          (sweep.stats_sample27)
annotations27           <- normsample27@meta.data$ClusteringResults
homotypic.prop27        <- modelHomotypic   (annotations27)
nExp_poi27              <- round            (0.017 * length (normsample27@meta.data$ID))  ## Assuming 1.6% doublet formation rate
nExp_poi.adj27          <- round            (nExp_poi27 *   (1-homotypic.prop27)       )
normsample27            <- doubletFinder_v3 (normsample27,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi27,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample27            <- doubletFinder_v3 (normsample27,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj27,
                                             reuse.pANN = "pANN_0.25_0.09_36",
                                             sct        = FALSE)
```

```{r}
normsample27@meta.data
DimPlot (normsample27, group.by = "DF.classifications_0.25_0.09_36")
table   (normsample27$DF.classifications_0.25_0.09_36)
```

```{r}
# Subset to remove identified cell doublets
fltsample27 <- subset (normsample27, subset = DF.classifications_0.25_0.09_36 == 'Singlet')
fltsample27
table (fltsample27$DF.classifications_0.25_0.09_36)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample27)
rm (sweep.res.list_sample27)
rm (sweep.stats_sample27)
rm (bcmvn_sample27)
rm (annotations27)
rm (homotypic.prop27)
rm (nExp_poi27)
rm (nExp_poi.adj27)
rm (normsample27)
rm (ldat27)
```

```{r}
# Patient D12

# Read in the data for decidual sample file nine
sample28 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Vento-Tormo/Vento-Tormo_D12/CD45+")

# Read in the respective velocity data
          ldat28            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/FCA7511881.loom")
colnames (ldat28$spliced)   <- paste        (substring (colnames (ldat28$spliced),   12, 27), sep = "")
colnames (ldat28$spliced)   <- paste                   (colnames (ldat28$spliced),   "-1",    sep = "")
colnames (ldat28$unspliced) <- paste        (substring (colnames (ldat28$unspliced), 12, 27), sep = "")
colnames (ldat28$unspliced) <- paste                   (colnames (ldat28$unspliced), "-1",    sep = "")
colnames (ldat28$ambiguous) <- paste        (substring (colnames (ldat28$ambiguous), 12, 27), sep = "")
colnames (ldat28$ambiguous) <- paste                   (colnames (ldat28$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample28 <- CreateSeuratObject (counts = sample28, min.cells = 1, min.features = 1, project = "VT_D12.2")

for (i in names (x = ldat28)) { sample28[[i]] <- CreateAssayObject (counts = ldat28[[i]]) }

# Add the sample metadata
sample28 <- AddMetaData (sample28, metadata = "MPI",      col.name = "Model"    )
sample28 <- AddMetaData (sample28, metadata = "VT_D12",   col.name = "ID"       )
sample28 <- AddMetaData (sample28, metadata = "VT_D12.2", col.name = "rawID"    )
sample28 <- AddMetaData (sample28, metadata = "0",        col.name = "nSex"     )
sample28 <- AddMetaData (sample28, metadata = "Female",   col.name = "Sex"      )
sample28 <- AddMetaData (sample28, metadata = "10.0",     col.name = "rawGA"    )
sample28 <- AddMetaData (sample28, metadata = "Mid",      col.name = "binGA"    )
sample28 <- AddMetaData (sample28, metadata = "2",        col.name = "nGA"      )
sample28 <- AddMetaData (sample28, metadata = "Late",     col.name = "GA"       )
sample28 <- AddMetaData (sample28, metadata = "Decidua",  col.name = "Tissue"   )
sample28 <- AddMetaData (sample28, metadata = "3",        col.name = "Batch"    )
sample28 <- AddMetaData (sample28, metadata = "2",        col.name = "Source"   )
sample28 <- AddMetaData (sample28, metadata = "in vivo",  col.name = "Type"     )
sample28 <- AddMetaData (sample28, metadata = "Fourth",   col.name = "Pregnancy")
sample28 <- AddMetaData (sample28, metadata = "CD45+",    col.name = "Gate"     )
sample28 <- AddMetaData (sample28, metadata = "First",    col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample28 <- NormalizeData    (sample28)
sample28 <- CellCycleScoring (sample28, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample28)
dim   (sample28)
sample28@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample28 [["percent.mt"]]   <- PercentageFeatureSet (sample28, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample28), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample28, slot   = 'counts')[ribo.genes, ]) / 
                               Matrix::colSums (x = GetAssayData (sample28, slot   = 'counts')              );
sample28 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample28, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample28, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample28, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample28@meta.data
```

```{r}
# Filter decidual file nine
fltsample28 <- subset (sample28, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample28, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample28, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample28, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample28
fltsample28
```

```{r}
normsample28 <- NormalizeData        ( fltsample28)
normsample28 <- ScaleData            (normsample28)
normsample28 <- FindVariableFeatures (normsample28, selection.method = "vst", nfeatures = 2000)
normsample28 <- RunPCA               (normsample28)
normsample28 <- RunUMAP              (normsample28, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample28 <- paramSweep_v3    (normsample28, PCs = 1:50, sct = FALSE)
sweep.stats_sample28    <- summarizeSweep   (sweep.res.list_sample28,  GT  = FALSE)
bcmvn_sample28          <- find.pK          (sweep.stats_sample28)
annotations28           <- normsample28@meta.data$ClusteringResults
homotypic.prop28        <- modelHomotypic   (annotations28)
nExp_poi28              <- round            (0.016 * length (normsample28@meta.data$ID))  ## Assuming 1.6% doublet formation rate
nExp_poi.adj28          <- round            (nExp_poi28 *   (1-homotypic.prop28)       )
normsample28            <- doubletFinder_v3 (normsample28,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi28,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample28            <- doubletFinder_v3 (normsample28,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj28,
                                             reuse.pANN = "pANN_0.25_0.09_20",
                                             sct        = FALSE)
```

```{r}
normsample28@meta.data
DimPlot (normsample28, group.by = "DF.classifications_0.25_0.09_20")
table   (normsample28$DF.classifications_0.25_0.09_20)
```

```{r}
# Subset to remove identified cell doublets
fltsample28 <- subset (normsample28, subset = DF.classifications_0.25_0.09_20 == 'Singlet')
fltsample28
table (fltsample28$DF.classifications_0.25_0.09_20)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample28)
rm (sweep.res.list_sample28)
rm (sweep.stats_sample28)
rm (bcmvn_sample28)
rm (annotations28)
rm (homotypic.prop28)
rm (nExp_poi28)
rm (nExp_poi.adj28)
rm (normsample28)
rm (ldat28)
```

```{r}
# Patient D12

# Read in the data for placental villi sample ten (file 3)
sample29 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/1_Vento-Tormo/Vento-Tormo_D12/HLA-G+")

# Read in the respective velocity data
          ldat29            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/FCA7511883.loom")
colnames (ldat29$spliced)   <- paste        (substring (colnames (ldat29$spliced),   12, 27), sep = "")
colnames (ldat29$spliced)   <- paste                   (colnames (ldat29$spliced),   "-1",    sep = "")
colnames (ldat29$unspliced) <- paste        (substring (colnames (ldat29$unspliced), 12, 27), sep = "")
colnames (ldat29$unspliced) <- paste                   (colnames (ldat29$unspliced), "-1",    sep = "")
colnames (ldat29$ambiguous) <- paste        (substring (colnames (ldat29$ambiguous), 12, 27), sep = "")
colnames (ldat29$ambiguous) <- paste                   (colnames (ldat29$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample29 <- CreateSeuratObject (counts = sample29, min.cells = 1, min.features = 1, project = "VT_D12.3")

for (i in names (x = ldat29)) { sample29[[i]] <- CreateAssayObject (counts = ldat29[[i]]) }

# Add the sample metadata
sample29 <- AddMetaData (sample29, metadata = "MPI",      col.name = "Model"    )
sample29 <- AddMetaData (sample29, metadata = "VT_D12",   col.name = "ID"       )
sample29 <- AddMetaData (sample29, metadata = "VT_D12.3", col.name = "rawID"    )
sample29 <- AddMetaData (sample29, metadata = "0",        col.name = "nSex"     )
sample29 <- AddMetaData (sample29, metadata = "Female",   col.name = "Sex"      )
sample29 <- AddMetaData (sample29, metadata = "10.0",     col.name = "rawGA"    )
sample29 <- AddMetaData (sample29, metadata = "Mid",      col.name = "binGA"    )
sample29 <- AddMetaData (sample29, metadata = "2",        col.name = "nGA"      )
sample29 <- AddMetaData (sample29, metadata = "Late",     col.name = "GA"       )
sample29 <- AddMetaData (sample29, metadata = "Decidua",  col.name = "Tissue"   )
sample29 <- AddMetaData (sample29, metadata = "3",        col.name = "Batch"    )
sample29 <- AddMetaData (sample29, metadata = "2",        col.name = "Source"   )
sample29 <- AddMetaData (sample29, metadata = "in vivo",  col.name = "Type"     )
sample29 <- AddMetaData (sample29, metadata = "Fourth",   col.name = "Pregnancy")
sample29 <- AddMetaData (sample29, metadata = "HLA-G+",   col.name = "Gate"     )
sample29 <- AddMetaData (sample29, metadata = "First",    col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample29 <- NormalizeData    (sample29)
sample29 <- CellCycleScoring (sample29, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample29)
dim   (sample29)
sample29@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample29 [["percent.mt"]]   <- PercentageFeatureSet (sample29, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample29), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample29, slot   = 'counts')[ribo.genes, ]) / 
                              Matrix::colSums (x = GetAssayData (sample29, slot   = 'counts')              );
sample29 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample29, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample29, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample29, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample29@meta.data
```

```{r}
# Filter sample nine (file 2)
fltsample29 <- subset (sample29, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample29, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample29, features = c("percent.mt"),   group.by = "ID")
VlnPlot (fltsample29, features = c("percent.ribo"), group.by = "ID")

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample29
fltsample29
```

```{r}
normsample29 <- NormalizeData        ( fltsample29)
normsample29 <- ScaleData            (normsample29)
normsample29 <- FindVariableFeatures (normsample29, selection.method = "vst", nfeatures = 2000)
normsample29 <- RunPCA               (normsample29)
normsample29 <- RunUMAP              (normsample29, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample29 <- paramSweep_v3    (normsample29, PCs = 1:50, sct = FALSE)
sweep.stats_sample29    <- summarizeSweep   (sweep.res.list_sample29,  GT  = FALSE)
bcmvn_sample29          <- find.pK          (sweep.stats_sample29)
annotations29           <- normsample29@meta.data$ClusteringResults
homotypic.prop29        <- modelHomotypic   (annotations29)
nExp_poi29              <- round            (0.004 * length (normsample29@meta.data$ID))  ## Assuming 0.4% doublet formation rate
nExp_poi.adj29          <- round            (nExp_poi29 *   (1 - homotypic.prop29)     )
normsample29            <- doubletFinder_v3 (normsample29,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi29,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample29            <- doubletFinder_v3 (normsample29,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj29,
                                             reuse.pANN = "pANN_0.25_0.09_1",
                                             sct        = FALSE)
```

```{r}
normsample29@meta.data
DimPlot (normsample29, group.by = "DF.classifications_0.25_0.09_1")
table   (normsample29$DF.classifications_0.25_0.09_1)
```

```{r}
# Subset to remove identified cell doublets
fltsample29 <- subset (normsample29, subset = DF.classifications_0.25_0.09_1 == 'Singlet')
fltsample29
table (fltsample29$DF.classifications_0.25_0.09_1)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample29)
rm (sweep.res.list_sample29)
rm (sweep.stats_sample29)
rm (bcmvn_sample29)
rm (annotations29)
rm (homotypic.prop29)
rm (nExp_poi29)
rm (nExp_poi.adj29)
rm (normsample29)
rm (ldat29)
```

# Human Trophoblast Stem Cell Derived Organoid Dataset Prepation, Quality Control, and Sample Integration

## Read in the hTSC organoid data

### Undifferentiated

```{r}
# Read in the data for trophoblast organoid sample one
sample30 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/3_Organoid/Undifferentiated/Stem_Cell_Derived/CT27")

# Read in the respective velocity data
          ldat30            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/TS_Organoids__CT27.loom")
colnames (ldat30$spliced)   <- paste        (substring (colnames (ldat30$spliced),   20, 35), sep = "")
colnames (ldat30$spliced)   <- paste                   (colnames (ldat30$spliced),   "-1",    sep = "")
colnames (ldat30$unspliced) <- paste        (substring (colnames (ldat30$unspliced), 20, 35), sep = "")
colnames (ldat30$unspliced) <- paste                   (colnames (ldat30$unspliced), "-1",    sep = "")
colnames (ldat30$ambiguous) <- paste        (substring (colnames (ldat30$ambiguous), 20, 35), sep = "")
colnames (ldat30$ambiguous) <- paste                   (colnames (ldat30$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample30 <- CreateSeuratObject (counts = sample30, min.cells = 1, min.features = 1, project = "CT27")

for (i in names (x = ldat30)) { sample30[[i]] <- CreateAssayObject(counts = ldat30[[i]]) }

# Add the sample metadata
sample30 <- AddMetaData (sample30, metadata = "hTSC.TOrg",        col.name = "Model"    )
sample30 <- AddMetaData (sample30, metadata = "CT27",             col.name = "ID"       )
sample30 <- AddMetaData (sample30, metadata = "UCT27",            col.name = "rawID"    )
sample30 <- AddMetaData (sample30, metadata = "Stem Cell",        col.name = "Source"   )
sample30 <- AddMetaData (sample30, metadata = "0",                col.name = "nSex"     )
sample30 <- AddMetaData (sample30, metadata = "Female",           col.name = "Sex"      )
sample30 <- AddMetaData (sample30, metadata = "1",                col.name = "Time"     )
sample30 <- AddMetaData (sample30, metadata = "Undifferentiated", col.name = "State"    )
sample30 <- AddMetaData (sample30, metadata = "hTSC Organoid",    col.name = "Tissue"   )
sample30 <- AddMetaData (sample30, metadata = "4",                col.name = "Batch"    )
sample30 <- AddMetaData (sample30, metadata = "4",                col.name = "Source"   )
sample30 <- AddMetaData (sample30, metadata = "in vitro",         col.name = "Type"     )
sample30 <- AddMetaData (sample30, metadata = "All",              col.name = "Gate"     )
sample30 <- AddMetaData (sample30, metadata = "in vitro",         col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample30 <- NormalizeData    (sample30)
sample30 <- CellCycleScoring (sample30, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample30)
dim   (sample30)
sample30@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample30 [["percent.mt"]]   <- PercentageFeatureSet (sample30, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample30), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample30, slot   = 'counts')[ribo.genes, ]) / 
                               Matrix::colSums (x = GetAssayData (sample30, slot   = 'counts')              );
sample30 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample30, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample30, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample30, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample30@meta.data
```

```{r}
# Filter trophoblast organoid sample one
fltsample30 <- subset (sample30, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample30, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample30, features = c("percent.mt"))
VlnPlot (fltsample30, features = c("percent.ribo"))

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample30
fltsample30
```

```{r}
normsample30 <- NormalizeData        ( fltsample30)
normsample30 <- ScaleData            (normsample30)
normsample30 <- FindVariableFeatures (normsample30, selection.method = "vst", nfeatures = 2000)
normsample30 <- RunPCA               (normsample30)
normsample30 <- RunUMAP              (normsample30, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample30 <- paramSweep_v3    (normsample30, PCs = 1:50, sct = FALSE)
sweep.stats_sample30    <- summarizeSweep   (sweep.res.list_sample30,  GT  = FALSE)
bcmvn_sample30          <- find.pK          (sweep.stats_sample30)
annotations30           <- normsample30@meta.data$ClusteringResults
homotypic.prop30        <- modelHomotypic   (annotations30)
nExp_poi30              <- round            (0.016 * length (normsample30@meta.data$ID))  ## Assuming 1.6% doublet formation rate
nExp_poi.adj30          <- round            (nExp_poi30 *    (1 - homotypic.prop30)    )
normsample30            <- doubletFinder_v3 (normsample30,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi30,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample30            <- doubletFinder_v3 (normsample30,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09, 
                                             nExp       = nExp_poi.adj30,
                                             reuse.pANN = "pANN_0.25_0.09_29",
                                             sct        = FALSE)
```

```{r}
normsample30@meta.data
DimPlot (normsample30, group.by = "DF.classifications_0.25_0.09_29")
table   (normsample30$DF.classifications_0.25_0.09_29)
```

```{r}
# Subset to remove identified cell doublets
fltsample30 <- subset (normsample30, subset = DF.classifications_0.25_0.09_29 == 'Singlet')
fltsample30
table (fltsample30$DF.classifications_0.25_0.09_29)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample30)
rm (sweep.res.list_sample30)
rm (sweep.stats_sample30)
rm (bcmvn_sample30)
rm (annotations30)
rm (homotypic.prop30)
rm (nExp_poi30)
rm (nExp_poi.adj30)
rm (normsample30)
rm (ldat30)
```

```{r}
# Read in the data for trophoblast organoid sample two
sample31 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/3_Organoid/Undifferentiated/Stem_Cell_Derived/CT29")

# Read in the respective velocity data
          ldat31            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/TS_Organoids__CT29.loom")
colnames (ldat31$spliced)   <- paste        (substring (colnames (ldat31$spliced),   20, 35), sep = "")
colnames (ldat31$spliced)   <- paste                   (colnames (ldat31$spliced),   "-1",    sep = "")
colnames (ldat31$unspliced) <- paste        (substring (colnames (ldat31$unspliced), 20, 35), sep = "")
colnames (ldat31$unspliced) <- paste                   (colnames (ldat31$unspliced), "-1",    sep = "")
colnames (ldat31$ambiguous) <- paste        (substring (colnames (ldat31$ambiguous), 20, 35), sep = "")
colnames (ldat31$ambiguous) <- paste                   (colnames (ldat31$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample31 <- CreateSeuratObject (counts = sample31, min.cells = 1, min.features = 1, project = "CT29")

for (i in names (x = ldat31)) { sample31[[i]] <- CreateAssayObject(counts = ldat31[[i]]) }

# Add the sample metadata
sample31 <- AddMetaData (sample31, metadata = "hTSC.TOrg",        col.name = "Model"    )
sample31 <- AddMetaData (sample31, metadata = "CT29",             col.name = "ID"       )
sample31 <- AddMetaData (sample31, metadata = "UCT29",            col.name = "rawID"    )
sample31 <- AddMetaData (sample31, metadata = "Stem Cell",        col.name = "Source"   )
sample31 <- AddMetaData (sample31, metadata = "1",                col.name = "nSex"     )
sample31 <- AddMetaData (sample31, metadata = "Male",             col.name = "Sex"      )
sample31 <- AddMetaData (sample31, metadata = "1",                col.name = "Time"     )
sample31 <- AddMetaData (sample31, metadata = "Undifferentiated", col.name = "State"    )
sample31 <- AddMetaData (sample31, metadata = "hTSC Organoid",    col.name = "Tissue"   )
sample31 <- AddMetaData (sample31, metadata = "4",                col.name = "Batch"    )
sample31 <- AddMetaData (sample31, metadata = "4",                col.name = "Source"   )
sample31 <- AddMetaData (sample31, metadata = "in vitro",         col.name = "Type"     )
sample31 <- AddMetaData (sample31, metadata = "All",              col.name = "Gate"     )
sample31 <- AddMetaData (sample31, metadata = "in vitro",         col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample31 <- NormalizeData    (sample31)
sample31 <- CellCycleScoring (sample31, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample31)
dim   (sample31)
sample31@meta.data
```

```{r} 
# Add sample quality control metrics (% MT and % RP data)
sample31 [["percent.mt"]]   <- PercentageFeatureSet (sample31, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample31), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample31, slot   = 'counts')[ribo.genes, ]) / 
                               Matrix::colSums (x = GetAssayData (sample31, slot   = 'counts')              );
sample31 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample31, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample31, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample31, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample31@meta.data
```

```{r}
# Filter trophoblast organoid sample two
fltsample31 <- subset (sample31, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample31, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample31, features = c("percent.mt"))
VlnPlot (fltsample31, features = c("percent.ribo"))

# View the structure of sample two before and after filtering to observe the effect of filtering on this sample
   sample31
fltsample31
```

```{r}
normsample31 <- NormalizeData        ( fltsample31)
normsample31 <- ScaleData            (normsample31)
normsample31 <- FindVariableFeatures (normsample31, selection.method = "vst", nfeatures = 2000)
normsample31 <- RunPCA               (normsample31)
normsample31 <- RunUMAP              (normsample31, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample31 <- paramSweep_v3    (normsample31, PCs = 1:50, sct = FALSE)
sweep.stats_sample31    <- summarizeSweep   (sweep.res.list_sample31,  GT  = FALSE)
bcmvn_sample31          <- find.pK          (sweep.stats_sample31)
annotations31           <- normsample31@meta.data$ClusteringResults
homotypic.prop31        <- modelHomotypic   (annotations31)
nExp_poi31              <- round            (0.003 * length (normsample31@meta.data$ID))  ## Assuming 0.3% doublet formation rate
nExp_poi.adj31          <- round            (nExp_poi31 *    (1 - homotypic.prop31)    )
normsample31            <- doubletFinder_v3 (normsample31,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi31,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample31            <- doubletFinder_v3 (normsample31,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj31,
                                             reuse.pANN = "pANN_0.25_0.09_1",
                                             sct        = FALSE)
```

```{r}
normsample31@meta.data
DimPlot (normsample31, group.by = "DF.classifications_0.25_0.09_1")
table   (normsample31$DF.classifications_0.25_0.09_1)
```

```{r}
# Subset to remove identified cell doublets
fltsample31 <- subset (normsample31, subset = DF.classifications_0.25_0.09_1 == 'Singlet')
fltsample31
table (fltsample31$DF.classifications_0.25_0.09_1)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample31)
rm (sweep.res.list_sample31)
rm (sweep.stats_sample31)
rm (bcmvn_sample31)
rm (annotations31)
rm (homotypic.prop31)
rm (nExp_poi31)
rm (nExp_poi.adj31)
rm (normsample31)
rm (ldat31)
```

```{r}
# Read in the data for trophoblast organoid sample three
sample32 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/3_Organoid/Undifferentiated/Stem_Cell_Derived/CT30")

# Read in the respective velocity data
          ldat32            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/TS_Organoids__CT30.loom")
colnames (ldat32$spliced)   <- paste        (substring (colnames (ldat32$spliced),   20, 35), sep = "")
colnames (ldat32$spliced)   <- paste                   (colnames (ldat32$spliced),   "-1",    sep = "")
colnames (ldat32$unspliced) <- paste        (substring (colnames (ldat32$unspliced), 20, 35), sep = "")
colnames (ldat32$unspliced) <- paste                   (colnames (ldat32$unspliced), "-1",    sep = "")
colnames (ldat32$ambiguous) <- paste        (substring (colnames (ldat32$ambiguous), 20, 35), sep = "")
colnames (ldat32$ambiguous) <- paste                   (colnames (ldat32$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample32 <- CreateSeuratObject (counts = sample32, min.cells = 1, min.features = 1, project = "CT30")

for (i in names (x = ldat32)) { sample32[[i]] <- CreateAssayObject(counts = ldat32[[i]]) }

# Add the sample metadata
sample32 <- AddMetaData (sample32, metadata = "hTSC.TOrg",        col.name = "Model"    )
sample32 <- AddMetaData (sample32, metadata = "CT30",             col.name = "ID"       )
sample32 <- AddMetaData (sample32, metadata = "UCT30",            col.name = "rawID"    )
sample32 <- AddMetaData (sample32, metadata = "Stem Cell",        col.name = "Source"   )
sample32 <- AddMetaData (sample32, metadata = "0",                col.name = "nSex"     )
sample32 <- AddMetaData (sample32, metadata = "Female",           col.name = "Sex"      )
sample32 <- AddMetaData (sample32, metadata = "1",                col.name = "Time"     )
sample32 <- AddMetaData (sample32, metadata = "Undifferentiated", col.name = "State"    )
sample32 <- AddMetaData (sample32, metadata = "hTSC Organoid",    col.name = "Tissue"   )
sample32 <- AddMetaData (sample32, metadata = "4",                col.name = "Batch"    )
sample32 <- AddMetaData (sample32, metadata = "4",                col.name = "Source"   )
sample32 <- AddMetaData (sample32, metadata = "in vitro",         col.name = "Type"     )
sample32 <- AddMetaData (sample32, metadata = "All",              col.name = "Gate"     )
sample32 <- AddMetaData (sample32, metadata = "in vitro",         col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample32 <- NormalizeData    (sample32)
sample32 <- CellCycleScoring (sample32, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample32)
dim   (sample32)
sample32@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample32 [["percent.mt"]]   <- PercentageFeatureSet (sample32, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample32), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample32, slot   = 'counts')[ribo.genes, ]) / 
                               Matrix::colSums (x = GetAssayData (sample32, slot   = 'counts')              );
sample32 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample32, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample32, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample32, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample32@meta.data
```

```{r}
# Filter trophoblast organoid sample three
fltsample32 <- subset (sample32, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample32, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample32, features = c("percent.mt"))
VlnPlot (fltsample32, features = c("percent.ribo"))

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample32
fltsample32
```

```{r}
normsample32 <- NormalizeData        ( fltsample32)
normsample32 <- ScaleData            (normsample32)
normsample32 <- FindVariableFeatures (normsample32, selection.method = "vst", nfeatures = 2000)
normsample32 <- RunPCA               (normsample32)
normsample32 <- RunUMAP              (normsample32, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample32 <- paramSweep_v3    (normsample32, PCs = 1:50, sct = FALSE)
sweep.stats_sample32    <- summarizeSweep   (sweep.res.list_sample32,  GT  = FALSE)
bcmvn_sample32          <- find.pK          (sweep.stats_sample32)
annotations32           <- normsample32@meta.data$ClusteringResults
homotypic.prop32        <- modelHomotypic   (annotations32)
nExp_poi32              <- round            (0.006 * length (normsample32@meta.data$ID))  ## Assuming 0.6% doublet formation rate
nExp_poi.adj32          <- round            (nExp_poi32 *    (1 - homotypic.prop32)     )
normsample32            <- doubletFinder_v3 (normsample32,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi32,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample32            <- doubletFinder_v3 (normsample32,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj32,
                                             reuse.pANN = "pANN_0.25_0.09_4",
                                             sct        = FALSE)
```

```{r}
normsample32@meta.data
DimPlot (normsample32, group.by = "DF.classifications_0.25_0.09_4")
table   (normsample32$DF.classifications_0.25_0.09_4)
```

```{r}
# Subset to remove identified cell doublets
fltsample32 <- subset (normsample32, subset = DF.classifications_0.25_0.09_4 == 'Singlet')
fltsample32
table (fltsample32$DF.classifications_0.25_0.09_4)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample32)
rm (sweep.res.list_sample32)
rm (sweep.stats_sample32)
rm (bcmvn_sample32)
rm (annotations32)
rm (homotypic.prop32)
rm (nExp_poi32)
rm (nExp_poi.adj32)
rm (normsample32)
rm (ldat32)
```

### Differentiated

```{r}
# Read in the data for trophoblast organoid sample four
sample33 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/3_Organoid/Differentiated/Stem_Cell_Derived/CT27")

# Read in the respective velocity data
          ldat33            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/TS_Organoids__CT27_EVT.loom")
colnames (ldat33$spliced)   <- paste        (substring (colnames (ldat33$spliced),   24, 39), sep = "")
colnames (ldat33$spliced)   <- paste                   (colnames (ldat33$spliced),   "-1",    sep = "")
colnames (ldat33$unspliced) <- paste        (substring (colnames (ldat33$unspliced), 24, 39), sep = "")
colnames (ldat33$unspliced) <- paste                   (colnames (ldat33$unspliced), "-1",    sep = "")
colnames (ldat33$ambiguous) <- paste        (substring (colnames (ldat33$ambiguous), 24, 39), sep = "")
colnames (ldat33$ambiguous) <- paste                   (colnames (ldat33$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample33 <- CreateSeuratObject (counts = sample33, min.cells = 1, min.features = 1, project = "CT27")

for (i in names (x = ldat33)) { sample33[[i]] <- CreateAssayObject (counts = ldat33[[i]]) }

# Add the sample metadata
sample33 <- AddMetaData (sample33, metadata = "hTSC.TOrg",      col.name = "Model"    )
sample33 <- AddMetaData (sample33, metadata = "CT27",           col.name = "ID"       )
sample33 <- AddMetaData (sample33, metadata = "DCT27",          col.name = "rawID"    )
sample33 <- AddMetaData (sample33, metadata = "Stem Cell",      col.name = "Source"   )
sample33 <- AddMetaData (sample33, metadata = "0",              col.name = "nSex"     )
sample33 <- AddMetaData (sample33, metadata = "Female",         col.name = "Sex"      )
sample33 <- AddMetaData (sample33, metadata = "2",              col.name = "Time"     )
sample33 <- AddMetaData (sample33, metadata = "Differentiated", col.name = "State"    )
sample33 <- AddMetaData (sample33, metadata = "hTSC Organoid",  col.name = "Tissue"   )
sample33 <- AddMetaData (sample33, metadata = "5",              col.name = "Batch"    )
sample33 <- AddMetaData (sample33, metadata = "4",              col.name = "Source"   )
sample33 <- AddMetaData (sample33, metadata = "in vitro",       col.name = "Type"     )
sample33 <- AddMetaData (sample33, metadata = "All",            col.name = "Gate"     )
sample33 <- AddMetaData (sample33, metadata = "in vitro",       col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample33 <- NormalizeData    (sample33)
sample33 <- CellCycleScoring (sample33, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample33)
dim   (sample33)
sample33@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample33 [["percent.mt"]]   <- PercentageFeatureSet (sample33, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample33), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample33, slot   = 'counts')[ribo.genes, ]) / 
                               Matrix::colSums (x = GetAssayData (sample33, slot   = 'counts')              );
sample33 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample33, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample33, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample33, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample33@meta.data
```

```{r}
# Filter trophoblast organoid sample four
fltsample33 <- subset (sample33, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample33, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample33, features = c("percent.mt"))
VlnPlot (fltsample33, features = c("percent.ribo"))

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample33
fltsample33
```

```{r}
normsample33 <- NormalizeData        ( fltsample33)
normsample33 <- ScaleData            (normsample33)
normsample33 <- FindVariableFeatures (normsample33, selection.method = "vst", nfeatures = 2000)
normsample33 <- RunPCA               (normsample33)
normsample33 <- RunUMAP              (normsample33, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample33 <- paramSweep_v3    (normsample33, PCs = 1:50, sct = FALSE)
sweep.stats_sample33    <- summarizeSweep   (sweep.res.list_sample33,  GT  = FALSE)
bcmvn_sample33          <- find.pK          (sweep.stats_sample33)
annotations33           <- normsample33@meta.data$ClusteringResults
homotypic.prop33        <- modelHomotypic   (annotations33)
nExp_poi33              <- round            (0.006 * length (normsample33@meta.data$ID))  ## Assuming 0.6% doublet formation rate
nExp_poi.adj33          <- round            (nExp_poi33 *    (1 - homotypic.prop33)    )
normsample33            <- doubletFinder_v3 (normsample33,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi33,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample33            <- doubletFinder_v3 (normsample33,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj33,
                                             reuse.pANN = "pANN_0.25_0.09_4",
                                             sct        = FALSE)
```

```{r}
normsample33@meta.data
DimPlot (normsample33, group.by = "DF.classifications_0.25_0.09_4")
table   (normsample33$DF.classifications_0.25_0.09_4)
```

```{r}
# Subset to remove identified cell doublets
fltsample33 <- subset (normsample33, subset = DF.classifications_0.25_0.09_4 == 'Singlet')
fltsample33
table (fltsample33$DF.classifications_0.25_0.09_4)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample33)
rm (sweep.res.list_sample33)
rm (sweep.stats_sample33)
rm (bcmvn_sample33)
rm (annotations33)
rm (homotypic.prop33)
rm (nExp_poi33)
rm (nExp_poi.adj33)
rm (normsample33)
rm (ldat33)
```

```{r}
# Read in the data for trophoblast organoid sample five
sample34 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/3_Organoid/Differentiated/Stem_Cell_Derived/CT29")

# Read in the respective velocity data
          ldat34            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/TS_Organoids__CT29_EVT.loom")
colnames (ldat34$spliced)   <- paste        (substring (colnames (ldat34$spliced),   24, 39), sep = "")
colnames (ldat34$spliced)   <- paste                   (colnames (ldat34$spliced),   "-1",    sep = "")
colnames (ldat34$unspliced) <- paste        (substring (colnames (ldat34$unspliced), 24, 39), sep = "")
colnames (ldat34$unspliced) <- paste                   (colnames (ldat34$unspliced), "-1",    sep = "")
colnames (ldat34$ambiguous) <- paste        (substring (colnames (ldat34$ambiguous), 24, 39), sep = "")
colnames (ldat34$ambiguous) <- paste                   (colnames (ldat34$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample34 <- CreateSeuratObject (counts = sample34, min.cells = 1, min.features = 1, project = "CT29")

for (i in names (x = ldat34)) { sample34[[i]] <- CreateAssayObject (counts = ldat34[[i]]) }

# Add the sample metadata
sample34 <- AddMetaData (sample34, metadata = "hTSC.TOrg",      col.name = "Model"    )
sample34 <- AddMetaData (sample34, metadata = "CT29",           col.name = "ID"       )
sample34 <- AddMetaData (sample34, metadata = "DCT29",          col.name = "rawID"    )
sample34 <- AddMetaData (sample34, metadata = "Stem Cell",      col.name = "Source"   )
sample34 <- AddMetaData (sample34, metadata = "1",              col.name = "nSex"     )
sample34 <- AddMetaData (sample34, metadata = "Male",           col.name = "Sex"      )
sample34 <- AddMetaData (sample34, metadata = "2",              col.name = "Time"     )
sample34 <- AddMetaData (sample34, metadata = "Differentiated", col.name = "State"    )
sample34 <- AddMetaData (sample34, metadata = "hTSC Organoid",  col.name = "Tissue"   )
sample34 <- AddMetaData (sample34, metadata = "5",              col.name = "Batch"    )
sample34 <- AddMetaData (sample34, metadata = "4",              col.name = "Source"   )
sample34 <- AddMetaData (sample34, metadata = "in vitro",       col.name = "Type"     )
sample34 <- AddMetaData (sample34, metadata = "All",            col.name = "Gate"     )
sample34 <- AddMetaData (sample34, metadata = "in vitro",       col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample34 <- NormalizeData    (sample34)
sample34 <- CellCycleScoringv(sample34, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample34)
dim   (sample34)
sample34@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample34 [["percent.mt"]]   <- PercentageFeatureSet (sample34, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample34), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample34, slot   = 'counts')[ribo.genes, ]) / 
                               Matrix::colSums (x = GetAssayData (sample34, slot   = 'counts')              );
sample34 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample34, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample34, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample34, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample34@meta.data
```

```{r}
# Filter trophoblast organoid sample five
fltsample34 <- subset (sample34, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample34, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample34, features = c("percent.mt"))
VlnPlot (fltsample34, features = c("percent.ribo"))

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample34
fltsample34
```

```{r}
normsample34 <- NormalizeData        ( fltsample34)
normsample34 <- ScaleData            (normsample34)
normsample34 <- FindVariableFeatures (normsample34, selection.method = "vst", nfeatures = 2000)
normsample34 <- RunPCA               (normsample34)
normsample34 <- RunUMAP              (normsample34, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample34 <- paramSweep_v3    (normsample34, PCs = 1:50, sct = FALSE)
sweep.stats_sample34    <- summarizeSweep   (sweep.res.list_sample34,  GT  = FALSE)
bcmvn_sample34          <- find.pK          (sweep.stats_sample34)
annotations34           <- normsample34@meta.data$ClusteringResults
homotypic.prop34        <- modelHomotypic   (annotations34)
nExp_poi34              <- round            (0.012 * length (normsample34@meta.data$ID))  ## Assuming 1.2% doublet formation rate
nExp_poi.adj34          <- round            (nExp_poi34 *    (1 - homotypic.prop34)    )
normsample34            <- doubletFinder_v3 (normsample34,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi34,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample34            <- doubletFinder_v3 (normsample34,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj34,
                                             reuse.pANN = "pANN_0.25_0.09_16",
                                             sct        = FALSE)
```

```{r}
normsample34@meta.data
DimPlot (normsample34, group.by = "DF.classifications_0.25_0.09_16")
table   (normsample34$DF.classifications_0.25_0.09_16)
```

```{r}
# Subset to remove identified cell doublets
fltsample34 <- subset (normsample34, subset = DF.classifications_0.25_0.09_16 == 'Singlet')
fltsample34
table (fltsample34$DF.classifications_0.25_0.09_16)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample34)
rm (sweep.res.list_sample34)
rm (sweep.stats_sample34)
rm (bcmvn_sample34)
rm (annotations34)
rm (homotypic.prop34)
rm (nExp_poi34)
rm (nExp_poi.adj34)
rm (normsample34)
rm (ldat34)
```

```{r}
# Read in the data for trophoblast organoid sample six
sample35 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/3_Organoid/Differentiated/Stem_Cell_Derived/CT30")

# Read in the respective velocity data
          ldat35            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/TS_Organoids__CT30_EVT.loom")
colnames (ldat35$spliced)   <- paste        (substring (colnames (ldat35$spliced),   24, 39), sep = "")
colnames (ldat35$spliced)   <- paste                   (colnames (ldat35$spliced),   "-1",    sep = "")
colnames (ldat35$unspliced) <- paste        (substring (colnames (ldat35$unspliced), 24, 39), sep = "")
colnames (ldat35$unspliced) <- paste                   (colnames (ldat35$unspliced), "-1",    sep = "")
colnames (ldat35$ambiguous) <- paste        (substring (colnames (ldat35$ambiguous), 24, 39), sep = "")
colnames (ldat35$ambiguous) <- paste                   (colnames (ldat35$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample35 <- CreateSeuratObject (counts = sample35, min.cells = 1, min.features = 1, project = "CT30")

for (i in names (x = ldat35)) { sample35[[i]] <- CreateAssayObject(counts = ldat35[[i]]) }

# Add the sample metadata
sample35 <- AddMetaData (sample35, metadata = "hTSC.TOrg",      col.name = "Model"    )
sample35 <- AddMetaData (sample35, metadata = "CT30",           col.name = "ID"       )
sample35 <- AddMetaData (sample35, metadata = "DCT30",          col.name = "rawID"    )
sample35 <- AddMetaData (sample35, metadata = "Stem Cell",      col.name = "Source"   )
sample35 <- AddMetaData (sample35, metadata = "0",              col.name = "nSex"     )
sample35 <- AddMetaData (sample35, metadata = "Female",         col.name = "Sex"      )
sample35 <- AddMetaData (sample35, metadata = "2",              col.name = "Time"     )
sample35 <- AddMetaData (sample35, metadata = "Differentiated", col.name = "State"    )
sample35 <- AddMetaData (sample35, metadata = "hTSC Organoid",  col.name = "Tissue"   )
sample35 <- AddMetaData (sample35, metadata = "5",              col.name = "Batch"    )
sample35 <- AddMetaData (sample35, metadata = "4",              col.name = "Source"   )
sample35 <- AddMetaData (sample35, metadata = "in vitro",       col.name = "Type"     )
sample35 <- AddMetaData (sample35, metadata = "All",            col.name = "Gate"     )
sample35 <- AddMetaData (sample35, metadata = "in vitro",       col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample35 <- NormalizeData      (sample35)
sample35 <- CellCycleScoring   (sample35, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample35)
dim   (sample35)
sample35@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample35 [["percent.mt"]]   <- PercentageFeatureSet (sample35, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample35), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample35, slot   = 'counts')[ribo.genes, ]) / 
                               Matrix::colSums (x = GetAssayData (sample35, slot   = 'counts')              );
sample35 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample35, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample35, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample35, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample35@meta.data
```

```{r}
# Filter trophoblast organoid sample six
fltsample35 <- subset (sample35, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample35, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample35, features = c("percent.mt"))
VlnPlot (fltsample35, features = c("percent.ribo"))

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample35
fltsample35
```

```{r}
normsample35 <- NormalizeData        ( fltsample35)
normsample35 <- ScaleData            (normsample35)
normsample35 <- FindVariableFeatures (normsample35, selection.method = "vst", nfeatures = 2000)
normsample35 <- RunPCA               (normsample35)
normsample35 <- RunUMAP              (normsample35, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample35 <- paramSweep_v3    (normsample35, PCs = 1:50, sct = FALSE)
sweep.stats_sample35    <- summarizeSweep   (sweep.res.list_sample35,  GT  = FALSE)
bcmvn_sample35          <- find.pK          (sweep.stats_sample35)
annotations35           <- normsample35@meta.data$ClusteringResults
homotypic.prop35        <- modelHomotypic   (annotations35)
nExp_poi35              <- round            (0.018 * length (normsample35@meta.data$ID))  ## Assuming 1.8% doublet formation rate
nExp_poi.adj35          <- round            (nExp_poi35 *    (1 - homotypic.prop35)    )
normsample35            <- doubletFinder_v3 (normsample35,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi35,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample35            <- doubletFinder_v3 (normsample35,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj35,
                                             reuse.pANN = "pANN_0.25_0.09_36",
                                             sct        = FALSE)
```

```{r}
normsample35@meta.data
DimPlot (normsample35, group.by = "DF.classifications_0.25_0.09_36")
table   (normsample35$DF.classifications_0.25_0.09_36)
```

```{r}
# Subset to remove identified cell doublets
fltsample35 <- subset (normsample35, subset = DF.classifications_0.25_0.09_36 == 'Singlet')
fltsample35
table (fltsample35$DF.classifications_0.25_0.09_36)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample35)
rm (sweep.res.list_sample35)
rm (sweep.stats_sample35)
rm (bcmvn_sample35)
rm (annotations35)
rm (homotypic.prop35)
rm (nExp_poi35)
rm (nExp_poi.adj35)
rm (normsample35)
rm (ldat35)
```

# Patient Derived Organoid Dataset Prepation, Quality Control, and Sample Integration

## Read in the patient organoid data

### Undifferentiated

```{r}
# Read in the data for trophoblast organoid sample five
sample36 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/3_Organoid/Undifferentiated/Patient_Derived/PT100")

# Read in the respective velocity data
          ldat36            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/TS_Organoids__PT100.loom")
colnames (ldat36$spliced)   <- paste        (substring (colnames (ldat36$spliced),   21, 36), sep = "")
colnames (ldat36$spliced)   <- paste                   (colnames (ldat36$spliced),   "-1",    sep = "")
colnames (ldat36$unspliced) <- paste        (substring (colnames (ldat36$unspliced), 21, 36), sep = "")
colnames (ldat36$unspliced) <- paste                   (colnames (ldat36$unspliced), "-1",    sep = "")
colnames (ldat36$ambiguous) <- paste        (substring (colnames (ldat36$ambiguous), 21, 36), sep = "")
colnames (ldat36$ambiguous) <- paste                   (colnames (ldat36$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample36 <- CreateSeuratObject (counts = sample36, min.cells = 1, min.features = 1, project = "PT100")

for (i in names (x = ldat36)) { sample36[[i]] <- CreateAssayObject (counts = ldat36[[i]]) }

# Add the sample metadata
sample36 <- AddMetaData (sample36, metadata = "PD.TOrg",          col.name = "Model"    )
sample36 <- AddMetaData (sample36, metadata = "PT100",            col.name = "ID"       )
sample36 <- AddMetaData (sample36, metadata = "UPT100",           col.name = "rawID"    )
sample36 <- AddMetaData (sample36, metadata = "Patient",          col.name = "Source"   )
sample36 <- AddMetaData (sample36, metadata = "3",                col.name = "nSex"     )
sample36 <- AddMetaData (sample36, metadata = "Mixed",            col.name = "Sex"      )
sample36 <- AddMetaData (sample36, metadata = "1",                col.name = "Time"     )
sample36 <- AddMetaData (sample36, metadata = "Undifferentiated", col.name = "State"    )
sample36 <- AddMetaData (sample36, metadata = "Patient Organoid", col.name = "Tissue"   )
sample36 <- AddMetaData (sample36, metadata = "4",                col.name = "Batch"    )
sample36 <- AddMetaData (sample36, metadata = "4",                col.name = "Source"   )
sample36 <- AddMetaData (sample36, metadata = "in vitro",         col.name = "Type"     )
sample36 <- AddMetaData (sample36, metadata = "All",              col.name = "Gate"     )
sample36 <- AddMetaData (sample36, metadata = "in vitro",         col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample36 <- NormalizeData    (sample36)
sample36 <- CellCycleScoring (sample36, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample36)
dim   (sample36)
sample36@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample36 [["percent.mt"]]   <- PercentageFeatureSet (sample36, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample36), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample36, slot   = 'counts')[ribo.genes, ]) / 
                               Matrix::colSums (x = GetAssayData (sample36, slot   = 'counts')              );
sample36 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample36, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample36, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample36, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample36@meta.data
```

```{r}
# Filter trophoblast organoid sample five
fltsample36 <- subset (sample36, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample36, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample36, features = c("percent.mt"))
VlnPlot (fltsample36, features = c("percent.ribo"))

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample36
fltsample36
```

```{r}
normsample36 <- NormalizeData        ( fltsample36)
normsample36 <- ScaleData            (normsample36)
normsample36 <- FindVariableFeatures (normsample36, selection.method = "vst", nfeatures = 2000)
normsample36 <- RunPCA               (normsample36)
normsample36 <- RunUMAP              (normsample36, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample36 <- paramSweep_v3    (normsample36, PCs = 1:50, sct = FALSE)
sweep.stats_sample36    <- summarizeSweep   (sweep.res.list_sample36,  GT  = FALSE)
bcmvn_sample36          <- find.pK          (sweep.stats_sample36)
annotations36           <- normsample36@meta.data$ClusteringResults
homotypic.prop36        <- modelHomotypic   (annotations36)
nExp_poi36              <- round            (0.028 * length (normsample36@meta.data$ID))  ## Assuming 2.8% doublet formation rate
nExp_poi.adj36          <- round            (nExp_poi36 *    (1 - homotypic.prop36)    )
normsample36            <- doubletFinder_v3 (normsample36,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi36,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample36            <- doubletFinder_v3 (normsample36,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj36,
                                             reuse.pANN = "pANN_0.25_0.09_88",
                                             sct        = FALSE)
```

```{r}
normsample36@meta.data
DimPlot (normsample36, group.by = "DF.classifications_0.25_0.09_88")
table   (normsample36$DF.classifications_0.25_0.09_88)
```

```{r}
# Subset to remove identified cell doublets
fltsample36 <- subset (normsample36, subset = DF.classifications_0.25_0.09_88 == 'Singlet')
fltsample36
table (fltsample36$DF.classifications_0.25_0.09_88)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample36)
rm (sweep.res.list_sample36)
rm (sweep.stats_sample36)
rm (bcmvn_sample36)
rm (annotations36)
rm (homotypic.prop36)
rm (nExp_poi36)
rm (nExp_poi.adj36)
rm (normsample36)
rm (ldat36)
```

### Differentiated

```{r}
# Read in the data for trophoblast organoid sample six
sample37 <- Read10X (data.dir = "/scratch/st-aberista-1/MJS/R/3_Organoid/Differentiated/Patient_Derived/PT100")

# Read in the respective velocity data
          ldat37            <- ReadVelocity ("/scratch/st-aberista-1/MJS/R/5_scVelo/TS_Organoids__PT100_EVT.loom")
colnames (ldat37$spliced)   <- paste        (substring (colnames (ldat37$spliced),   24, 39), sep = "")
colnames (ldat37$spliced)   <- paste                   (colnames (ldat37$spliced),   "-1",    sep = "")
colnames (ldat37$unspliced) <- paste        (substring (colnames (ldat37$unspliced), 24, 39), sep = "")
colnames (ldat37$unspliced) <- paste                   (colnames (ldat37$unspliced), "-1",    sep = "")
colnames (ldat37$ambiguous) <- paste        (substring (colnames (ldat37$ambiguous), 24, 39), sep = "")
colnames (ldat37$ambiguous) <- paste                   (colnames (ldat37$ambiguous), "-1",    sep = "")

# Convert to a seurat object for analysis
sample37 <- CreateSeuratObject (counts = sample37, min.cells = 1, min.features = 1, project = "PT100")

for (i in names (x = ldat37)) { sample37[[i]] <- CreateAssayObject(counts = ldat37[[i]]) }

# Add the sample metadata
sample37 <- AddMetaData (sample37, metadata = "PD.TOrg",          col.name = "Model"    )
sample37 <- AddMetaData (sample37, metadata = "PT100",            col.name = "ID"       )
sample37 <- AddMetaData (sample37, metadata = "DPT100",           col.name = "rawID"    )
sample37 <- AddMetaData (sample37, metadata = "Patient",          col.name = "Source"   )
sample37 <- AddMetaData (sample37, metadata = "3",                col.name = "nSex"     )
sample37 <- AddMetaData (sample37, metadata = "Mixed",            col.name = "Sex"      )
sample37 <- AddMetaData (sample37, metadata = "2",                col.name = "Time"     )
sample37 <- AddMetaData (sample37, metadata = "Differentiated",   col.name = "State"    )
sample37 <- AddMetaData (sample37, metadata = "Patient Organoid", col.name = "Tissue"   )
sample37 <- AddMetaData (sample37, metadata = "5",                col.name = "Batch"    )
sample37 <- AddMetaData (sample37, metadata = "4",                col.name = "Source"   )
sample37 <- AddMetaData (sample37, metadata = "in vitro",         col.name = "Type"     )
sample37 <- AddMetaData (sample37, metadata = "All",              col.name = "Gate"     )
sample37 <- AddMetaData (sample37, metadata = "in vitro",         col.name = "Trimester")

# Normalize the sample data and add cell cycle scores
sample37 <- NormalizeData    (sample37)
sample37 <- CellCycleScoring (sample37, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Observe the sample data for quality control
class (sample37)
dim   (sample37)
sample37@meta.data
```

```{r}
# Add sample quality control metrics (% MT and % RP data)
sample37 [["percent.mt"]]   <- PercentageFeatureSet (sample37, pattern = "^MT-")
ribo.genes                  <- grep (pattern = "^RP[SL][[:digit:]]", x = rownames (sample37), value = TRUE);
percent.ribo                <- Matrix::colSums (x = GetAssayData (sample37, slot   = 'counts')[ribo.genes, ]) / 
                               Matrix::colSums (x = GetAssayData (sample37, slot   = 'counts')              );
sample37 [['percent.ribo']] <- percent.ribo

# Visualize QC metrics as a violin plot, cells with >20% mtRNA are taken as poor quality and removed 
VlnPlot (sample37, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (sample37, features = c("percent.mt"),   group.by = "ID")
VlnPlot (sample37, features = c("percent.ribo"), group.by = "ID")

# View sample metadata
sample37@meta.data
```

```{r}
# Filter trophoblast organoid sample six
fltsample37 <- subset (sample37, subset = nFeature_RNA > 500 & percent.mt < 20)

# Visualize QC metrics as a violin plot again (confirm filtering success)
VlnPlot (fltsample37, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot (fltsample37, features = c("percent.mt"))
VlnPlot (fltsample37, features = c("percent.ribo"))

# View the structure of sample one before and after filtering to observe the effect of filtering on this sample
   sample37
fltsample37
```

```{r}
normsample37 <- NormalizeData        ( fltsample37)
normsample37 <- ScaleData            (normsample37)
normsample37 <- FindVariableFeatures (normsample37, selection.method = "vst", nfeatures = 2000)
normsample37 <- RunPCA               (normsample37)
normsample37 <- RunUMAP              (normsample37, dims = 1:50)
```

```{r}
# Remove doublets from this sample
sweep.res.list_sample37 <- paramSweep_v3    (normsample37, PCs = 1:50, sct = FALSE)
sweep.stats_sample37    <- summarizeSweep   (sweep.res.list_sample37,  GT  = FALSE)
bcmvn_sample37          <- find.pK          (sweep.stats_sample37)
annotations37           <- normsample37@meta.data$ClusteringResults
homotypic.prop37        <- modelHomotypic   (annotations37)
nExp_poi37              <- round            (0.007 * length (normsample37@meta.data$ID))  ## Assuming 0.7% doublet formation rate
nExp_poi.adj37          <- round            (nExp_poi37 *    (1 - homotypic.prop37)    )
normsample37            <- doubletFinder_v3 (normsample37,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi37,
                                             reuse.pANN = FALSE,
                                             sct        = FALSE)
normsample37            <- doubletFinder_v3 (normsample37,
                                             PCs        = 1:50,
                                             pN         = 0.25,
                                             pK         = 0.09,
                                             nExp       = nExp_poi.adj37,
                                             reuse.pANN = "pANN_0.25_0.09_5",
                                             sct        = FALSE)
```

```{r}
normsample37@meta.data
DimPlot (normsample37, group.by = "DF.classifications_0.25_0.09_5")
table   (normsample37$DF.classifications_0.25_0.09_5)
```

```{r}
# Subset to remove identified cell doublets
fltsample37 <- subset (normsample37, subset = DF.classifications_0.25_0.09_5 == 'Singlet')
fltsample37
table (fltsample37$DF.classifications_0.25_0.09_5)
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (sample37)
rm (sweep.res.list_sample37)
rm (sweep.stats_sample37)
rm (bcmvn_sample37)
rm (annotations37)
rm (homotypic.prop37)
rm (nExp_poi37)
rm (nExp_poi.adj37)
rm (normsample37)
rm (ldat37)
```

# Merge the Beristain and Vento-Tormo Placenta and Decidua datasets

```{r # Run in Terminal}
# Merge the twenty-one placental villi seurat objects into one large data file
new_interface <- merge (fltsample1, y = c(fltsample2 ,
                                          fltsample3 ,
                                          fltsample4 ,
                                          fltsample5 ,
                                          fltsample6 ,
                                          fltsample7 ,
                                          fltsample8 ,
                                          fltsample9 ,
                                          fltsample10,
                                          fltsample11,
                                          fltsample12,
                                          fltsample13,
                                          fltsample14,
                                          fltsample15,
                                          fltsample16,
                                          fltsample17,
                                          fltsample18,
                                          fltsample19,
                                          fltsample20,
                                          fltsample21,
                                          fltsample22,
                                          fltsample23,
                                          fltsample24,
                                          fltsample25,
                                          fltsample26,
                                          fltsample27,
                                          fltsample28,
                                          fltsample29,
                                          fltsample30,
                                          fltsample31,
                                          fltsample32,
                                          fltsample33,
                                          fltsample34,
                                          fltsample35,
                                          fltsample36,
                                          fltsample37 ),
                        add.cell.ids  = c("1",  "2",  "3",  "4",  "5",  "6",  "7",                # Beristain   Placenta  Datasets
                                          "8",  "9",  "10", "11", "12", "13", "14", "15", "16",   # Vento-Tormo Placenta  Datasets
                                          "17", "18", "19", "22", "23", "25",                     # Vento-Tormo Decidua   Datasets
                                          "26", "27", "28", "29",                                 # Vento-Tormo Decidua   Datasets
                                          "20", "21", "24",                                       # Vento-Tormo Blood     Datasets
                                          "30", "31", "32", "33", "34", "35",                     # Beristain   hTSC-TOrg Datasets
                                          "36", "37"),                                            # Beristain   PD-TOrg   Datasets
                        merge.data    = TRUE,
                        project       = "10X_Integrated_Interface")

# View the merged placental villi seurat object for visual inspection before downstream analyses
new_interface
```

```{r}
# Remove files no longer needed before proceeding (conserves computational space)
rm (fltsample1)
rm (fltsample2)
rm (fltsample3)
rm (fltsample4)
rm (fltsample5)
rm (fltsample6)
rm (fltsample7)
rm (fltsample8)
rm (fltsample9)
rm (fltsample10)
rm (fltsample11)
rm (fltsample12)
rm (fltsample13)
rm (fltsample14)
rm (fltsample15)
rm (fltsample16)
rm (fltsample17)
rm (fltsample18)
rm (fltsample19)
rm (fltsample20)
rm (fltsample21)
rm (fltsample22)
rm (fltsample23)
rm (fltsample24)
rm (fltsample25)
rm (fltsample26)
rm (fltsample27)
rm (fltsample28)
rm (fltsample29)
rm (fltsample30)
rm (fltsample31)
rm (fltsample32)
rm (fltsample33)
rm (fltsample34)
rm (fltsample35)
rm (fltsample36)
rm (fltsample37)
rm (i)
rm (g2m.genes)
rm (s.genes)
rm (ribo.genes)
rm (percent.ribo)
```

```{r}
# View the merged sample data stratified by sample ID
table (new_interface$ID)
```

```{r}
# View the dimensionality of the merged dataset
dim (new_interface)
```

```{r}
# Visualize the distribution of cell cycle markers across the merged dataset
RidgePlot (new_interface, features = c("PCNA", "TOP2A", "MCM6", "CDK1", "MKI67", "CCNA2"), ncol = 3)
```

```{r}
# Add metadata for the cell cycle difference scores, this will allow us to normalize for any variation present within a cell cycle stage but not for variation that exists between cell cycle stages (allows us to still use cell proliferation as an indicator for cell clustering without confounding variance of cell cycle gene expression changes that exist within distinct cell cycle states)
new_interface$CC.Difference <- new_interface$S.Score - new_interface$G2M.Score
```

## Visualize metadata and mitochondrial DNA information

```{r}
# View the merged metadata
new_interface@meta.data
```

```{r}
# Visualize final dataset QC metrics using violoin plots
VlnPlot (object   = new_interface,
         features = c("nCount_RNA", "nFeature_RNA", "percent.mt", "percent.ribo"),
         group.by =   "ID",
         ncol     = 2)
```

```{r}
# FeatureScatter to visualize feature-feature relationships
P1 <- FeatureScatter (object = new_interface, feature1 = "nCount_RNA", feature2 = "percent.mt"  ) + NoLegend()
P2 <- FeatureScatter (object = new_interface, feature1 = "nCount_RNA", feature2 = "percent.ribo") + NoLegend()
P3 <- FeatureScatter (object = new_interface, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + NoLegend()

CombinePlots (plots = list (P1, P2, P3))
rm           (P1)
rm           (P2)
rm           (P3)
```

## Integrate All Datasets into One Seurat File

```{r}
# Integrate the data based on the "Batch" variable
new_interface.list <- SplitObject (new_interface, split.by = "Batch")
new_interface.list <- new_interface.list [ c("1", "2", "3", "4", "5") ]

rm (new_interface)
```

```{r}
# Normalize the data and regress out variation resulting from the % MT and CC Difference variables
for (i in 1:length (new_interface.list)) {

new_interface.list[[i]] <- SCTransform (new_interface.list[[i]],
                                        verbose         = FALSE,
                                        vars.to.regress = c("percent.mt",
                                                            "CC.Difference"))
                                          }
```

```{r}
# Increase global max size to sustain heavy computational loads
options (future.globals.maxSize = 10000000000000000000000000000000000000000000000000000000000000000000000000000000)
```

```{r}
# Integrate the data
new_interface.features <- SelectIntegrationFeatures (object.list     = new_interface.list, nfeatures = 3000)

new_interface.list     <- PrepSCTIntegration        (object.list     = new_interface.list,
                                                     anchor.features = new_interface.features,
                                                     verbose         = FALSE)
``` 

```{r}
# Integrate the data (continued)
new_interface.anchors <- FindIntegrationAnchors (object.list          = new_interface.list,
                                                 normalization.method = "SCT",
                                                 anchor.features      = new_interface.features,
                                                 verbose              = FALSE)
rm (new_interface.list)
rm (new_interface.features)
```

```{r}
# Integrate the data (continued)
AllData  <- IntegrateData (anchorset            = new_interface.anchors,
                           normalization.method = "SCT",
                           verbose              = FALSE)
rm (new_interface.anchors)
```

```{r}
# Run principal component analysis on integrated data
 AllData  <- RunPCA (object         = AllData,
                     do.print       = TRUE,
                     pcs.print      = 1:5,
                     features.print = 5)
```

```{r}
# Visualize explained variation as a function of the number of principal components
ElbowPlot (AllData, ndims = 50)
```

```{r}
# Principal component analysis visualiztion, colour-coded by selected metadata variables
PCAPlot (AllData, group.by = "ID")
PCAPlot (AllData, group.by = "GA")
PCAPlot (AllData, group.by = "Sex")
PCAPlot (AllData, group.by = "Type")
PCAPlot (AllData, group.by = "Phase")
PCAPlot (AllData, group.by = "Source")
```

```{r}
# Additional data visualization
VizDimLoadings (AllData, dims = 1:4)
```

```{r}
# Save the raw data
save (AllData, file = "Model_Comparison_Data_Velocities.Rdata")
```

```{r}
AllData
```

```{r}
# Cluster the maternal fetal interface dataset
resolution.range     <- seq (from = 0, to = 1, by = 0.1)

AllData <- FindNeighbors (AllData)
AllData <- FindClusters  (AllData, resolution = resolution.range)
```

```{r}
# Select maternal fetal interface organoid cluster resolution
clustree (AllData, prefix = "integrated_snn_res.")
```

```{r}
AllData <- FindClusters (AllData, resolution = 0.90)
# Reduce maternal fetal interface data dimensionality
AllData <- RunUMAP      (AllData, dims       = 1:50)
AllData <- RunTSNE      (AllData, dims       = 1:50)
```

```{r}
# Visualize the maternal fetal interface data by TSNE
TSNEPlot    (AllData, label = TRUE,  pt.size = 2) + NoLegend ()
# Visualize the maternal fetal interface data by UMAP
UMAPPlot    (AllData, label = TRUE,  pt.size = 2) + NoLegend ()

FeaturePlot (AllData, features = c("nFeature_RNA"), min.cutoff = 0, pt.size = 2, order = TRUE, cols = c("grey", "red"))
FeaturePlot (AllData, features = c("nCount_RNA"),   min.cutoff = 0, pt.size = 2, order = TRUE, cols = c("grey", "red"))
FeaturePlot (AllData, features = c("percent.mt"),   min.cutoff = 0, pt.size = 2, order = TRUE, cols = c("grey", "red"))
FeaturePlot (AllData, features = c("percent.ribo"), min.cutoff = 0, pt.size = 2, order = TRUE, cols = c("grey", "red"))
```

## Isolate and Purify for Trophoblasts

```{r}
AllData[[]]
```

```{r}
DefaultAssay (AllData) <- "RNA"

# Subset cells expressing known trophoblast lineage and gene markers
vel_trophoblasts <- subset (AllData,   subset =  KRT7      > 0 |
                                                 EGFR      > 0 |
                                                 TP63      > 0 |
                                                 TEAD4     > 0 |
                                                 TFAP2A    > 0 |
                                                 TFAP2C    > 0 |
                                                 GATA2     > 0 |
                                                 GATA3     > 0 |
                                                `HLA-G`    > 0 |
                                                `ERVFRD-1` > 0 | Type == 'in vitro')
vel_trophoblasts
rm (AllData)
```

```{r}
DefaultAssay  (vel_trophoblasts) <- "RNA"

# Subset to remove cells expressing genes that are not expressed by trophoblasts (remove any non-trophoblast contamination)
trophoblasts <- subset (vel_trophoblasts, subset = VIM       == 0 &
                                                   WARS      == 0 &
                                                   PTPRC     == 0 &
                                                   DCN       == 0 &
                                                   CD34      == 0 &
                                                   CD14      == 0 &
                                                   CD86      == 0 &
                                                   CD163     == 0 &
                                                   NKG7      == 0 &
                                                   KLRD1     == 0 &
                                                  `HLA-DPA1` == 0 &
                                                  `HLA-DPB1` == 0 &
                                                  `HLA-DRA`  == 0 &
                                                  `HLA-DRB1` == 0 &
                                                  `HLA-DRB5` == 0 &
                                                  `HLA-DQA1` == 0 | Type == 'in vitro' )
DefaultAssay  (trophoblasts) <- "integrated"
rm            (vel_trophoblasts)
```

```{r}
# Cluster the trophoblast dataset
resolution.range <- seq           (from = 0, to = 1, by = 0.05)
#redo find neighbours/clusters
trophoblasts     <- FindNeighbors (trophoblasts)
trophoblasts     <- FindClusters  (trophoblasts, resolution = resolution.range)
trophoblasts
```

```{r}
# Select trophoblast cluster resolution
clustree (trophoblasts, prefix = "integrated_snn_res.")
```

```{r}
# Visualize explained variation as a function of the number of principal components
ElbowPlot (trophoblasts, ndims = 50)
```

```{r}
trophoblasts <- FindClusters (         trophoblasts, resolution = 0.375)
# Reduce trophoblast data dimensionality
trophoblasts <- RunUMAP      (object = trophoblasts, reduction  = "pca", dims = 1:30)
trophoblasts <- RunTSNE      (object = trophoblasts, reduction  = "pca", dims = 1:30)
trophoblasts
```

```{r}
# Visualize the trophoblast data by TSNE
TSNEPlot    (trophoblasts,                     label = TRUE, pt.size = 3) + NoLegend ()
# Visualize the trophoblast data by UMAP
UMAPPlot    (trophoblasts,                     label = TRUE, pt.size = 3) + NoLegend ()
UMAPPlot    (trophoblasts, split.by = 'Model', label = TRUE, pt.size = 3) + NoLegend ()

# Visualize expression of trophoblast QC metrics by UMAP
FeaturePlot (trophoblasts, features = c("nFeature_RNA"), min.cutoff = 0, order = TRUE, pt.size = 3, cols = c("grey", "red"))
FeaturePlot (trophoblasts, features = c("nCount_RNA"),   min.cutoff = 0, order = TRUE, pt.size = 3, cols = c("grey", "red"))
FeaturePlot (trophoblasts, features = c("percent.mt"),   min.cutoff = 0, order = TRUE, pt.size = 3, cols = c("grey", "red"))
```

```{r}
Trophoblasts <- subset (trophoblasts, idents = c("0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10"))
Trophoblasts
```

```{r}
# Cluster the trophoblast dataset
resolution.range <- seq           (from = 0, to = 1, by = 0.05)
#redo find neighbours/clusters
Trophoblasts     <- RunPCA        (Trophoblasts, npcs       = 50, verbose = FALSE)
Trophoblasts     <- FindNeighbors (Trophoblasts)
Trophoblasts     <- FindClusters  (Trophoblasts, resolution = resolution.range)
```

```{r}
# Select trophoblast cluster resolution
clustree (Trophoblasts, prefix = "integrated_snn_res.")
```

```{r}
# Visualize explained variation as a function of the number of principal components
ElbowPlot (Trophoblasts, ndims = 50)
```

```{r}
Trophoblasts <- FindClusters (Trophoblasts, resolution = 0.375)

# Reduce trophoblast data dimensionality
Trophoblasts <- RunUMAP      (Trophoblasts, reduction  = "pca", dims = 1:30)
Trophoblasts <- RunTSNE      (Trophoblasts, reduction  = "pca", dims = 1:30)
Trophoblasts
```

```{r}
# Visualize the trophoblast data by TSNE
TSNEPlot    (Trophoblasts,                     label = TRUE, pt.size = 3) + NoLegend ()
# Visualize the trophoblast data by UMAP
UMAPPlot    (Trophoblasts,                     label = TRUE, pt.size = 3) + NoLegend ()
UMAPPlot    (Trophoblasts, split.by = 'Model', label = TRUE, pt.size = 3) + NoLegend ()

# Visualize expression of trophoblast QC metrics by UMAP
FeaturePlot (Trophoblasts, features = c("nFeature_RNA"), min.cutoff = 0, order = TRUE, pt.size = 3, cols = c("grey", "red"))
FeaturePlot (Trophoblasts, features = c("nCount_RNA"),   min.cutoff = 0, order = TRUE, pt.size = 3, cols = c("grey", "red"))
FeaturePlot (Trophoblasts, features = c("percent.mt"),   min.cutoff = 0, order = TRUE, pt.size = 3, cols = c("grey", "red"))
```

```{r}
DefaultAssay  (object = Trophoblasts) <- "RNA"

Trophoblast.markers <- FindAllMarkers (object   = Trophoblasts,
                                       only.pos = TRUE,
                                       min.pct  = 0.25, 
                                       test.use = "MAST")
```

```{r, Table S3A}
Trophoblast.markers %>% group_by (cluster) %>% top_n (20, avg_log2FC)
```

```{r}
# Label all trophoblast clusters by cell type
       new.cluster.ids       <- c("CTB 1",     # Cytotrophoblast state 1
                                  "CTB 4",     # Cytotrophoblast state 4
                                  "TSC",       # hTSC-enriched state
                                  "CTB 2",     # Cytotrophoblast state 2
                                  "cCTB 1",    # Column cytotrophoblast state 1
                                  "EVT",       # Extravillous trophoblast state
                                  "cCTB 3",    # Column cytotrophoblast state 3
                                  "SCTp 1",    # Syncytiotrophoblast precursor state 1
                                  "SCTp 2",    # Syncytiotrophoblast precursor state 2
                                  "CTB 3",     # Cytotrophoblast state 3
                                  "cCTB 2")    # Column cytotrophoblast state 2

names (new.cluster.ids)      <- levels       (Trophoblasts)
Trophoblasts$seurat_clusters <- Idents       (Trophoblasts)
Trophoblasts                 <- RenameIdents (Trophoblasts, new.cluster.ids)

UMAPPlot (Trophoblasts, label = TRUE, pt.size = 3) + NoLegend ()
```

```{r}
# Save the integrated villous trophoblast and patient organoid and hTSC organoid dataset
save (Trophoblasts, file = "Trophoblast_Data.Rdata")
```